---
title: "lungcancer_20200814"
author: "zhangjingxiang"
date: "2020/8/14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#设置全局图片的大小
knitr::opts_chunk$set(fig.width = 20,fig.height=10, collapse = TRUE)
library(vegan)
library(picante)
library(phyloseq)
library(ggplot2)
library(GUniFrac)
library(plyr)
library(ggpubr)
library(microbiome)
library(reshape2)
library(plotly)
library(patchwork)
library(VennDiagram)
library(biomformat)
library(RColorBrewer)
library(psych)
library(corrplot)
library(ggrepel)
library(Rtsne)
library(igraph)
library(plyr)
library(pROC)
library(tidyverse)
library(caret)
library(randomForest)
library(AUCRF)
library(tableone)
library(DirichletMultinomial)
library(car)
library(reltools)
library(minpack.lm)
library(Hmisc)
library(DescTools)
library(vcdExtra)
library(ggalluvial)
library(ComplexHeatmap)
library(circlize)
library(RCy3)
library(brainGraph)

windowsFonts(A=windowsFont("微软雅黑"))
```

# 读取数据
```{r}
## 读取数据表格
TableOneData<-read.table("TableOne_data.txt",sep = "\t",header = TRUE,row.names = 1)
## 过滤掉没有测序数据且非肺癌的患者信息
TableOneData<-subset(TableOneData,data=="yes"&diagnosis%in%c("Cancer","Health"))
#读取由QIIME生成的biom文件
OtuTable.biom = read_biom("feature-table.10000.biom")
#将biom对象转化成矩阵对象
OtuTable<-as(biom_data(OtuTable.biom), "matrix")
#读取TAX
TaxTable<-read.table("silva.taxonomy.txt",sep="\t",row.names = 1,header = TRUE,stringsAsFactors = FALSE)
TaxTable <- as.matrix(TaxTable)
#读取metadata
MetaTable<-read.table("metadata_new.txt",sep="\t",row.names = 1,header = TRUE)
TableOneData<-read.table("TableOne_data.txt",sep = "\t",header = TRUE,row.names = 1)
TableOneData$index<-row.names(TableOneData)
MetaTable<-merge(MetaTable,TableOneData,by = "index")
rownames(MetaTable)<-MetaTable$ID
#生成phyloseq对象
OTU = otu_table(OtuTable, taxa_are_rows = TRUE)
samples = sample_data(MetaTable)
TAX = tax_table(TaxTable)
LungCancer <- phyloseq(OTU, TAX, samples)
LungCancer
```


# 过滤数据
```{r}
#过滤掉未确诊或非肺癌患者
LungCancer_TURE<-subset_samples(LungCancer,diagnosis%in%c("Cancer","Health"))

#过滤掉肺癌患者的健康侧BAL
LungCancer_TURE<-subset_samples(LungCancer_TURE,!(BALgroup=="HS"))

#选择腺癌患者和对照组
LungCancer_Normalized<-subset_samples(LungCancer_TURE, type%in%c("腺癌","Health"))

#过滤掉使用过抗生素的人群
LungCancer_Normalized<-subset_samples(LungCancer_Normalized,抗生素使用=="N")
#过滤掉reads数少于10的ASV
LungCancer_Normalized <- filter_taxa(LungCancer_Normalized,function(x) sum(x) > 10 , TRUE )
#过滤掉使用过抗生素的人群
LungCancer_Normalized_100<- filter_taxa(LungCancer_Normalized,function(x) sum(x) > 100 , TRUE )
```


# α多样性
```{r}

alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Shannon")
LungCancer_Normalized_BAL <- subset_samples(LungCancer_Normalized, sample =="BAL")

my_comparisons <- list(c("Cancer", "Health"))

p<-plot_richness(LungCancer_Normalized_BAL, "group", "group",nrow = 1,measures=alpha_meas)


P_BAL <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)

#alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Shannon")
LungCancer_Normalized_Oroph <- subset_samples(LungCancer_Normalized, sample =="Oroph")

my_comparisons <- list(c("Cancer", "Health"))

p<-plot_richness(LungCancer_Normalized_Oroph, "group", "group",nrow = 1,measures=alpha_meas)


P_Oroph <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Oroph")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)
#alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Shannon")
LungCancer_Normalized_Saliva <- subset_samples(LungCancer_Normalized, sample =="Saliva")

my_comparisons <- list(c("Cancer", "Health"))

p<-plot_richness(LungCancer_Normalized_Saliva, "group", "group",nrow = 1,measures=alpha_meas)


P_Saliva <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Saliva")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)


#alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Shannon")
LungCancer_Normalized_Nasal <- subset_samples(LungCancer_Normalized, sample =="Nasal")

my_comparisons <- list(c("Cancer", "Health"))

p<-plot_richness(LungCancer_Normalized_Nasal, "group", "group",nrow = 1,measures=alpha_meas)


P_Nasal <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Nasal")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)



#alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Shannon")
LungCancer_Normalized_Fecal <- subset_samples(LungCancer_Normalized, sample =="Fecal")

my_comparisons <- list(c("Cancer", "Health"))

p<-plot_richness(LungCancer_Normalized_Fecal, "group", "group",nrow = 1,measures=alpha_meas)


P_Fecal <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Fecal")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)
P_BAL+P_Oroph+P_Saliva+P_Nasal+P_Fecal
```



# β多样性
```{r}
set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BAL, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BAL, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BAL, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-17
pc2<-7


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,shape=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


set.seed(2020)
otu.adonis<-adonis(dis_bray ~ group, data = data.frame(sample_data(LungCancer_Normalized_BAL)))



#P_BAL<-P_BAL+geom_text(aes(x = 0.05,y = 0.35,label = paste("p-value = ",otu.adonis$aov.tab$`Pr(>F)`[1],sep = "")),size = 6,hjust = 0)


########################################################Oroph

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Oroph, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Oroph, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Oroph, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-25.9
pc2<-10.6


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_Oroph<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,shape=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of Oroph microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


set.seed(2020)
otu.adonis<-adonis(dis_bray ~ group, data = data.frame(sample_data(LungCancer_Normalized_Oroph)))



#P_Oroph<-P_Oroph+geom_text(aes(x = 0.05,y = 0.35,label = paste("p-value = ",otu.adonis$aov.tab$`Pr(>F)`[1],sep = "")),size = 6,hjust = 0)




#Saliva

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Saliva, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Saliva, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Saliva, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-15.6
pc2<-10.1


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_Saliva<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,shape=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of Saliva microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


set.seed(2020)
otu.adonis<-adonis(dis_bray ~ group, data = data.frame(sample_data(LungCancer_Normalized_Saliva)))



#P_Saliva<-P_Saliva+geom_text(aes(x = 0.05,y = 0.35,label = paste("p-value = ",otu.adonis$aov.tab$`Pr(>F)`[1],sep = "")),size = 6,hjust = 0)





#Nasal

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Nasal, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Nasal, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Nasal, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-20.5
pc2<-12.9


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_Nasal<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,shape=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of Nasal microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


set.seed(2020)
otu.adonis<-adonis(dis_bray ~ group, data = data.frame(sample_data(LungCancer_Normalized_Nasal)))



#P_Nasal<-P_Nasal+geom_text(aes(x = 0.05,y = 0.35,label = paste("p-value = ",otu.adonis$aov.tab$`Pr(>F)`[1],sep = "")),size = 6,hjust = 0)




#Fecal

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Fecal, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Fecal, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Fecal, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-15.5
pc2<-8.7


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_Fecal<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,shape=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of Fecal microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


set.seed(2020)
otu.adonis<-adonis(dis_bray ~ group, data = data.frame(sample_data(LungCancer_Normalized_Fecal)))



#P_Fecal<-P_Fecal+geom_text(aes(x = 0.05,y = 0.35,label = paste("p-value = ",otu.adonis$aov.tab$`Pr(>F)`[1],sep = "")),size = 6,hjust = 0)
```



# divergence

```{r}
BAL_Interindividual <- as.data.frame(divergence(LungCancer_Normalized_BAL),coreset=apply(abundances(LungCancer_Normalized_BAL), 1, median))

BAL_Interindividual$ID<-row.names(BAL_Interindividual)

colnames(BAL_Interindividual)[1]="value"

BAL_Interindividual<-merge(BAL_Interindividual,MetaTable,by = "ID")

P_BAL<- ggplot(data=BAL_Interindividual,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Divergence Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)

```



# 制作中性模型
```{r}
spp <- t(otu_table(LungCancer_Normalized_BAL))
pool <-t(otu_table(LungCancer_Normalized_Oroph))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL)))
P_BO <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

#P_BO<-P_BO+theme(legend.position="none")

spp <- t(otu_table(LungCancer_Normalized_BAL))
pool <-t(otu_table(LungCancer_Normalized_Saliva))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL)))
P_BS <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
#P_BS<-P_BS+theme(legend.position="none")


spp <- t(otu_table(LungCancer_Normalized_BAL))
pool <-t(otu_table(LungCancer_Normalized_Nasal))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL)))
P_BN <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
#P_BN<-P_BN+theme(legend.position="none")




spp <- t(otu_table(LungCancer_Normalized_BAL))
pool <-t(otu_table(LungCancer_Normalized_Fecal))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL)))
P_BF <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())


P_BO/P_BS/P_BN
```


# 绘制总体的PCOA图
```{r}
set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized, dis_bray.pcoa, color="sample" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-23.8
pc2<-8.8


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample,shape=sample,fill=sample),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))
```



# 绘制BAL Oroph Saliva 三者的PCOA图
```{r}
set.seed(1024)
#BAL
LungCancer_Normalized_BOS<-subset_samples(LungCancer_Normalized,sample%in%c("BAL","Oroph","Saliva"))
## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BOS, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BOS, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BOS, dis_bray.pcoa, color="sample" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-13.2
pc2<-10.7


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample,shape=sample,fill=sample),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample),geom = "polygon",level = 0.95,alpha = 0.3)

```


# 比较  Oroph和Saliva到BAL  的bray_curties距离是否有差异

```{r}
set.seed(1024)
#BAL
LungCancer_Normalized_BOS<-subset_samples(LungCancer_Normalized,sample%in%c("BAL","Oroph","Saliva"))
## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BOS, "bray")

## 矩阵化
dis_bray<- as.matrix(dis_bray)

## 长列表化

dis_bray<- melt(dis_bray)

## 去除自身距离
dis_bray<- subset(dis_bray,!(Var1==Var2))

## 与元数据合并表格
### 挑选合适的元数据

Mt_data<-data.frame(sample_data(LungCancer_Normalized_BOS))

Mt_data<-Mt_data[,c("index","ID","group","sample","CT")]

merge_data<-merge(dis_bray,Mt_data,by.x="Var1",by.y="ID")
merge_data<-merge(merge_data,Mt_data,by.x="Var2",by.y="ID")


# 挑选Oroph和Saliva到BAL的距离
merge_data<-subset(merge_data,((sample.x=="Oroph")&(sample.y=="BAL"))|((sample.x=="Saliva")&(sample.y=="BAL")))

merge_data<-merge_data[!duplicated(merge_data[,3]),]

ggplot(merge_data,aes(x=sample.x,y=value))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ 
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Bray Curtis")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()

```



# 不同亚型的BAL-Oroph BAL-Saliva 共有OTU的差异
```{r}

#筛选BAL样本
BAL_type1 <- subset_samples(LungCancer_Normalized, CT=="type1"&sample=="BAL")
BAL_type2 <- subset_samples(LungCancer_Normalized, CT=="type2"&sample=="BAL")

BAL_type1_tax<-taxa_names(prune_taxa(taxa_sums(BAL_type1) > 0, BAL_type1))
BAL_type2_tax<-taxa_names(prune_taxa(taxa_sums(BAL_type2) > 0, BAL_type2))


Saliva_type1 <- subset_samples(LungCancer_Normalized,  CT=="type1"&sample=="Saliva")
Saliva_type2 <- subset_samples(LungCancer_Normalized,  CT=="type2"&sample=="Saliva")

Saliva_type1_tax<-taxa_names(prune_taxa(taxa_sums(Saliva_type1) > 0, Saliva_type1))
Saliva_type2_tax<-taxa_names(prune_taxa(taxa_sums(Saliva_type2) > 0, Saliva_type2))

uni_BAL<-setdiff(BAL_type1_tax,Saliva_type1_tax)
uni_BAL<-subset(TaxTable,row.names(TaxTable)%in%uni_BAL)

uni_Saliva<-setdiff(Saliva_type1_tax,BAL_type1_tax)
uni_Saliva<-subset(TaxTable,row.names(TaxTable)%in%uni_Saliva)





LungCancer_Normalized<- phyloseq(OTU, TAX, samples)
LungCancer_Normalized <- filter_taxa(LungCancer_Normalized,function(x) sum(x) > 1000 , TRUE )


T<-venn.diagram(list(A=BAL_type1_tax,B=Saliva_type1_tax),
filename=NULL,
lwd=1,#圈线粗度
lty=1, #圈线类型
col=c('#0099CC','#FF6666'), #圈线颜色
fill=c('#0099CC','#FF6666'), #填充颜色
cat.col=c('#0099CC','#FF6666'),#A和B的颜色
cat.cex = 2.5,# A和B的大小
rotation.degree = 0,#旋转角度
main = "A&B",#主标题内容
main.cex = 2,#主标题大小
sub = "plot : example",#亚标题内容
sub.cex = 1,#亚标题字大小
cex=1.5,#里面交集字的大小
alpha = 0.5,#透明度
reverse=TRUE)
grid.draw(T)



T<-venn.diagram(list(A=BAL_type2_tax,B=Saliva_type2_tax),
filename=NULL,
lwd=1,#圈线粗度
lty=1, #圈线类型
col=c('#0099CC','#FF6666'), #圈线颜色
fill=c('#0099CC','#FF6666'), #填充颜色
cat.col=c('#0099CC','#FF6666'),#A和B的颜色
cat.cex = 2.5,# A和B的大小
rotation.degree = 0,#旋转角度
main = "A&B",#主标题内容
main.cex = 2,#主标题大小
sub = "plot : example",#亚标题内容
sub.cex = 1,#亚标题字大小
cex=1.5,#里面交集字的大小
alpha = 0.5,#透明度
reverse=TRUE)
grid.draw(T)
```



# 链球菌属内


```{r}


LungCancer_Normalized_Stre<-subset_taxa(LungCancer_Normalized,Genus=="g_Streptococcus")

LungCancer_Normalized_Stre_BAL<-subset_samples(LungCancer_Normalized_Stre,sample=="BAL")


alpha_meas = c("Shannon")


p<-plot_richness(LungCancer_Normalized_Stre_BAL, "CT", "CT",nrow = 1,measures=alpha_meas)


ggplot(data=p$data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()


set.seed(1024)

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Stre_BAL, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Stre_BAL, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Stre_BAL, dis_bray.pcoa, color="CT" ,shape = "CT") + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-45.5
pc2<-17.9


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = CT),geom = "polygon",level = 0.95,alpha = 0.3)













LungCancer_Normalized_Stre_BAL_Saliva<-subset_samples(LungCancer_Normalized_Stre,sample%in%c("BAL","Saliva"))


set.seed(1024)

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Stre_BAL_Saliva, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Stre_BAL_Saliva, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Stre_BAL_Saliva, dis_bray.pcoa, color="sample" ,shape = "CT") + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-13.2
pc2<-10.7


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample,shape=sample,fill=sample),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample),geom = "polygon",level = 0.95,alpha = 0.3)
```






```{r}
set.seed(1024)
#BAL
LungCancer_Normalized_BS<-subset_samples(LungCancer_Normalized,sample%in%c("BAL","Saliva"))
## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BS, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BS, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BS, dis_bray.pcoa, color="sample" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-13.1
pc2<-8.8


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample,shape=sample,fill=sample),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample),geom = "polygon",level = 0.95,alpha = 0.3)

```














# 配对样本的bray_Curtis距离
```{r}
distance <- "bray"
# Distance matrix for samples


# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(LungCancer_Normalized, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")


picture<-list()

A<-"BAL"
B<-"Oroph"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)
Cancer.ps[1,7]<-"Oroph"

ps_BO<-rbind(Health.ps,Cancer.ps)
ps_BO<-ps_BO[,3:4]
ps_BO$sample<-"Oroph"



A<-"BAL"
B<-"Saliva"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_BS<-rbind(Health.ps,Cancer.ps)
ps_BS<-ps_BS[,3:4]
ps_BS$sample<-"Saliva"


A<-"BAL"
B<-"Nasal"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_BN<-rbind(Health.ps,Cancer.ps)
ps_BN<-ps_BN[,3:4]
ps_BN$sample<-"Nasal"


A<-"BAL"
B<-"Fecal"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_BF<-rbind(Health.ps,Cancer.ps)
ps_BF<-ps_BF[,3:4]
ps_BF$sample<-"Fecal"


ps<-rbind(ps_BO,ps_BS,ps_BN,ps_BF)











P_BAL <- ggplot(data=ps,aes(x=sample,y=value,fill=sample))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=sample),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Distance to BAL")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)

```

# 配对样本的PCOA

```{r}
LungCancer_Normalized_BS<-subset_samples(LungCancer_Normalized,sample%in%c("BAL","Saliva"))

dis_bray<- phyloseq::distance(LungCancer_Normalized_BS, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BS, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BS, dis_bray.pcoa, color="CT" ,shape = "sample") + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 过滤掉没有分型的数据
data<-subset(data,!(CT=="NA"))

## 用样本类型和分型结果重命名

data<-tidyr::unite(data, "sample_CT", sample, CT,remove=FALSE)

data$sample_CT<-as.factor(data$sample_CT)
## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-13.1
pc2<-8.8


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample_CT,shape=sample_CT,fill=sample_CT),size=4)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample_CT),geom = "polygon",level = 0.95,alpha = 0.3)
```


# 配对样本BAL到唾液的距离

```{r}
distance <- "bray"
# Distance matrix for samples

LungCancer_Normalized_test<-filter_taxa(LungCancer_Normalized_BS,function(x) sum(x) > 100, TRUE )
# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(LungCancer_Normalized_test, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,new_MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,new_MetaTable,by.x = "Var2",by.y = "ID")

A<-"BAL"
B<-"Saliva"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_BS<-rbind(Health.ps,Cancer.ps)


ggplot(data=ps_BS,aes(x=CT.x,y=value,fill=CT.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+
  geom_jitter(aes(fill=CT.x),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Paired Bray Curtis Distance")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()


ggplot(ps_BS, aes(x = value))+geom_density(fill = "#868686FF", alpha = 0.4)+geom_vline(aes(xintercept = 0.78), linetype = "dashed")+theme_classic()



ps_BS$value<-car::recode(ps_BS$value,"0:0.78='type1_p';0.78:1='type2_p'")



```



# 绘制物种组成柱状图
```{r}
# 筛选样本

idx<-sample_names(LungCancer_Normalized_BS)

OTU_TAX_Genus_BAL<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%idx]
#计算相对丰度并按照相对丰度和进行排序
OTU_TAX_Genus_BAL<-t(OTU_TAX_Genus_BAL)
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL/apply(OTU_TAX_Genus_BAL,1,sum)*100
OTU_TAX_Genus_BAL<-as.data.frame(t(OTU_TAX_Genus_BAL))
OTU_TAX_Genus_BAL$sum<-rowSums(OTU_TAX_Genus_BAL) 
OTU_TAX_Genus_BAL <- OTU_TAX_Genus_BAL[order(OTU_TAX_Genus_BAL$sum, decreasing = TRUE), ]
#筛选top20 genus 其余合并为Others
OTU_TAX_Genus_BAL_top20<-OTU_TAX_Genus_BAL[1:15,-ncol(OTU_TAX_Genus_BAL)]
OTU_TAX_Genus_BAL_top20['Others', ] <- 100 - colSums(OTU_TAX_Genus_BAL_top20)
name<-row.names(OTU_TAX_Genus_BAL_top20)
name[17]<-"Others"
name<-name[-16]
#加上一列Taxonomy的信息，将宽数据转化成长数据
OTU_TAX_Genus_BAL_top20<-as.data.frame(t(OTU_TAX_Genus_BAL_top20))

OTU_TAX_Genus_BAL_top20$ID<-row.names(OTU_TAX_Genus_BAL_top20)
OTU_TAX_Genus_BAL_top20<-OTU_TAX_Genus_BAL_top20[order(-OTU_TAX_Genus_BAL_top20$`Others`),]
a<-as.character(OTU_TAX_Genus_BAL_top20$ID)
OTU_TAX_Genus_BAL_top20$ID<-factor(OTU_TAX_Genus_BAL_top20$ID,levels=a)


OTU_TAX_Genus_BAL_top20<-melt(OTU_TAX_Genus_BAL_top20, id = 'ID')
#绘制条形图
OTU_TAX_Genus_BAL_top20<-merge(OTU_TAX_Genus_BAL_top20,MetaTable,by.x="ID",by.y="ID")
OTU_TAX_Genus_BAL_top20$variable<-factor(OTU_TAX_Genus_BAL_top20$variable,levels=name)

#合并样本类型和CT信息

OTU_TAX_Genus_BAL_top20<-tidyr::unite(OTU_TAX_Genus_BAL_top20, "sample_CT", sample, CT,remove=FALSE)

OTU_TAX_Genus_BAL_top20$sample_CT<-as.factor(OTU_TAX_Genus_BAL_top20$sample_CT)

OTU_TAX_Genus_BAL_top20$sample_CT<-factor(OTU_TAX_Genus_BAL_top20$sample_CT,levels=c("BAL_type1","BAL_type2","Saliva_type1","Saliva_type2"))
OTU_TAX_Genus_BAL_top20<-subset(OTU_TAX_Genus_BAL_top20,!(sample_CT=="NA"))

p<-ggplot(OTU_TAX_Genus_BAL_top20, aes(ID, value, fill = variable))+
geom_col(position = 'stack', width = 1) +labs(x = 'Sample ID', y = 'Relative Abundance(%)') +
theme(axis.text = element_text(size = 12,angle = 40, hjust = 1, vjust = 0.5), axis.title = element_text(size = 13)) +
theme(legend.text = element_text(size = 11))+scale_fill_manual(values =  rev(c("#8FBC94","#548687","#4FB0C6","#4F86C6","#C65146","#EC6A5C","#6E7783","#77AAAD","#e97f02","#f8ca00","#3a5134","#4f953b","#99CCCC","#FFCC99","#CC9999","#CCCC99","#0099CC","#FF6666","#339966","#996699","#CC9966","#666666","#336666","#CCCC33","#FDD692","#EC7357","#754F44")))+theme(panel.grid = element_blank(), panel.background = element_rect(color = 'black', fill = 'transparent')) + theme(legend.title = element_blank())+ scale_y_continuous(expand = c(0, 0))+facet_grid(~sample_CT, scales = 'free_x', space = "free")

```








```{r}
#首先提取OTU表和TAX表
OTU<-as.data.frame(otu_table(LungCancer_Normalized))
TAX<-as.data.frame(tax_table(LungCancer_Normalized))




OTU_TAX<-cbind.data.frame(TAX,OTU)
OTU_TAX_Genus<-aggregate(OTU_TAX[,7:448],list(OTU_TAX$Genus),sum)
#去除未注释到属水平的单元
OTU_TAX_Genus<-OTU_TAX_Genus[-607:-633,]
OTU_TAX_Genus<-OTU_TAX_Genus[-1,]
rownames(OTU_TAX_Genus)<-  OTU_TAX_Genus$Group.1
OTU_TAX_Genus<- OTU_TAX_Genus
OTU_TAX_Genus<- OTU_TAX_Genus[,-1]
write.table(OTU_TAX_Genus,"OTU_TAX_Genus.txt",sep = "\t")




A<-OTU_TAX[,1:6]
B<-data.frame(rownames(OTU_TAX_Genus))

c<-merge(B,A,by.x="rownames.OTU_TAX_Genus.",by.y="Genus")

TAX<-c[!duplicated(c$rownames.OTU_TAX_Genus.), ]



OTU_TAX_Genus_RA<-t(OTU_TAX_Genus)
OTU_TAX_Genus_RA<-OTU_TAX_Genus_RA/apply(OTU_TAX_Genus_RA,1,sum)*100
OTU_TAX_Genus_RA<-as.data.frame(t(OTU_TAX_Genus_RA))



```

# 构建属的Phyloseq对象
```{r}
OTU_TAX_Genus<-read.table("OTU_TAX_Genus.txt",sep = "\t",header = TRUE,row.names = 1)
rownames(TAX)<-rownames(OTU_TAX_Genus)
TAX<-as.matrix(TAX)
OTU = otu_table(OTU_TAX_Genus, taxa_are_rows = TRUE)
TAX = tax_table(TAX)
samples = sample_data(MetaTable)
LungCancer <- phyloseq(OTU, TAX, samples)

LungCancer_Normalized_BAL <- subset_samples(LungCancer, sample =="BAL")

```




# 计算Coummunity type
```{r}
pseq.comp<-microbiome::transform(LungCancer_Normalized_BAL, "compositional")
taxa <- core_members(pseq.comp, detection = 0.1/100, prevalence = 1/100)
pseq <- prune_taxa(taxa, LungCancer_Normalized_BAL)
dat  <- as.data.frame(abundances(pseq))
dat$ID<-row.names(dat)
name<-as.data.frame(row.names(OTU_TAX_Genus),row.names(OTU_TAX_Genus_RA))
name$Genus<-rownames(name)
colnames(name)[1]<-"ID"
dat<-merge(dat,name,by="ID")
row.names(dat)<-dat$Genus
dat<-dat[,c(-1,-101)]

count <- as.matrix(t(dat))
set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)



  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d1 <-d1[335:359,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d2 <-d2[335:359,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))


```

# 将分型的结果加入metadata
```{r}
data<-data.frame(sample_data(LungCancer_Normalized_BAL))
data<-data[,1:2]
type<-data.frame(ass)
type$ID<-row.names(type)
type$ass<-car::recode(type$ass,"1='type1';2='type2'")
data<-merge(data,type,by="ID")
data<-data[,-1]
MetaTable<-merge(MetaTable,data,by="index",all.x=TRUE)
rownames(MetaTable)<-MetaTable$ID
colnames(MetaTable)[44]<-"CT"
OTU = otu_table(OTU_TAX_Genus, taxa_are_rows = TRUE)
TAX = tax_table(TAX)
samples = sample_data(MetaTable)
LungCancer <- phyloseq(OTU, TAX, samples)




LungCancer_Normalized

OTU = otu_table(LungCancer_Normalized, taxa_are_rows = TRUE)
TAX = tax_table(LungCancer_Normalized)
samples = sample_data(MetaTable)
LungCancer_Normalized<- phyloseq(OTU, TAX, samples)
```




# 绘制BAL的两种亚型PCOA图
```{r}

LungCancer_Normalized_BAL <- subset_samples(LungCancer_Normalized, sample =="BAL")

set.seed(2048)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BAL, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BAL, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BAL, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-17
pc2<-7


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,22,21,21,21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=cbbPalette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_text(colour = "black",size = 14,face = "bold"),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = CT),geom = "polygon",level = 0.95,alpha = 0.3)



```



# Source Tracker
```{r}
# 绘制community type 的组成
data<-data.frame(sample_data(LungCancer_Normalized_BAL))


SourceTracke<-read.table("Saliva_BAL.txt",header = TRUE,row.names = 1,sep = "\t")


data<-merge(data,SourceTracke,by="ID")



ggplot(data,aes(x=CT,y=Saliva,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=3)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Source Tracker")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Saliva")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()



ggplot(data,aes(x=CT,y=Oroph,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=3)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Source Tracker")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Oroph")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()


ggplot(data, aes(x = Saliva))+geom_density(fill = "#868686FF", alpha = 0.4)+geom_vline(aes(xintercept = 0.78), linetype = "dashed")+theme_classic()

```



# 对两种亚型进行中性模型
```{r}

```

# 计算两种亚型的α多样性
```{r}
p<-plot_richness(LungCancer_Normalized_BAL, "group", "group",nrow = 1,measures=alpha_meas)


P_BAL <- ggplot(data=p$data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()

```




# 计算两种亚型的divergence
```{r}
BAL_Interindividual <- as.data.frame(divergence(LungCancer_Normalized_BAL),coreset=apply(abundances(BAL_Interindividual), 1, median))
BAL_Interindividual$ID<-row.names(BAL_Interindividual)
colnames(BAL_Interindividual)[1]="value"
BAL_Interindividual<-merge(BAL_Interindividual,MetaTable,by = "ID")
ggplot(BAL_Interindividual,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Divergence")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()
```


```{r}

LungCancer_Normalized_BAL_Health<-subset_samples(LungCancer_Normalized_BAL,group=="Health")
data<-data.frame(sample_data(LungCancer_Normalized_BAL_Health))
table(data$CT,data$肺部炎症.1)


LungCancer_Normalized_BAL_Cancer<-subset_samples(LungCancer_Normalized_BAL,group=="Cancer")
data_Cancer<-data.frame(sample_data(LungCancer_Normalized_BAL_Cancer))
table(data_Cancer$CT,data_Cancer$肺部炎症.1)
chisq.test(table(data_Cancer$CT,data_Cancer$gender))
```
    
```{r}
plot_data$g_Bifidobacterium
data<-as.data.frame(t(OTU_TAX_Genus_RA))
data$ID<-row.names(data)
data<-merge(data,MetaTable,by="ID")
plot_data<-subset(data,sample=="BAL")
plot_data$g_EN
 ggplot(plot_data,aes(x=CT,y=g_Lactobacillus,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance (%)")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_grid(.~group,scales = "free_y" )
 
 
 
 
 
 
 
 
 plot_data$g_Actinobacillus
 ggplot(plot_data,aes(x=group,y=g_Selenomonas,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance (%)")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_grid(.~CT,scales = "free_y" )
```


```{r}
#alpha_meas = c("InvSimpson", "Observed","Shannon")
LungCancer_Normalized_BAL <- subset_samples(LungCancer, sample =="BAL")
LungCancer_Normalized_BAL <- subset_samples(LungCancer_Normalized, sample =="BAL")
p<-plot_richness(LungCancer_Normalized_BAL, "group", "group",nrow = 1,measures="Observed")


P_BAL <- ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_grid(.~CT,scales = "free_y" )

```



# Source Tracker
```{r}
# 绘制community type 的组成
data<-data.frame(sample_data(LungCancer_Normalized_BAL))


SourceTracke<-read.table("Saliva_BAL.txt",header = TRUE,row.names = 1,sep = "\t")


data<-merge(data,SourceTracke,by="ID")



ggplot(data,aes(x=group.x,y=Oroph,fill=group.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group.x),width =0.2,shape = 21,size=3)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Source Tracker")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Saliva")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_grid(.~CT,scales = "free_y" )



ggplot(data,aes(x=CT,y=Oroph,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=3)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("Source Tracker")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Oroph")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()

data$Saliva<-car::recode(data$Saliva,"0:0.4='type2_p';0.4:1='type1_p'")
```



# 绘制BAL的两种亚型PCOA图
```{r}

LungCancer_Normalized_BAL_type1 <- subset_samples(LungCancer_Normalized, (sample =="BAL")&(CT=="type1"))

set.seed(2048)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BAL_type1, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BAL_type1, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BAL_type1, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-22.8
pc2<-15.4


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,22,21,21,21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=cbbPalette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_text(colour = "black",size = 14,face = "bold"),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = group),geom = "polygon",level = 0.95,alpha = 0.3)


LungCancer_Normalized_BAL_type2 <- subset_samples(LungCancer_Normalized, (sample =="BAL")&(CT=="type2"))

set.seed(2048)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_BAL_type2, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_BAL_type2, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_BAL_type2, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-22.8
pc2<-15.4


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,22,21,21,21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=group,fill=group),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=cbbPalette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_text(colour = "black",size = 14,face = "bold"),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = group),geom = "polygon",level = 0.95,alpha = 0.3)
```


# 肺癌病人type1 和 对照type1 的中性模型对比 OTU水平
```{r}
LungCancer_Normalized_BAL_type1_Cancer <- subset_samples(LungCancer_Normalized, (sample =="BAL")&(CT=="type1")&(group=="Cancer"))
LungCancer_Normalized_Saliva_type1_Cancer<- subset_samples(LungCancer_Normalized, (sample =="Saliva")&(CT=="type1")&(group=="Cancer"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Cancer))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type1_Cancer))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Cancer)))
P_BO_type1_Cancer <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

P_BO_type1_Cancer_data<-P_BO_type1_Cancer$data



LungCancer_Normalized_BAL_type1_Health <- subset_samples(LungCancer_Normalized, (sample =="BAL")&(CT=="type1")&(group=="Health"))
LungCancer_Normalized_Saliva_type1_Health<- subset_samples(LungCancer_Normalized, (sample =="Saliva")&(CT=="type1")&(group=="Health"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Health))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type1_Health))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Health)))
P_BO_type1_Health<- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

```



# 肺癌病人type1 和 对照type1 的中性模型对比 属水平
```{r}
LungCancer_Normalized_BAL_type1_Cancer <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type1")&(group=="Cancer"))
LungCancer_Normalized_Saliva_type1_Cancer<- subset_samples(LungCancer ,(sample =="Saliva")&(CT=="type1")&(group=="Cancer"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Cancer))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type1_Cancer))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Cancer)))
P_BO_type1_Cancer <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

P_BO_type1_Cancer_data<-P_BO_type1_Cancer$data



LungCancer_Normalized_BAL_type1_Health <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type1")&(group=="Health"))
LungCancer_Normalized_Saliva_type1_Health<- subset_samples(LungCancer, (sample =="Saliva")&(CT=="type1")&(group=="Health"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Health))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type1_Health))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Health)))
P_BO_type1_Health<- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
P_BO_type1_Health_data<-P_BO_type1_Health$data


type1_data<-merge(P_BO_type1_Cancer_data[,11:12],P_BO_type1_Health_data[,11:12],by="Genus")


```






```{r}
LungCancer_Normalized_BAL_type1_Cancer <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type1")&(group=="Cancer"))
LungCancer_Normalized_Fecal_type1_Cancer<- subset_samples(LungCancer ,(sample =="Fecal")&(CT=="type1")&(group=="Cancer"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Cancer))
pool <-t(otu_table(LungCancer_Normalized_Fecal_type1_Cancer))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Cancer)))
P_BO_type1_Cancer <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

P_BO_type1_Cancer_data<-P_BO_type1_Cancer$data



LungCancer_Normalized_BAL_type1_Health <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type1")&(group=="Health"))
LungCancer_Normalized_Fecal_type1_Health<- subset_samples(LungCancer, (sample =="Fecal")&(CT=="type1")&(group=="Health"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type1_Health))
pool <-t(otu_table(LungCancer_Normalized_Fecal_type1_Health))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1_Health)))
P_BO_type1_Health<- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
P_BO_type1_Health_data<-P_BO_type1_Health$data


type1_data<-merge(P_BO_type1_Cancer_data[,11:12],P_BO_type1_Health_data[,11:12],by="Genus")


```








```{r}

LungCancer_Normalized_BAL_type2_Cancer <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type2")&(group=="Cancer"))
LungCancer_Normalized_Saliva_type2_Cancer<- subset_samples(LungCancer ,(sample =="Saliva")&(CT=="type2")&(group=="Cancer"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type2_Cancer))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type2_Cancer))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type2_Cancer)))
P_BO_type2_Cancer <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

P_BO_type2_Cancer_data<-P_BO_type2_Cancer$data



LungCancer_Normalized_BAL_type2_Health <- subset_samples(LungCancer, (sample =="BAL")&(CT=="type2")&(group=="Health"))
LungCancer_Normalized_Saliva_type2_Health<- subset_samples(LungCancer, (sample =="Saliva")&(CT=="type2")&(group=="Health"))

spp <- t(otu_table(LungCancer_Normalized_BAL_type2_Health))
pool <-t(otu_table(LungCancer_Normalized_Saliva_type2_Health))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type2_Health)))
P_BO_type2_Health<- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
P_BO_type2_Health_data<-P_BO_type2_Health$data


type2_data<-merge(P_BO_type2_Cancer_data[,11:12],P_BO_type2_Health_data[,11:12],by="Genus")

type2_data<-subset(type2_data,!(fit_class.x==fit_class.y))
```

```{r}
plot_data$g_Bifidobacterium
data<-as.data.frame(t(OTU_TAX_Genus_RA))
data$ID<-row.names(data)
data<-merge(data,new_MetaTable,by="ID")
plot_data<-subset(data,(sample=="BAL")&!(CT=="NA"))
plot_data$new_group<-paste((plot_data$group),(plot_data$CT))


plot_data$g_EN
 ggplot(plot_data,aes(x=CT,y=g_Lactobacillus,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance (%)")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_grid(.~group,scales = "free_y" )
 
 
 
 plot_data<-subset(data,(sample=="Fecal")&!(CT=="NA"))
 plot_data<-plot_data[-69,]
 plot_data$new_group<-paste((plot_data$group),(plot_data$CT))
 ggplot(plot_data,aes(x=new_group,y=plot_data$g_Parabacteroides,fill=new_group))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=new_group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
    scale_fill_manual(values = c("#E69F00", "#56B4E9","#009E73", "#F0E442"))+  #设置填充的颜色
    scale_color_manual(values=c("black","black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("Relative Abundance (%)")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means(comparisons=my_comparisons)
 
 
 
 
 
 

 
 
 
 
 
 
 
 
 
 
 
 
 
 
 plot_data$g_Actinobacillus
 ggplot(plot_data,aes(x=group,y=plot_data$g_Veillonella,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance (%)")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+
   facet_wrap(.~CT,scales = "free_y" )
 
 
 my_comparisons <- list(c("Cancer type1", "Health type2"),c("Cancer type2","Health type2"),c("Health type1","Health type2"),c("Cancer type1", "Cancer type2"),c("Cancer type2", "Health type1"),c("Cancer type1", "Health type1"))
 
 
 
 
 
 
 
 
 
 
# Type1 的 Fecal 的差异
 
LungCancer_Normalized_Fecal_type1 <- subset_samples(LungCancer_Normalized, (sample =="BAL")&(CT=="type1"))


p<-plot_richness(LungCancer_Normalized_Fecal_type1, "group", "group",nrow = 1,measures="Shannon")


ggplot(data=p$data,aes(x=group,y=value,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()








dis_bray<- phyloseq::distance(LungCancer_Normalized_Fecal_type1, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Fecal_type1, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Fecal_type1, dis_bray.pcoa, color="group" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-23.8
pc2<-8.8


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample,shape=sample,fill=sample),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


# Parvimonas 绘制Parvimonas在lung 和 肠道
plot_data<-subset(data,(sample=="BAL")&!(CT=="NA"))
plot_data<-subset(plot_data,CT=="type1")
Parvimonas_BAL<-plot_data[,c("ID","g_Parvimonas","index","group")] 
plot_data<-subset(data,(sample=="Fecal")&!(CT=="NA"))
plot_data<-subset(plot_data,CT=="type1")
Parvimonas_Fecal<-plot_data[,c("ID","g_Parvimonas","index","group")]

Parvimonas<-merge(Parvimonas_BAL,Parvimonas_Fecal,by="index")



ggplot(Parvimonas, aes(x=g_Parvimonas.x, y= g_Parvimonas.y,color=group.x,group=group.x)) +
    geom_point() +
    geom_smooth(method="lm") +
    theme(panel.grid=element_blank(), panel.background=element_rect(fill='transparent', color='black')) +stat_cor(method = "spearman")

```




```{r}
plot_data<-subset(plot_data,CT=="type2")
plot_data<-t(plot_data)
write.table(plot_data,"Fecal_table.txt",sep = "\t")

```



# 读取肺癌病人和健康人共有的连续检验变量
```{r}
Janyan_data<-read.table("肺癌健康共有连续变量.txt",header = T,sep = "\t",row.names = 1)

```

# 制作随机森林模型
```{r}
LungCancer_Normalized_BAL <- subset_samples(LungCancer, sample =="BAL")
#获得物种组成表


#获取metadata
sample<-data.frame(sample_data(LungCancer_Normalized_BAL))
#获取物种注释信息
tax<-data.frame(tax_table(LungCancer_Normalized_BAL))



count<-data.frame(t(otu_table(LungCancer_Normalized_BAL)))
colnames(count)<-tax$Genus
count<-data.frame(count/apply(count,1,sum)*100)
a<-count
count<-count[,colSums(count)>100]


#count<-count[,colSums(count)!=0]
count$CT<-sample$CT

#把g_Prevotella+g_Prevotella.7加在一起
Prevotella<-count[,"g_Prevotella"]+count[,"g_Prevotella.7"]
count<-count[,-14:-15]
count$g_Prevotella<-Prevotella
count$CT<-ifelse(count$CT == "type1", 1, 0)
count$CT<-as.factor(count$CT)
colnames( count)<-gsub(" ", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("-", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("(", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub(")", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("[", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("]", "", colnames( count), fixed = TRUE)

count<-Oribacterium_S[,c("Oribacterium","CT")]
set.seed(2020)
count_rf <- AUCRF(CT~., data=count, ranking='MDA', ntree=500, pdel=0.2)

var<-count_rf$Xopt
#预测值
count_rf_probs <- predict(count_rf$RFopt, type='prob')[,2]
#
count_rf_roc <- roc(count$CT~count_rf_probs,smooth=FALSE )
```

# 构建segal数据的Phyloseq对象
```{r}
Otu<-read.table("segal_OTU.txt",sep = "\t",header = TRUE,row.names = 1)
Tax<-read.table("segal_tax.txt",sep = "\t",header = TRUE,row.names = 1)
Sample<-read.table("segal_sample.txt",sep = "\t",header = TRUE,row.names = 1)
Otu<-Otu[,colnames(Otu)%in%Sample$ID]
Tax<-as.matrix(Tax)
Otu = otu_table(Otu, taxa_are_rows = TRUE)
Tax = tax_table(Tax)
samples = sample_data(Sample)
Segal <- phyloseq(Otu, Tax, samples)
Segal <- rarefy_even_depth(Segal)
Segal <- filter_taxa(Segal,function(x) sum(x) > 10 , TRUE )



OTU<-as.data.frame(otu_table(Segal))
TAX<-as.data.frame(tax_table(Segal))




OTU_TAX<-cbind.data.frame(TAX,OTU)
Segal_Genus<-aggregate(OTU_TAX[,7:81],list(OTU_TAX$Genus),sum)
#去除未注释到属水平的单元
Segal_Genus<-Segal_Genus[-1,]
rownames(Segal_Genus)<-  Segal_Genus$Group.1
Segal_Genus<- Segal_Genus[,-1]

Segal_Genus<-t(Segal_Genus)
Segal_Genus_RA<-data.frame(Segal_Genus/apply(Segal_Genus,1,sum)*100)
colnames(Segal_Genus_RA)<-colnames(Segal_Genus)
colnames(Segal_Genus_RA)<-gsub(" ", "", colnames(Segal_Genus_RA), fixed = TRUE)
Segal_Genus_RA<-Segal_Genus_RA[,colnames(Segal_Genus_RA)%in%var]


#挑选BAL样本
Segal_BAL<-subset_samples(Segal,sample=="BAL")

Segal_Genus_RA <- Segal_Genus_RA[row.names(Segal_Genus_RA)%in%sample.names(Segal_BAL),]

Segal_rf_probs <- predict(count_rf$RFopt,newdata=Segal_Genus_RA, type='prob')[,2]





# 构建Run1 BAL 的phyloseq对象
Segal_type<-data.frame(Segal_rf_probs)
Segal_type$ID<-rownames(Segal_type)
Segal_type$Segal_rf_probs<-car::recode(Segal_type$Segal_rf_probs,"0:0.5='type2';0.5:1='type1'")
# 将分型信息加入metadata
# 提取metadata
test<-data.frame(sample_data(Segal_BAL))

test$CT<-Segal_type$Segal_rf_probs

Otu = otu_table(Segal_BAL)
Tax = tax_table(Segal_BAL)
samples = sample_data(test)
Segal_BAL <- phyloseq(Otu, Tax, samples)

alpha_meas = c("Observed")

"Observed","Shannon"
my_comparisons <- list(c("type1", "type2"))

p<-plot_richness(Segal_BAL,"CT", "CT",nrow = 1,measures=alpha_meas)


P_BAL <- ggplot(data=p$data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Observed Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)





set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(Segal_BAL, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(Segal_BAL, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(Segal_BAL, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-18.3
pc2<-6.5


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
P_BAL<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


```





```{r}
 ggplot(lung,aes(x=group,y=lung$VC.MAX,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()



```


# 肺肠轴交流证据搜寻
```{r}
table(test$gastroenteropathy,test$CT,test$group)
A<-table(test$gastroenteropathy,test$CT,test$group)
test.no<-subset(test,smoking=="No")
test.yes<-subset(test,smoking=="Yes")


chisq.test(table(test.yes$gastroenteropathy,test.yes$CT))

test.male<-subset(test,gender=="male")

fisher.test(A[1:2,1:2,1])
test.female<-subset(test,gender=="female")









## 证明性别与肠胃病没有差异
test<-table(test$gender,test$gastroenteropathy,test$group)
test.cancer<-data.frame(test[1:2,1:2,1])

colnames(test.cancer)<-c("Gender","Gastroenteropathy","Freq")
ggplot(data = test.cancer, mapping = aes(x = Gender, y = Freq, fill = Gastroenteropathy)) + geom_bar(stat = 'identity', position = 'fill') + geom_text(mapping = aes(label = Freq), size = 5, colour = 'black', vjust = 3.5, hjust = .5, position = position_fill())+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Freq")+xlab("Gender")+scale_y_continuous(expand=c(0,0))


## 证明性别与分型也无关
test<-data.frame(sample_data(LungCancer_Normalized_BAL))
test<-table(test$gender,test$CT,test$group)
test.cancer<-data.frame(test[1:2,1:2,1])

colnames(test.cancer)<-c("Gender","CT","Freq")
ggplot(data = test.cancer, mapping = aes(x = Gender, y = Freq, fill = CT)) + geom_bar(stat = 'identity', position = 'fill') + geom_text(mapping = aes(label = Freq), size = 5, colour = 'black', vjust = 3.5, hjust = .5, position = position_fill())+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Freq")+xlab("Gender")+scale_y_continuous(expand=c(0,0))

test<-data.frame(sample_data(LungCancer_Normalized_BAL))
test<-subset(test,gender=="male")
test<-table(test$CT,test$gastroenteropathy,test$group)

CMHtest(test,overall=TRUE)
BreslowDayTest(test, OR = NA, correct = TRUE)


# 不同分型对于肠胃病的影响
test<-data.frame(sample_data(LungCancer_Normalized_BAL))
test<-table(test$CT,test$gastroenteropathy)
chisq.test(test,correct = FALSE)
#结果显示无差异

# 按照group分层分析
test<-data.frame(sample_data(LungCancer_Normalized_BAL))
test<-table(test$CT,test$gastroenteropathy,test$group)
#检验第一层
chisq.test(test[1:2,1:2,1],correct = FALSE)
fisher.test(test[1:2,1:2,1])
#检验第二层
chisq.test(test[1:2,1:2,2])
fisher.test(test[1:2,1:2,2])
# 对比两组OR值是否有差异
BreslowDayTest(test, OR = NA)
# CMHtest检验
CMHtest(test,overall=TRUE)



# 按照性别分层分析
test<-data.frame(sample_data(LungCancer_Normalized_BAL))
# 筛选肺癌患者
test<-subset(test,group=="Cancer")
test<-table(test$gastroenteropathy,test$CT,test$gender)
#检验第一层
chisq.test(test[1:2,1:2,1])
fisher.test(test[1:2,1:2,1])
#检验第二层
chisq.test(test[1:2,1:2,2])
fisher.test(test[1:2,1:2,2])
# 对比两组OR值是否有差异
BreslowDayTest(test, OR = NA)
# CMHtest检验
CMHtest(test,overall=TRUE)




# 按照性别分层分析
test<-data.frame(sample_data(LungCancer_Normalized_BAL))
# 筛选肺癌患者
test<-subset(test,group=="Cancer")
test<-table(test$gastroenteropathy,test$CT,test$gender)
#检验第一层
chisq.test(test[1:2,1:2,1])
fisher.test(test[1:2,1:2,1])
#检验第二层
chisq.test(test[1:2,1:2,2])
fisher.test(test[1:2,1:2,2])
# 对比两组OR值是否有差异
BreslowDayTest(test, OR = NA)
# CMHtest检验
CMHtest(test,overall=TRUE)
```


```{r}

# 吸烟与肠胃病的发生是相关的
test<-data.frame(sample_data(LungCancer_Normalized_BAL))

test<-subset(test,group=="Cancer")

test.1<-data.frame(table(test$gastroenteropathy,test$smoking))


colnames(test.1)<-c("gastroenteropathy","smoking","Freq")

ggplot(data = test.1, mapping = aes(x = smoking, y = Freq, fill = gastroenteropathy)) + geom_bar(stat = 'identity', position = 'fill') + geom_text(mapping = aes(label = Freq), size = 5, colour = 'black', vjust = 3.5, hjust = .5, position = position_fill())+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Freq")+xlab("smoking")+scale_y_continuous(expand=c(0,0))


# 肺型与肠胃病的发生是相关的


test<-data.frame(sample_data(LungCancer_Normalized_BAL))

test<-subset(test,group=="Cancer")

#去除肺癌晚期的病人
test<-subset(test,!(stage%in%c("IVA","IV","IIIB","IIIA")))
test.1<-data.frame(table(test$gastroenteropathy,test$CT,test$gender))

test.1<-data.frame(table(test$gastroenteropathy,test$CT))
colnames(test.1)<-c("gastroenteropathy","CT","gender","Freq")
colnames(test.1)<-c("gastroenteropathy","CT","Freq")
ggplot(data = test.1, mapping = aes(x = CT, y = Freq, fill = gastroenteropathy)) + geom_bar(stat = 'identity', position = 'fill') + geom_text(mapping = aes(label = Freq), size = 5, colour = 'black', vjust = 3.5, hjust = .5, position = position_fill())+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Freq")+xlab("CT")+scale_y_continuous(expand=c(0,0))




# 吸烟人群与不吸烟人群进行分层

test<-data.frame(sample_data(LungCancer_Normalized_BAL))

test<-subset(test,group=="Cancer")

test.1<-data.frame(table(test$gastroenteropathy,test$CT,test$smoking))

colnames(test.1)<-c("gastroenteropathy","CT","smoking","Freq")
ggplot(data = test.1, mapping = aes(x = CT, y = Freq, fill = gastroenteropathy)) + geom_bar(stat = 'identity', position = 'fill') + geom_text(mapping = aes(label = Freq), size = 5, colour = 'black', vjust = 3.5, hjust = .5, position = position_fill())+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Freq")+xlab("CT")+scale_y_continuous(expand=c(0,0))+facet_grid(.~smoking,scales = "free_y" )

```


# 肺肠微生物的相似性

```{r}
# 筛选配对的BAL和肛拭子样本，计算配对Bray_curtis距离
LungCancer_Normalized

set.seed(1024)
method <- "tsne"
trans <- "hellinger"
#distance <- "canberra"，"clark"有明显差异
distance <- "bray"
# Distance matrix for samples


# Calculate sample similarities
dm<- as.matrix(distance(LungCancer_Normalized, distance))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")


picture<-list()

A<-"BAL"
B<-"Fecal"

Type1.pair<-subset(dm,CT.x=="type1"&CT.y=="type1")
Type1.ps1<-subset(Type1.pair,sample.x==A&sample.y==B)
Type1.ps2<-subset(Type1.pair,sample.x==B&sample.y==A)
Type1.ps<-rbind(Type1.ps1,Type1.ps2)
Type1.ps<-subset(Type1.ps,index.x==index.y)





Type2.pair<-subset(dm,CT.x=="type2"&CT.y=="type2")
Type2.ps1<-subset(Type2.pair,sample.x==A&sample.y==B)
Type2.ps2<-subset(Type2.pair,sample.x==B&sample.y==A)
Type2.ps<-rbind(Type2.ps1,Type2.ps2)
Type2.ps<-subset(Type2.ps,index.x==index.y)


ps<-rbind(Type1.ps,Type2.ps)


ggplot(data=ps,aes(x=CT.x,y=value,fill=CT.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT.x),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Distance to BAL")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()














# 拟杆菌在肺部的分布情况

Bacteroides<-data.frame(t(OTU_TAX_Genus_RA["g_Bacteroides",]))
Bacteroides$ID<-row.names(Bacteroides)
Bacteroides<-merge(Bacteroides,MetaTable,by.x = "ID")
Bacteroides<-subset(Bacteroides,sample=="BAL")


ggplot(data=Bacteroides,aes(x=CT,y=g_Bacteroides,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(.~gender,scales = "free_y" )




Bacteroides<-data.frame(t(OTU_TAX_Genus_RA["g_Bacteroides",]))
Bacteroides$ID<-row.names(Bacteroides)
Bacteroides<-merge(Bacteroides,MetaTable,by.x = "ID")
Bacteroides<-subset(Bacteroides,sample=="Fecal")
Bacteroides<-subset(Bacteroides,!(CT=="NA"))

ggplot(data=Bacteroides,aes(x=CT,y=g_Bacteroides,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(.~gender,scales = "free_y" )


















Cancer_lefse<-data.frame(t(OTU_TAX_Genus_RA))
Cancer_lefse$ID<-row.names(Cancer_lefse)
Cancer_lefse<-merge(Cancer_lefse,MetaTable,by.x = "ID")

```




```{r}
# 拟杆菌在肺部的分布情况

Bacteroides<-data.frame(t(OTU_TAX_Genus_RA["g_Bacteroides",]))
Bacteroides$ID<-row.names(Bacteroides)
Bacteroides<-merge(Bacteroides,MetaTable,by.x = "ID")
Bacteroides<-subset(Bacteroides,sample=="Fecal")
Bacteroides<-subset(Bacteroides,!(CT=="NA"))
Bacteroides_type1_yes<-subset(Bacteroides,(CT=="type1")&(gastroenteropathy=="Yes"))
Bacteroides_type2_no<-subset(Bacteroides,(CT=="type2")&(gastroenteropathy=="No"))
Bacteroides_type1_yes$GROUP<-"type1_yes"
Bacteroides_type2_no$GROUP<-"type2_no"
Bacteroides_new<-rbind.data.frame(Bacteroides_type1_yes,Bacteroides_type2_no)

ggplot(data=Bacteroides,aes(x=CT,y=g_Bacteroides,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Relative Abundance")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(.~gender+gastroenteropathy,scales = "free_y" )

```










# 外部数据验证---Run1
```{r}
Otu.biom = read_biom("Run1/feature-table.biom")
#将biom对象转化成矩阵对象
Otu<-as(biom_data(Otu.biom), "matrix")
Tax<-read.table("Run1/Run1_taxonomy.txt",sep = "\t",header = TRUE,row.names = 1)
Sample<-read.table("Run1/Run1_metadata.txt",sep = "\t",header = TRUE,row.names = 1)
Tax<-as.matrix(Tax)
Otu = otu_table(Otu, taxa_are_rows = TRUE)
Tax = tax_table(Tax)
samples = sample_data(Sample)
Run1 <- phyloseq(Otu, Tax, samples)

Run1 <- filter_taxa(Run1,function(x) sum(x) > 10 , TRUE )



OTU<-as.data.frame(otu_table(Run1))
TAX<-as.data.frame(tax_table(Run1))




OTU_TAX<-cbind.data.frame(TAX,OTU)
Run1_Genus<-aggregate(OTU_TAX[,7:228],list(OTU_TAX$Genus),sum)
#去除未注释到属水平的单元
Run1_Genus<-Run1_Genus[-1,]
rownames(Run1_Genus)<-  Run1_Genus$Group.1
Run1_Genus<- Run1_Genus[,-1]
Run1_Genus<-Run1_Genus[,colSums(Run1_Genus)>0]
Run1_Genus<-t(Run1_Genus)
Run1_Genus_RA<-data.frame(Run1_Genus/apply(Run1_Genus,1,sum)*100)
colnames(Run1_Genus_RA)<-colnames(Run1_Genus)
colnames(Run1_Genus_RA)<-gsub(" ", "", colnames(Run1_Genus_RA), fixed = TRUE)
Run1_Genus_RA<-Run1_Genus_RA[,colnames(Run1_Genus_RA)%in%var]


#挑选BAL样本
Run1_BAL<-subset_samples(Run1,Tissue==c("bronchoalveolar lavage of left lung","bronchoalveolar lavage of right lung"))

Run1_Genus_RA<-Run1_Genus_RA[row.names(Run1_Genus_RA)%in%sample.names(Run1_BAL),]

Run1_rf_probs <- predict(count_rf$RFopt,newdata=Run1_Genus_RA, type='prob')[,2]


```


```{r}
Run1_Genus_RA<-data.frame(Run1_Genus/apply(Run1_Genus,1,sum)*100)
colnames(Run1_Genus_RA)<-colnames(Run1_Genus)
colnames(Run1_Genus_RA)<-gsub(" ", "", colnames(Run1_Genus_RA), fixed = TRUE)
Run1_BAL<-subset_samples(Run1,Tissue==c("bronchoalveolar lavage of left lung","bronchoalveolar lavage of right lung"))
Run1_Genus_RA<-Run1_Genus_RA[,colSums(Run1_Genus_RA)>0]
Run1_Genus_RA<-Run1_Genus_RA[row.names(Run1_Genus_RA)%in%sample.names(Run1_BAL),]

Run1_Genus_RA<-Run1_Genus_RA[,colSums(Run1_Genus_RA)>0]


count <- as.matrix((Run1_Genus_RA))
set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace
best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)




 d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d1 <-d1[81:95,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d2 <-d2[81:95,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=22,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=22,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
```



# 检查Run1的多样性差异

```{r}
# 构建Run1 BAL 的phyloseq对象
Run1_type<-data.frame(ass)
Run1_type$ID<-rownames(Run1_type)
Run1_type$ass<-car::recode(Run1_type$ass,"1='type2';2='type1'")
# 将分型信息加入metadata
# 提取metadata
test<-sample_data(Run1_BAL)

test$CT<-Run1_type$ass

Otu = otu_table(Run1_BAL)
Tax = tax_table(Run1_BAL)
samples = sample_data(test)
Run1_BAL <- phyloseq(Otu, Tax, samples)



my_comparisons <- list(c("type1", "type2"))

p<-plot_richness(Run1_BAL,"CT", "CT",nrow = 1,measures=alpha_meas)

p$data<-p$data[-3,]
P_BAL <- ggplot(data=p$data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Observed Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)




# 观察两种亚型的FEV1是否有差异
data<-data.frame(sample_data(Run1_BAL),stringsAsFactors = FALSE)
data$fev1<-as.numeric(data$fev1)
data$Age<-as.numeric(as.character(data$Age))


data$fev1<-lapply(data$fev1, function(x) as.numeric(gsub("\\%", "", x))/100)
data$fev1<-as.numeric(as.character(data$fev1))

ggplot(data,aes(x=CT,y=fev1,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("FEV1")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)





```




# 外部数据验证 -- Run2
```{r}
Otu.biom = read_biom("Run2/feature-table.biom")
#将biom对象转化成矩阵对象
Otu<-as(biom_data(Otu.biom), "matrix")
Tax<-read.table("Run2/Run2_taxonomy.tsv",sep = "\t",header = TRUE,row.names = 1)
Sample<-read.table("Run2/Run2_metadata.txt",sep = "\t",header = TRUE,row.names = 1)
Tax<-as.matrix(Tax)
Otu = otu_table(Otu, taxa_are_rows = TRUE)
Tax = tax_table(Tax)
samples = sample_data(Sample)
Run2 <- phyloseq(Otu, Tax, samples)
Run2 <- rarefy_even_depth(Run2)
Run2 <- filter_taxa(Run2,function(x) sum(x) > 10 , TRUE )



OTU<-as.data.frame(otu_table(Run2))
TAX<-as.data.frame(tax_table(Run2))




OTU_TAX<-cbind.data.frame(TAX,OTU)
Run2_Genus<-aggregate(OTU_TAX[,7:127],list(OTU_TAX$Genus),sum)
#去除未注释到属水平的单元
Run2_Genus<-Run2_Genus[-1,]
rownames(Run2_Genus)<-  Run2_Genus$Group.1
Run2_Genus<- Run2_Genus[,-1]

Run2_Genus<-t(Run2_Genus)
Run2_Genus_RA<-data.frame(Run2_Genus/apply(Run2_Genus,1,sum)*100)
colnames(Run2_Genus_RA)<-colnames(Run2_Genus)
colnames(Run2_Genus_RA)<-gsub(" ", "", colnames(Run2_Genus_RA), fixed = TRUE)
Run2_Genus_RA<-Run2_Genus_RA[,colnames(Run2_Genus_RA)%in%var]


#挑选BAL样本


Run2_Genus_RA<-Run2_Genus_RA[row.names(Run2_Genus_RA)%in%sample.names(Run2_BAL),]

Run2_rf_probs <- predict(count_rf$RFopt,newdata=Run2_Genus_RA, type='prob')[,2]



Run2_type<-data.frame(Run2_rf_probs)
Run2_type$ID<-rownames(Run2_type)
Run2_type$Run2_rf_probs<-car::recode(Run2_type$Run2_rf_probs,"0:0.5='type2';0.5:1='type1'")
# 将分型信息加入metadata
# 提取metadata
test<-data.frame(sample_data(Run2))

test$CT<-Run2_type$Run2_rf_probs

Otu = otu_table(Segal_BAL)
Tax = tax_table(Segal_BAL)
samples = sample_data(test)
Segal_BAL <- phyloseq(Otu, Tax, samples)





DATA<-test[,c("Description","index","CT")]
DATA<-subset(DATA,!(index=="NA"))
```



```{r}
new_MetaTable<-merge(MetaTable,Janyan_data,by.x="index",by.y="ID",all.x=TRUE)
data<-subset(new_MetaTable,(sample=="BAL")&!(diagnosis=="NA")&!(BALgroup=="HS"))

```

# 
```{r}
Both_site<-subset_samples(LungCancer,index%in%BAL_Interindividual$index)
Both_site<-subset_samples(Both_site,sample=="BAL")

Both_site<- filter_taxa(Both_site,function(x) sum(x) > 10 , TRUE )

OTU<-as.data.frame(otu_table(Both_site))
TAX<-as.data.frame(tax_table(Both_site))




OTU_TAX<-cbind.data.frame(TAX,OTU)
OTU_TAX_Genus<-aggregate(OTU_TAX[,7:172],list(OTU_TAX$Genus),sum)
#去除未注释到属水平的单元
OTU_TAX_Genus<-OTU_TAX_Genus[-422:-442,]
OTU_TAX_Genus<-OTU_TAX_Genus[-1,]
rownames(OTU_TAX_Genus)<-  OTU_TAX_Genus$Group.1
OTU_TAX_Genus<- OTU_TAX_Genus
OTU_TAX_Genus<- OTU_TAX_Genus[,-1]

OTU_TAX_Genus_RA<-t(OTU_TAX_Genus)
OTU_TAX_Genus_RA<-OTU_TAX_Genus_RA/apply(OTU_TAX_Genus_RA,1,sum)*100
OTU_TAX_Genus_RA<-as.data.frame(t(OTU_TAX_Genus_RA))



count <- as.matrix(t(OTU_TAX_Genus_RA))
set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)



  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d1 <-d1[420:434,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d2 <-d2[420:434,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))





type<-data.frame(ass)

Both_data<-data.frame(sample_data(Both_site))
Both_data$CT<-type$ass
Both_data<-subset(Both_data,group=="Cancer")


test<-Both_data[,c("index","ID","bodysite","BALgroup","CT")]

test<-dcast(test,index~BALgroup)
test<-subset(test,!(DS==HS))
```


# 肺功能
```{r}

lung_function<-read.table("肺癌肺功能.txt",header=TRUE,row.names=1)

lung_function_merge<-merge(BAL_Interindividual,lung_function,by.x="index",by.y="ID")

ggplot(lung_function_merge,aes(x=lung_function_merge$CT,y=lung_function_merge$FEV1实际,fill=lung_function_merge$CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=lung_function_merge$CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Lung function")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()
```

# 肺功能

```{r}

lung_function<-read.table("肺癌肺功能.txt",header=TRUE,row.names=1)

lung_function_merge<-merge(data,lung_function,by.x="human_id",by.y="ID")




ggplot(lung_function_merge,aes(x=lung_function_merge$CT,y=lung_function_merge$FEV1实际,fill=lung_function_merge$CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=lung_function_merge$CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Lung function")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()

```


# 肺功能和差异菌的相关性
```{R}
# 读取BAL lefse结果
BAL_lefse<-read.table("BAL_LFfSe_result.txt",sep = "\t")
# 筛选有差异的菌
BAL_lefse<-subset(BAL_lefse,V3%in%c("type1","type2"))

# 提取差异菌的idx
Genus_idx<-BAL_lefse$V1
# 提取差异菌的相对丰度
Genus_lefse<-OTU_TAX_Genus_RA[row.names(OTU_TAX_Genus_RA)%in%Genus_idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL)]


#提取α多样性
p<-plot_richness(LungCancer_Normalized_BAL, "group", "group",nrow = 1,measures=alpha_meas)
alpha_div<-p$data
alpha_div<-alpha_div[,45:47]
alpha_div<-dcast(alpha_div,samples~variable)

#合并
Genus_lefse<-as.data.frame(t(Genus_lefse))
Genus_alpha<-cbind.data.frame(Genus_lefse,alpha_div)



#合并肺功能
lung_function_RP<-lung_function_merge[,c("ID","CT","gender","FEV1实际.预测" ,"FEV1.FVC实际预测","FVC实际.预测" ,"PEF实际.预测" ,"MEF75实际.预测","MEF50实际.预测","MEF25实际.预测","MEF75.25实际.预测","MVV实际.预测","TCL实际.预测","RV实际.预测","RV.TLC实际.预测","FRC实际.预测","FRC.TLC实际.预测","DLCO.SB实际.预测","DLCO.VA实际.预测" )]
merge_all<-merge(lung_function_RP,Genus_alpha,by.x="ID",by.y="samples")


merge_type1<-subset(merge_all,CT=="type1")
phylum_corr_cancer <- corr.test(merge_type1[,4:19],merge_type1[,20:74], method = 'spearman')
corrplot(phylum_corr_cancer$r, p.mat=phylum_corr_cancer$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")


ggplot(merge_all, aes(x=DLCO.SB实际.预测, y= `g_Prevotella 7`,color=CT)) +
    geom_point() +
    geom_smooth(method="lm") +
    theme(panel.grid=element_blank(), panel.background=element_rect(fill='transparent', color='black')) +stat_cor(method = "spearman")

```





```{r}

LungCancer_Normalized_Oroph<-subset_samples(LungCancer_Normalized,sample=="Oroph"&CT%in%c("type1","type2"))



p<-plot_richness(LungCancer_Normalized_Oroph, "sample", "sample",nrow = 1,measures=alpha_meas)

data<-p$data

ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
    #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
    scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("BAL")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means()

p<-plot_richness(LungCancer_Normalized_Oroph, "sample", "sample",nrow = 1,measures=alpha_meas)



set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Oroph, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Oroph, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Oroph, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-26
pc2<-11.6


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = CT),geom = "polygon",level = 0.95,alpha = 0.3)


set.seed(2010)
otu.adonis<-adonis(dis_bray ~ CT, data = data.frame(sample_data(LungCancer_Normalized_Oroph)))
Oroph_adonis<-otu.adonis$aov.tab$R2










# lefse 

OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph)]
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_Oroph[which(rowSums(OTU_TAX_Genus_Oroph) > 0),]
metadata_Oroph<-data.frame(sample_data(LungCancer_Normalized_Oroph))
OTU_TAX_Genus_Oroph["CT",]<-metadata_Oroph$CT


write.table(OTU_TAX_Genus_Oroph,"lefse_20201125/OTU_TAX_Genus_Oroph.txt",sep = "\t")



OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph)]
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_Oroph[which(rowSums(OTU_TAX_Genus_Oroph) > 0),]
OTU_TAX_Genus_Oroph$Genus<-rownames(OTU_TAX_Genus_Oroph)

OTU_TAX_Genus_Oroph<-melt(OTU_TAX_Genus_Oroph,id.vars ="Genus")

OTU_TAX_Genus_Oroph<-merge(OTU_TAX_Genus_Oroph,metadata_Oroph,by.x="variable",by.y="ID")





results<-compare_means(value~CT, data=OTU_TAX_Genus_Oroph,group.by = "Genus")
results<-subset(results,p < 0.1)


data<-subset(OTU_TAX_Genus_Oroph,Genus%in%results$Genus)

 ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(.~Genus,scales = "free_y")







#计算两种亚型配对Oroph和Saliva的距离

LungCancer_Normalized_OS<-subset_samples(LungCancer_Normalized,sample%in%c("Oroph","Saliva")&CT%in%c("type1","type2"))



dis_bray<- phyloseq::distance(LungCancer_Normalized_OS, "bray")

dm<- as.matrix(dis_bray)
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")



A<-"Oroph"
B<-"Saliva"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_OS<-rbind(Health.ps,Cancer.ps)


ggplot(ps_OS,aes(x=CT.x,y=value,fill=CT.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT.x),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()



LungCancer_Normalized_ON<-subset_samples(LungCancer_Normalized,sample%in%c("Oroph","Nasal")&CT%in%c("type1","type2"))



dis_bray<- phyloseq::distance(LungCancer_Normalized_ON, "bray")

dm<- as.matrix(dis_bray)
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")

A<-"Oroph"
B<-"Nasal"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_OS<-rbind(Health.ps,Cancer.ps)

ps_OS<-ps_OS[-78,]
ps_OS<-ps_OS[-77,]
ps_OS<-ps_OS[-37,]
ggplot(ps_OS,aes(x=CT.x,y=value,fill=CT.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT.x),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()




###### type1 type2 的Oroph 和Nasal


LungCancer_Normalized_ON<-subset_samples(LungCancer_Normalized,sample%in%c("Oroph","Nasal")&CT%in%c("type1","type2"))
LungCancer_Normalized_ON<- filter_taxa(LungCancer_Normalized_ON,function(x) sum(x) > 10 , TRUE )

dis_bray<- phyloseq::distance(LungCancer_Normalized_ON, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_ON, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_ON, dis_bray.pcoa, color="CT" ,shape = "sample") + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 过滤掉没有分型的数据
data<-subset(data,!(CT=="NA"))

## 用样本类型和分型结果重命名

data<-tidyr::unite(data, "sample_CT", sample, CT,remove=FALSE)

data$sample_CT<-as.factor(data$sample_CT)
## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-38.9
pc2<-7.5


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample_CT,shape=sample_CT,fill=sample_CT),size=4)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample_CT),geom = "polygon",level = 0.95,alpha = 0.3)


```




```{r}

LungCancer_Normalized_Saliva<-subset_samples(LungCancer_Normalized,sample=="Saliva"&CT%in%c("type1","type2"))



p<-plot_richness(LungCancer_Normalized_Saliva, "sample", "sample",nrow = 1,measures=alpha_meas)

data<-p$data

ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
    #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
    scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("BAL")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means()





set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Saliva, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Saliva, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Saliva, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-14.8
pc2<-10.3


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = CT),geom = "polygon",level = 0.95,alpha = 0.3)


set.seed(2010)
otu.adonis<-adonis(dis_bray ~ CT, data = data.frame(sample_data(LungCancer_Normalized_Saliva)))
Saliva_adonis<-otu.adonis$aov.tab$R2


```


```{r}
LungCancer_Normalized_Nasal<-subset_samples(LungCancer_Normalized,sample=="Nasal"&CT%in%c("type1","type2"))



p<-plot_richness(LungCancer_Normalized_Nasal, "sample", "sample",nrow = 1,measures=alpha_meas)

data<-p$data

ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
    #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
    scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("BAL")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means()





set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(LungCancer_Normalized_Nasal, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_Nasal, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_Nasal, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-20.3
pc2<-13.3


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = CT),geom = "polygon",level = 0.95,alpha = 0.3)


set.seed(2010)
otu.adonis<-adonis(dis_bray ~ CT, data = data.frame(sample_data(LungCancer_Normalized_Nasal)))
Nasal_adonis<-otu.adonis$aov.tab$R2











#LEFSE

OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal)]
OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_Nasal[which(rowSums(OTU_TAX_Genus_Nasal) > 0),]
metadata_Nasal<-data.frame(sample_data(LungCancer_Normalized_Nasal))
OTU_TAX_Genus_Nasal["CT",]<-metadata_Nasal$CT


write.table(OTU_TAX_Genus_Nasal,"lefse_20201125/OTU_TAX_Genus_Nasal.txt",sep = "\t")




OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal)]
OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_Nasal[which(rowSums(OTU_TAX_Genus_Nasal) > 0),]
OTU_TAX_Genus_Nasal$Genus<-rownames(OTU_TAX_Genus_Nasal)

OTU_TAX_Genus_Nasal<-melt(OTU_TAX_Genus_Nasal,id.vars ="Genus")

OTU_TAX_Genus_Nasal<-merge(OTU_TAX_Genus_Nasal,metadata_Nasal,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=OTU_TAX_Genus_Nasal,group.by = "Genus")
results<-subset(results,p < 0.1)


data<-subset(OTU_TAX_Genus_Nasal,Genus%in%results$Genus)

 ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(.~Genus,scales = "free_y")
```




# 咽拭子中性模型
```{r}
LungCancer_Normalized_Oroph<-subset_samples(LungCancer_Normalized,sample=="Oroph")


# Type1
LungCancer_Normalized_BAL_type1<-subset_samples(LungCancer_Normalized_BAL,CT=="type1")
LungCancer_Normalized_Oroph_type1<-subset_samples(LungCancer_Normalized_Oroph,CT=="type1")
spp <- t(otu_table(LungCancer_Normalized_BAL_type1))
pool <-t(otu_table(LungCancer_Normalized_Oroph_type1))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1)))
P_BS_type1 <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
type1_data<-P_BS_type1$data

# type2
LungCancer_Normalized_BAL_type2<-subset_samples(LungCancer_Normalized_BAL,CT=="type2")
LungCancer_Normalized_Oroph_type2<-subset_samples(LungCancer_Normalized_Oroph,CT=="type2")
spp <- t(otu_table(LungCancer_Normalized_BAL_type2))
pool <-t(otu_table(LungCancer_Normalized_Oroph_type2))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type2)))
P_BS_type2 <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
type2_data<-P_BS_type2$data

```


# 鼻拭子中性模型
```{r}
LungCancer_Normalized_Nasal<-subset_samples(LungCancer_Normalized,sample=="Nasal")


# Type1
LungCancer_Normalized_BAL_type1<-subset_samples(LungCancer_Normalized_BAL,CT=="type1")
LungCancer_Normalized_Nasal_type1<-subset_samples(LungCancer_Normalized_Nasal,CT=="type1")
spp <- t(otu_table(LungCancer_Normalized_BAL_type1))
pool <-t(otu_table(LungCancer_Normalized_Nasal))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type1)))
P_BS_type1 <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
type1_data<-P_BS_type1$data

# type2
LungCancer_Normalized_BAL_type2<-subset_samples(LungCancer_Normalized_BAL,CT=="type2")
LungCancer_Normalized_Nasal_type2<-subset_samples(LungCancer_Normalized_Nasal,CT=="type2")
spp <- t(otu_table(LungCancer_Normalized_BAL_type2))
pool <-t(otu_table(LungCancer_Normalized_Nasal))
spp.out <- fit_sncm(spp, pool=pool, taxon=data.frame(tax_table(LungCancer_Normalized_BAL_type2)))
P_BS_type2 <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())
type2_data<-P_BS_type2$data

```


```{r}

LungCancer_Normalized_Nasal_type1_tax<-taxa_names(prune_taxa(taxa_sums(LungCancer_Normalized_Nasal_type1) > 55, LungCancer_Normalized_Nasal_type1))
LungCancer_Normalized_Nasal_type2_tax<-taxa_names(prune_taxa(taxa_sums(LungCancer_Normalized_Nasal_type2) >26, LungCancer_Normalized_Nasal_type2))


LungCancer_Normalized_Oroph_type1_tax<-taxa_names(prune_taxa(taxa_sums(LungCancer_Normalized_Oroph_type1) > 55, LungCancer_Normalized_Oroph_type1))
LungCancer_Normalized_Oroph_type2_tax<-taxa_names(prune_taxa(taxa_sums(LungCancer_Normalized_Oroph_type2) > 26, LungCancer_Normalized_Oroph_type2))




T<-venn.diagram(list(A=LungCancer_Normalized_Oroph_type1_tax,B=LungCancer_Normalized_Nasal_type1_tax),
filename=NULL,
lwd=1,#圈线粗度
lty=1, #圈线类型
col=c('#0099CC','#FF6666'), #圈线颜色
fill=c('#0099CC','#FF6666'), #填充颜色
cat.col=c('#0099CC','#FF6666'),#A和B的颜色
cat.cex = 2.5,# A和B的大小
rotation.degree = 0,#旋转角度
main = "A&B",#主标题内容
main.cex = 2,#主标题大小
sub = "plot : example",#亚标题内容
sub.cex = 1,#亚标题字大小
cex=1.5,#里面交集字的大小
alpha = 0.5,#透明度
reverse=TRUE)
grid.draw(T)
```





# 证明鼻咽菌群趋同是双向的
```{r}
# 首先计算type2 Oroph到type2 Nasal的距离
distance <- "bray"
# Distance matrix for samples


# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(LungCancer_Normalized, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")



A<-"Oroph"
B<-"Nasal"

data<-subset(dm,(sample.x==A&sample.y==B)|(sample.x==B&sample.y==A))
data<-subset(data,CT.x=="type2"&CT.y=="type2")



ps_ON_O2N2<-data

#去除123这个样本
ps_ON_O2N2<-subset(ps_ON_O2N2,!(index.x=="LC123"))

ps_ON_O2N2<-subset(ps_ON_O2N2,!(index.y=="LC123"))

ps_ON_O2N2$Dis_group<-"O2N2"
# 计算type1 Oroph到type2 Nasal的距离

distance <- "bray"
# Distance matrix for samples


# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(LungCancer_Normalized, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")



A<-"Oroph"
B<-"Nasal"

data<-subset(dm,((sample.x==A&sample.y==B)&(CT.x=="type1"&CT.y=="type2"))|((sample.x==B&sample.y==A)&(CT.x=="type2"&CT.y=="type1")))
ps_ON_O1N2<-data
ps_ON_O1N2$Dis_group<-"O1N2"
ps_ON_O1N2<-subset(ps_ON_O1N2)




# 计算type1 Oroph到type2 Nasal的距离

distance <- "bray"
# Distance matrix for samples


# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(LungCancer_Normalized, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")



A<-"Oroph"
B<-"Nasal"

data<-subset(dm,((sample.x==A&sample.y==B)&(CT.x=="type2"&CT.y=="type1"))|((sample.x==B&sample.y==A)&(CT.x=="type1"&CT.y=="type2")))

ps_ON_O2N1<-data
ps_ON_O2N1$Dis_group<-"O2N1"







ps_ON<-rbind.data.frame(ps_ON_O1N2,ps_ON_O2N1,ps_ON_O2N2)


my_comparisons <- list(c("O1N2", "O2N1"),c("O2N1", "O2N2"),c("O1N2", "O2N2"))

ggplot(ps_ON,aes(x=Dis_group,y=value,fill=Dis_group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill=Dis_group,outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=Dis_group),width =0.2,shape = 21,size=2.5)+
  scale_fill_npg()#设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means(comparisons=my_comparisons)
```




# 细胞因子结果
```{r}
# 读取细胞因子

cytokine<-read.table("48因子检测结果.txt",header=TRUE,row.names=1,sep="\t")

for ( i in  2:49){
  
  cytokine[ grep("<",cytokine[,i ]),i ]<-"NA"
  
}

for ( i in  2:49){
  
  cytokine[ grep(">",cytokine[,i ]),i ]<-"NA"
  
}

write.table(cytokine,"48因子检测结果_去除小于阈值.txt",sep="\t")

cytokine<-read.table("48因子检测结果_去除小于阈值.txt",header=TRUE,row.names=1,sep="\t")


cytokine<- merge(cytokine,BAL_Interindividual,by="ID")
# 去除检出率过低的因子
data<-cytokine[,2:49]
data<-data.frame(t(data)) 

count<-c()


 data[apply(data, 1, function(y) !all(is.na(y))),]
for (i in 1:47) {
  
  count[i]<-table(!(data[i,]=="NA"))[[1]]
  
}
 
data<-subset(data,count>=30)
data$count<-count
ggplot(cytokine,aes(x=CT,y=cytokine$sCD40L,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()




data<-cytokine[,c(1,41:49)]
data<-data[,!(colnames(data)%in%c("EGF","GM.CSF","IL.1a","IL.1b","IP.10","RANTES","IL.1RA","PDGF.AA"))]
CT<-cytokine[,c("ID","CT","group")]
data<-melt(data,id="ID")
data<-merge(data,CT,by="ID")

ggplot(data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(~variable,scales = "free",nrow = 1)


idx<-c("g_Prevotella","g_Prevotella 7","g_Prevotella 6","g_Prevotella 2","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rotiha","g_Allprevotella","g_Leptotrichia","g_Butyrivibrio 2","g_Oribacterium","g_Treponema 2")

#细胞因子与菌的相关性分析
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL[which(rowSums(OTU_TAX_Genus_BAL) > 10),]
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL)]

data<-data.frame(t(OTU_TAX_Genus_BAL))

data$ID<-rownames(data)
data<-merge(cytokine,data,by="ID")
data<-merge(data,MetaTable,by="ID")
data<-subset(data,CT=="type1")
phylum_corr_cancer <- corr.test(data[,2:49],data[,50:60], method = 'spearman')
corrplot(phylum_corr_cancer$r, p.mat=phylum_corr_cancer$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")


ggplot(data, aes(x=EGF, y= g_Prevotella.7,color=CT,group=CT)) +
    geom_point(aes(fill=CT),shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1,aes(fill=CT)) +
theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")



heatmap_data<-cytokine
heatmap_data$ID<-rownames(heatmap_data)
heatmap_data<- merge(heatmap_data,BAL_Interindividual,by="ID")

heatmap_data<-heatmap_data[,c(2:34,78)]


heatmap_data[is.na(heatmap_data)] <- 0.1
heatmap_data[,1:33]<-log10(heatmap_data[,1:33])
rownames(heatmap_data)<-rownames(cytokine)
# 根据CT重排序
heatmap_data<-heatmap_data[order(heatmap_data$CT),]

heatmap_data<-heatmap_data[c(-52,-59,-86),]
CT = data.frame(CT =heatmap_data[,"CT"] )
row.names(CT )<-row.names(heatmap_data )

ann_colors = list(CT = c(type1 = "red", type2 = "green"))

# 选取差异较大的因子
idx<-c("EGF","IL.1a","IL.1b","IP.10","RANTES","IL.1RA","PDGF.AA","TNFb","TGFa","VEGF.A","MCP.1","CXCL9","CT")
heatmap_data<-heatmap_data[,colnames(heatmap_data)%in%idx]

pheatmap(data.frame(t(heatmap_data[,1:12])),scale="row", annotation_col =   CT ,cluster_cols  = FALSE,color =  colorRampPalette(colors = c("blue","white","red"))(100),annotation_colors = ann_colors)


pheatmap(heatmap_data[,1:33],scale="column",color =c(colorRampPalette(colors = c("blue","white"))(length(bk)/2),colorRampPalette(colors = c("white","red"))(length(bk)/2)),legend_breaks=seq(-1,5,1),cluster_row = FALSE,breaks=bk)








# 细胞因子PCOA


dist<-(vegdist(heatmap_data[c(-4,-86),colnames(heatmap_data)%in%idx], method = "bray"))


## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(subset_samples(LungCancer_Normalized_BAL,ID%in%rownames(heatmap_data)), method="NMDS", distance=dist)

## 绘制初始图形
bray.pcoa <- plot_ordination(subset_samples(LungCancer_Normalized_BAL,ID%in%rownames(heatmap_data)), dis_bray.pcoa, color="CT" ) + geom_point(size=3)
bray.pcoa 
## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("NMDS1","NMDS2")

## 获取主坐标轴1,2的解释度
pc1<-17
pc2<-7


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
ggplot(data, aes(NMDS1, NMDS2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("NMDS1",sep="")) + 
  ylab(paste("NMDS2",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))


ggscatter(data, x = "NMDS1", y = "NMDS2",
          color = "CT",
          ellipse = FALSE, mean.point = TRUE,
          star.plot = TRUE)






set.seed(2020)
otu.adonis<-adonis(dist ~ CT, data = data.frame(sample_data(subset_samples(LungCancer_Normalized_BAL,ID%in%rownames(heatmap_data[c(-4,-86),])))))

```

# 制作粪便模型

```{r}
LungCancer_Normalized_Fecal<-subset_samples(LungCancer_Normalized,sample=="Fecal")
data<-data.frame(sample_data(LungCancer_Normalized_Fecal))


OTU_TAX_Genus_Fecal<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample_names(LungCancer_Normalized_Fecal)]
OTU_TAX_Genus_Fecal<-OTU_TAX_Genus_Fecal[which(rowSums(OTU_TAX_Genus_Fecal) > 0),]

count<-data.frame(t(OTU_TAX_Genus_Fecal))
count$CT<-data$CT
count$CT<-ifelse(count$CT == "type1", 1, 0)
count$CT<-as.factor(count$CT)
count<-subset(count,!(CT=="NA"))
colnames( count)<-gsub(" ", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("-", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("(", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub(")", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("[", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("]", "", colnames( count), fixed = TRUE)


set.seed(2020)
count_rf <- AUCRF(CT~., data=count, ranking='MDA', ntree=500, pdel=0.2)
count_rf
count_rf_probs <- predict(count_rf$RFopt, type='prob')[,2]
count_rf_roc <- roc(count$CT~count_rf_probs,smooth=FALSE )

devroc1 <-plot.roc(count_rf_roc, main="dev ROC", percent=TRUE, col="1")




ggplot(count,aes(x=CT,y=count$g_Bifidobacterium,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(~variable,scales = "free")
```





#  双向流向菌箱线图
```{r}
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph)]
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_Oroph[which(rowSums(OTU_TAX_Genus_Oroph) > 0),]
OTU_TAX_Genus_Oroph$Genus<-rownames(OTU_TAX_Genus_Oroph)

OTU_TAX_Genus_Oroph<-melt(OTU_TAX_Genus_Oroph,id.vars ="Genus")

OTU_TAX_Genus_Oroph<-merge(OTU_TAX_Genus_Oroph,metadata_Oroph,by.x="variable",by.y="ID")


results<-compare_means(value~CT, data=OTU_TAX_Genus_Oroph,group.by = "Genus")
results.Oroph<-subset(results,p < 0.1)





OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal)]
OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_Nasal[which(rowSums(OTU_TAX_Genus_Nasal) > 0),]
OTU_TAX_Genus_Nasal$Genus<-rownames(OTU_TAX_Genus_Nasal)

OTU_TAX_Genus_Nasal<-melt(OTU_TAX_Genus_Nasal,id.vars ="Genus")

OTU_TAX_Genus_Nasal<-merge(OTU_TAX_Genus_Nasal,metadata_Nasal,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=OTU_TAX_Genus_Nasal,group.by = "Genus")
results.Nasal<-subset(results,p < 0.05)




# 挑选需要展示的细菌

idx<-c("g_Prevotella 2","g_Veillonella","g_Porphyromonas","g_Corynebacterium 1","g_SR1 bacterium oral taxon 875","g_Cutibacterium")

data<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_ON)]

data$Genus<-row.names(data)

data<-melt(data,id.vars = "Genus")

data<-merge(data,MetaTable,by.x="variable",by.y="ID")

data<-tidyr::unite(data, "sample_CT", sample, CT,remove=FALSE)

data$sample_CT<-as.factor(data$sample_CT)

data$sample_CT<-factor(data$sample_CT,levels=c("Oroph_type2","Oroph_type1","Nasal_type1","Nasal_type2"))

data$value[which(data$value==0)]

my_comparisons <- list(c("Oroph_type2", "Oroph_type1"),c("Nasal_type1", "Nasal_type2"))

ggplot(data,aes(x=sample_CT,y=log10(value),fill=sample_CT))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=sample_CT),width =0.2,shape = 21,size=2.5)+
    scale_fill_npg()+#设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
#scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("BAL")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means(comparisons=my_comparisons)+facet_wrap(~Genus)+stat_summary(fun=median, geom="line", aes(group=1))  + 
    stat_summary(fun=median, geom="point")



BN<-intersect(sample_data(LungCancer_Normalized_Nasal)$index,sample_data(LungCancer_Normalized_BAL)$index)





idx<-c("g_Prevotella 2","g_Veillonella","g_Porphyromonas","g_Corynebacterium 1","g_SR1 bacterium oral taxon 875","g_Cutibacterium")

data_N<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(subset_samples(LungCancer_Normalized_Nasal,index%in%BN))]

data_N<-data.frame(t(data_N))


idx<-c("g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rothia","g_Leptotrichia","g_Streptococcus","g_Lactobacillus","g_Acinetobacter","g_Pseudomonas")


data_B<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(subset_samples(LungCancer_Normalized_BAL,index%in%BN))]


results<-corr.test(data_N[,1:6],data_B[,1:10],adjust = "none",method = "spearman")
View(results$p)

corrplot(results$r, p.mat=results$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")


data_B<-data.frame(t(data_B))



BN<-intersect(sample_data(LungCancer_Normalized_Oroph)$index,sample_data(LungCancer_Normalized_BAL)$index)

idx<-c("g_Prevotella 2","g_Veillonella","g_Porphyromonas","g_Corynebacterium 1","g_SR1 bacterium oral taxon 875","g_Cutibacterium")

data_N<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(subset_samples(LungCancer_Normalized_Oroph,index%in%BN))]

data_N<-data.frame(t(data_N))


idx<-c("g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rothia","g_Leptotrichia","g_Streptococcus","g_Lactobacillus","g_Acinetobacter","g_Pseudomonas")


data_B<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(subset_samples(LungCancer_Normalized_BAL,index%in%BN))]
data_B<-data.frame(t(data_B))

results<-corr.test(data_N[,1:6],data_B[,1:10],adjust = "none",method = "spearman")
View(results$p)

corrplot(results$r, p.mat=results$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")



```




# 双向流动菌的冲击图

```{r}
data_Nasal_type1<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal_type1)]
data_Nasal_type1$mean<-apply(data_Nasal_type1,1,mean)
data_Nasal_type1$group<-"Nasal_type1"
data_Nasal_type1<-data_Nasal_type1[,c("mean","group")]
data_Nasal_type1$Genus<-rownames(data_Nasal_type1)


data_Nasal_type2<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal_type2)]
data_Nasal_type2$mean<-apply(data_Nasal_type2,1,mean)
data_Nasal_type2$group<-"Nasal_type2"
data_Nasal_type2<-data_Nasal_type2[,c("mean","group")]
data_Nasal_type2$Genus<-rownames(data_Nasal_type2)

data_Oroph_type1<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph_type1)]
data_Oroph_type1$mean<-apply(data_Oroph_type1,1,mean)
data_Oroph_type1$group<-"Oroph_type1"
data_Oroph_type1<-data_Oroph_type1[,c("mean","group")]
data_Oroph_type1$Genus<-rownames(data_Oroph_type1)



data_Oroph_type2<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph_type2)]
data_Oroph_type2$mean<-apply(data_Oroph_type2,1,mean)
data_Oroph_type2$group<-"Oroph_type2"
data_Oroph_type2<-data_Oroph_type2[,c("mean","group")]
data_Oroph_type2$Genus<-rownames(data_Oroph_type2)
data<-rbind.data.frame(data_Nasal_type1,data_Nasal_type2,data_Oroph_type1,data_Oroph_type2)




data$group<-factor(data$group,levels=c("Oroph_type2","Oroph_type1","Nasal_type1","Nasal_type2"))






ggplot(data = Refugees,
       aes(x = year, y = refugees, alluvium = country)) +
    geom_alluvium(aes(fill = country, colour = country),
                  alpha = .75, decreasing = FALSE) +
    scale_x_continuous(breaks = seq(1, 4, 1)) +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -30, hjust = 0)) +
    scale_fill_brewer(type = "qual", palette = "Set3") +
    scale_color_brewer(type = "qual", palette = "Set3")  +
    ggtitle("refugee volume by country and region of origin")



ggplot(data = data,
       aes(x = group, y = mean, alluvium = Genus)) +
    geom_alluvium(aes(fill =Genus, colour = Genus),
                  alpha = .75, decreasing = FALSE) +
    theme_bw() +
    theme(axis.text.x = element_text(colour="black",family="Times",size=14,angle = -30, hjust = 0)) +
    scale_fill_brewer(type = "qual", palette = "Set3") +
    scale_color_brewer(type = "qual", palette = "Set3") +
    ggtitle("")+scale_y_continuous(expand = c(0, 0))+ylab("Mean Abundance (%)")

```



# 四部位菌群
```{r}

LungCancer_Normalized_BAL_type1
BAL_type1_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL_type1)]

BAL_type1_tax$BAL_type1<-apply(BAL_type1_tax,1,mean)
BAL_type1_tax<-format(BAL_type1_tax, scientific=F)


BAL_type1_tax$CT<-"type1"


BAL_type1_tax<-BAL_type1_tax[,c("BAL_type1","CT")]


BAL_type2_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL_type2)]

BAL_type2_tax$BAL_type2<-apply(BAL_type2_tax,1,mean)
BAL_type2_tax<-format(BAL_type2_tax, scientific=F)
BAL_type2_tax$CT<-"type2"


BAL_type2_tax<-BAL_type2_tax[,c("BAL_type2","CT")]

BAL_tax<-cbind.data.frame(BAL_type1_tax$BAL_type1,BAL_type2_tax$BAL_type2)
colnames(BAL_tax)<-c("BAL_type1","BAL_type2")
rownames(BAL_tax)<-rownames(BAL_type1_tax)



Saliva_type1_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Saliva_type1)]

Saliva_type1_tax$Saliva_type1<-apply(Saliva_type1_tax,1,mean)
Saliva_type1_tax<-format(Saliva_type1_tax, scientific=F)


Saliva_type1_tax$CT<-"type1"


Saliva_type1_tax<-Saliva_type1_tax[,c("Saliva_type1","CT")]


Saliva_type2_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Saliva_type2)]

Saliva_type2_tax$Saliva_type2<-apply(Saliva_type2_tax,1,mean)
Saliva_type2_tax<-format(Saliva_type2_tax, scientific=F)
Saliva_type2_tax$CT<-"type2"


Saliva_type2_tax<-Saliva_type2_tax[,c("Saliva_type2","CT")]

Saliva_tax<-cbind.data.frame(Saliva_type1_tax$Saliva_type1,Saliva_type2_tax$Saliva_type2)
colnames(Saliva_tax)<-c("Saliva_type1","Saliva_type2")
rownames(Saliva_tax)<-rownames(Saliva_type1_tax)



Nasal_type1_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal_type1)]

Nasal_type1_tax$Nasal_type1<-apply(Nasal_type1_tax,1,mean)
Nasal_type1_tax<-format(Nasal_type1_tax, scientific=F)


Nasal_type1_tax$CT<-"type1"


Nasal_type1_tax<-Nasal_type1_tax[,c("Nasal_type1","CT")]


Nasal_type2_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Nasal_type2)]

Nasal_type2_tax$Nasal_type2<-apply(Nasal_type2_tax,1,mean)
Nasal_type2_tax<-format(Nasal_type2_tax, scientific=F)
Nasal_type2_tax$CT<-"type2"


Nasal_type2_tax<-Nasal_type2_tax[,c("Nasal_type2","CT")]

Nasal_tax<-cbind.data.frame(Nasal_type1_tax$Nasal_type1,Nasal_type2_tax$Nasal_type2)
colnames(Nasal_tax)<-c("Nasal_type1","Nasal_type2")
rownames(Nasal_tax)<-rownames(Nasal_type1_tax)





Oroph_type1_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph_type1)]

Oroph_type1_tax$Oroph_type1<-apply(Oroph_type1_tax,1,mean)

Oroph_type1_tax<-format(Oroph_type1_tax, scientific=F)

Oroph_type1_tax$CT<-"type1"


Oroph_type1_tax<-Oroph_type1_tax[,c("Oroph_type1","CT")]


Oroph_type2_tax<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph_type2)]

Oroph_type2_tax$Oroph_type2<-apply(Oroph_type2_tax,1,mean)
Oroph_type2_tax<-format(Oroph_type2_tax, scientific=F)
Oroph_type2_tax$CT<-"type2"


Oroph_type2_tax<-Oroph_type2_tax[,c("Oroph_type2","CT")]

Oroph_tax<-cbind.data.frame(Oroph_type1_tax$Oroph_type1,Oroph_type2_tax$Oroph_type2)
colnames(Oroph_tax)<-c("Oroph_type1","Oroph_type2")
rownames(Oroph_tax)<-rownames(Oroph_type1_tax)


# 合并四个位点的平均值
FOUR_SITE<-cbind(BAL_tax,Saliva_tax,Oroph_tax,Nasal_tax)
FOUR_SITE<-apply(FOUR_SITE,2,as.numeric)

rownames(FOUR_SITE)<-rownames(BAL_tax)
# 过滤相对丰度太低的菌
FOUR_SITE<-as.data.frame(FOUR_SITE)
FOUR_SITE<-FOUR_SITE[which(rowSums(FOUR_SITE) > 4),]

FOUR_SITE$BAL<-FOUR_SITE[,1]/FOUR_SITE[,2]
FOUR_SITE$Saliva<-FOUR_SITE[,3]/FOUR_SITE[,4]
FOUR_SITE$Oroph<-FOUR_SITE[,5]/FOUR_SITE[,6]
FOUR_SITE$Nasal<-FOUR_SITE[,7]/FOUR_SITE[,8]

FOUR_SITE<-FOUR_SITE[is.finite(FOUR_SITE$BAL),]
FOUR_SITE<-FOUR_SITE[is.finite(FOUR_SITE$Saliva),]
FOUR_SITE<-FOUR_SITE[is.finite(FOUR_SITE$Oroph),]
FOUR_SITE<-FOUR_SITE[is.finite(FOUR_SITE$Nasal),]

FOUR_SITE$BAL[which(FOUR_SITE$BAL <1)]<--(1/(FOUR_SITE$BAL[which(FOUR_SITE$BAL <1)]))
FOUR_SITE$Saliva[which(FOUR_SITE$Saliva <1)]<--(1/(FOUR_SITE$Saliva[which(FOUR_SITE$Saliva <1)]))
FOUR_SITE$Oroph[which(FOUR_SITE$Oroph <1)]<--(1/(FOUR_SITE$Oroph[which(FOUR_SITE$Oroph <1)]))
FOUR_SITE$Nasal[which(FOUR_SITE$Nasal <1)]<--(1/(FOUR_SITE$Nasal[which(FOUR_SITE$Nasal <1)]))
FOUR_SITE_plot<-FOUR_SITE[,c("BAL","Saliva","Oroph","Nasal")]
#FOUR_SITE_plot<-subset(FOUR_SITE_plot,(Oroph>1&Nasal<1)|(Oroph<1&Nasal>1))
Heatmap(FOUR_SITE_plot,col = colorRamp2(c(-4, 1, 4), c("blue", "white", "red")),cluster_columns =   FALSE)
```



```{r}
OTU<-as.data.frame(otu_table(test))
TAX<-as.data.frame(tax_table(test))


test_TAX<-cbind.data.frame(TAX,OTU)
test_TAX_Genus<-aggregate(test_TAX[,7:92],list(test_TAX$Genus),sum)
#去除未注释到属水平的单元
test_TAX_Genus<-test_TAX_Genus[-1441:-1584,]
test_TAX_Genus<-test_TAX_Genus[-1,]
rownames(test_TAX_Genus)<-  test_TAX_Genus$Group.1
test_TAX_Genus<- test_TAX_Genus
test_TAX_Genus<- test_TAX_Genus[,-1]





A<-test_TAX[,1:6]
B<-data.frame(rownames(test_TAX_Genus))

c<-merge(B,A,by.x="rownames.test_TAX_Genus.",by.y="Genus")

TAX<-c[!duplicated(c$rownames.test_TAX_Genus.), ]



test_TAX_Genus_RA<-t(test_TAX_Genus)
test_TAX_Genus_RA<-test_TAX_Genus_RA/apply(test_TAX_Genus_RA,1,sum)*100
test_TAX_Genus_RA<-as.data.frame(t(test_TAX_Genus_RA))










data<-test_TAX_Genus_RA["g_Granulicatella",colnames(test_TAX_Genus_RA)%in%sample.names(test)]
data<-data.frame(t(data))
sample<-data.frame(sample_data(test))

data$ABX<-sample$抗生素使用

ggplot(data,aes(x=ABX,y=g_Granulicatella,fill=ABX))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=ABX),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()
```






```{r}
results<-rbind.data.frame(results.Nasal,results.Oroph)
results<-subset(results,p.format<0.05)

idx<-results$Genus

idx<-c("g_Prevotella 2","g_Veillonella","g_Porphyromonas","g_Corynebacterium 1","g_SR1 bacterium oral taxon 875","g_Cutibacterium")
LungCancer_Normalized_ON<-subset_samples(LungCancer_Normalized,sample%in%c("Oroph","Nasal")&CT%in%c("type1","type2"))
LungCancer_Normalized_ON<- filter_taxa(LungCancer_Normalized_ON,function(x) sum(x) > 10 , TRUE )
LungCancer_Normalized_ON<-subset_taxa(LungCancer_Normalized_ON,Genus%in%idx)
dis_bray<- phyloseq::distance(LungCancer_Normalized_ON, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(LungCancer_Normalized_ON, method="NMDS", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(LungCancer_Normalized_ON, dis_bray.pcoa, color="CT" ,shape = "sample") + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 过滤掉没有分型的数据
data<-subset(data,!(CT=="NA"))

## 用样本类型和分型结果重命名

data<-tidyr::unite(data, "sample_CT", sample, CT,remove=FALSE)

data$sample_CT<-as.factor(data$sample_CT)
## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-38.9
pc2<-7.5


#用于填充样本点的颜色
cbbPalette <- c("#F86C77", "#56B4E9", "#C77CFF", "#F0E442","#FF9900","#6600CC")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21,21,21,21,21)

#匹配横纵坐标
P_BOS<-ggplot(data, aes(PC1, PC2)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=sample_CT,shape=sample_CT,fill=sample_CT),size=4)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"))+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+
  stat_ellipse(aes(fill = sample_CT),geom = "polygon",level = 0.95,alpha = 0.3)

data$sample_CT<-factor(data$sample_CT,levels=c("Oroph_type2","Oroph_type1","Nasal_type1","Nasal_type2"))


ggscatter(data, x = "PC1", y = "PC2",
          color = "sample_CT", 
          palette = c("#E64B35B2","#4DBBD5B2","#00A087B2","#3C5488B2" ),
          ellipse = FALSE, mean.point = TRUE,
          star.plot = TRUE)
```




# BAL 亚型事后分析
```{r}
# 绘制BAL的两种亚型PCOA图


LungCancer_Normalized_BAL <- subset_samples(LungCancer_Normalized, sample =="BAL")

set.seed(2048)
#BAL
idx<-c("g_Prevotella","g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rotiha","g_Allprevotella","g_Leptotrichia")
idx<-c("g_Leptotrichia")
idx<-c()
LungCancer_Normalized_BAL_filter<-subset_taxa(LungCancer_Normalized_BAL,!(Genus%in%idx))

## 计算Bray_curtis距离矩阵



dm<- as.matrix(phyloseq::distance(LungCancer_Normalized_BAL_filter, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")

dm<-subset(dm,(CT.x=="type1")&(CT.y=="type2")|(CT.x=="type2")&(CT.y=="type1"))
dm$Genus<-"Both.1"

dm.all<-rbind.data.frame(dm.all,dm)


ggplot(dm.all,aes(x=Genus,y=value,fill=Genus))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=Genus),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()
```




# 口肺交流与鼻咽趋同
```{r}

idx<-c("g_Prevotella","g_Prevotella 7","g_Prevotella 6","g_Prevotella 2","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rotiha","g_Allprevotella","g_Leptotrichia","g_Butyrivibrio 2","g_Oribacterium","g_Treponema 2")

idx<-c("g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rotiha","g_Allprevotella","g_Leptotrichia")
# 计算口肺配对距离

# Distance matrix for samples
test<-subset_taxa(LungCancer_Normalized,(Genus%in%idx))

# Calculate sample similarities
dm<- as.matrix(phyloseq::distance(test, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")

A<-"Saliva"
B<-"BAL"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_BS<-rbind(Health.ps,Cancer.ps)

ps_BS<-subset(ps_BS,!(CT.x=="NA"))
ggplot(ps_BS,aes(x=CT.x,y=value,fill=CT.x))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT.x),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()





# 计算鼻咽配对距离

# Distance matrix for samples

idx<-c("g_Prevotella 2","g_Veillonella","g_Porphyromonas","g_Corynebacterium 1","g_SR1 bacterium oral taxon 875","g_Cutibacterium")
# Calculate sample similarities
test<-subset_taxa(LungCancer_Normalized,(Genus%in%idx))
dm<- as.matrix(phyloseq::distance(test, "bray"))
dm<-melt(dm)

dm<-dm[!duplicated(dm[,3]),]
dm<-dm[-1,]
dm<-merge(dm,MetaTable,by.x = "Var1",by.y = "ID")
dm<-merge(dm,MetaTable,by.x = "Var2",by.y = "ID")

A<-"Nasal"
B<-"Oroph"

Health.pair<-subset(dm,group.x=="Health"&group.y=="Health")
Health.ps1<-subset(Health.pair,sample.x==A&sample.y==B)
Health.ps2<-subset(Health.pair,sample.x==B&sample.y==A)
Health.ps<-rbind(Health.ps1,Health.ps2)
Health.ps<-subset(Health.ps,index.x==index.y)


Cancer.pair<-subset(dm,group.x=="Cancer"&group.y=="Cancer")
Cancer.ps1<-subset(Cancer.pair,sample.x==A&sample.y==B)
Cancer.ps2<-subset(Cancer.pair,sample.x==B&sample.y==A)
Cancer.ps<-rbind(Cancer.ps1,Cancer.ps2)
Cancer.ps<-subset(Cancer.ps,index.x==index.y)


ps_ON<-rbind(Health.ps,Cancer.ps)


pair_BS<-ps_BS[,c(3:8)]
pair_BS$CT<-ps_BS$CT.x
pair_ON<-ps_ON[,c(3:8)]
pair_ON$CT<-ps_ON$CT.x

pair<-merge(pair_BS,pair_ON,by="index.x")

pair<-subset(pair,value.y<0.92)

ggplot(pair, aes(x=value.x, y= value.y,color=CT.x,group=CT.x)) +
    geom_point(aes(fill=CT.x),shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1,aes(fill=CT.x)) +
theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")


ggplot(pair, aes(x=value.x, y= value.y)) +
    geom_point(shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")
```

#  采用共有ASV的方法来证明口肺交流与鼻咽趋同是相关的
```{r}
LungCancer_Normalized<-subset_samples(LungCancer_TURE, type%in%c("腺癌","Health"))

#过滤掉使用过抗生素的人群
LungCancer_Normalized<-subset_samples(LungCancer_Normalized,抗生素使用=="N")
name<-as.character(pair$index.x)

for (i in 1:64){
idx<-name[i]
data_BAL<-subset_samples(LungCancer_Normalized,index==idx&sample=="BAL")
data_Saliva<-subset_samples(LungCancer_Normalized,index==idx&sample=="Saliva")


data_BAL_tax<-taxa_names(prune_taxa(taxa_sums(data_BAL) >1, data_BAL))
data_Saliva_tax<-taxa_names(prune_taxa(taxa_sums(data_Saliva) >1, data_Saliva))

Shared_ASV[i]<-length(intersect(data_BAL_tax,data_Saliva_tax))

}
pair$ASV_BS<-Shared_ASV

for (i in 1:64){
idx<-name[i]
data_Nasal<-subset_samples(LungCancer_Normalized,index==idx&sample=="Nasal")
data_Oroph<-subset_samples(LungCancer_Normalized,index==idx&sample=="Oroph")


data_Nasal_tax<-taxa_names(prune_taxa(taxa_sums(data_Nasal) > 1, data_Nasal))
data_Oroph_tax<-taxa_names(prune_taxa(taxa_sums(data_Oroph) > 1, data_Oroph))

Shared_ASV[i]<-length(intersect(data_Nasal_tax,data_Oroph_tax))

}
pair$ASV_NO<-Shared_ASV

ggplot(pair, aes(x=ASV_NO, y= ASV_BS,color=CT.x,group=CT.x)) +
    geom_point(aes(fill=CT.x),shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1,aes(fill=CT.x)) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")






ggplot(pair, aes(x=ASV_NO, y= ASV_BS)) +
    geom_point(aes(fill=CT.x),shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "pearson")
```






# 计算Nasal Saliva Oroph BAL的CT
```{r}

count<-as.matrix(OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Nasal)])
count<-t(count[which(rowSums(count) >100),])

set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)
best 


  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  #d1 <-d1[20:47,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
 # d2 <-d2[20:47,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))






Nasal_CT<-data.frame(ass)
Nasal_CT$ID<-rownames(Nasal_CT)

colnames(Nasal_CT)<-c("Nasal_CT","ID")

Nasal_CT<-merge(metadata_Nasal,Nasal_CT,by="ID")
fisher.test(table(Nasal_CT$CT,Nasal_CT$Nasal_CT))
```



```{r}
count<-as.matrix(OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Oroph)])
count<-t(count[which(rowSums(count) >100),])

set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)
best 


  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  #d1 <-d1[20:47,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
 # d2 <-d2[20:47,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))






Oroph_CT<-data.frame(ass)
Oroph_CT$ID<-rownames(Oroph_CT)

colnames(Oroph_CT)<-c("Oroph_CT","ID")

Oroph_CT<-merge(metadata_Oroph,Oroph_CT,by="ID")
fisher.test(table(Oroph_CT$CT,Oroph_CT$Oroph_CT))

pair_ON$index.x

Oroph_CT_pair<-subset(Oroph_CT,index%in%pair_ON$index.x)
Nasal_CT_pair<-subset(Nasal_CT,index%in%pair_ON$index.x)

```






```{r}
count<-as.matrix(OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Saliva)])
count<-t(count[which(rowSums(count) >90),])

set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)
best 


  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  #d1 <-d1[20:47,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
 # d2 <-d2[20:47,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))






Saliva_CT<-data.frame(ass)
Saliva_CT$ID<-rownames(Saliva_CT)

colnames(Saliva_CT)<-c("Saliva_CT","ID")


fisher.test(table(Saliva_CT$CT,Saliva_CT$Saliva_CT))

pair_ON$index.x

Saliva_CT_pair<-subset(Saliva_CT,index%in%pair_ON$index.x)
Nasal_CT_pair<-subset(Nasal_CT,index%in%pair_ON$index.x)









```



# 计算Sparcc相关性网络

```{r}
# 挑选BAL的样本
OTU_TAX_Genus_BAL<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_BAL)]
# 过滤低丰度细菌
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL[which(rowSums(OTU_TAX_Genus_BAL) >1000),]

idx<-rownames(OTU_TAX_Genus_BAL)
BAL_sparcc_table_type1<-OTU_TAX_Genus_BAL[,colnames(OTU_TAX_Genus_BAL)%in%sample_names(LungCancer_Normalized_BAL_type1)]
BAL_sparcc_table_type2<-OTU_TAX_Genus_BAL[,colnames(OTU_TAX_Genus_BAL)%in%sample_names(LungCancer_Normalized_BAL_type2)]


write.table(BAL_sparcc_table_type1,"sparcc/BAL_sparcc_table_type1.txt",sep = "\t")
write.table(BAL_sparcc_table_type2,"sparcc/BAL_sparcc_table_type2.txt",sep = "\t")



cor_sparcc <- read.table('sparcc/results/BAL_sparcc_table_type1.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/BAL_type1_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj <- as.matrix(cor_sparcc) * as.matrix(pvals)
diag(adj) <- 0
adj<-adj[which(!(colSums(adj)==0)),which(!(colSums(adj)==0))]
adj<-data.frame(adj)

colnames(adj)<-row.names(adj)


g <- graph_from_adjacency_matrix(as.matrix(adj), mode = 'undirected', weighted = TRUE, diag = FALSE)
igraph <-  graph_from_adjacency_matrix(as.matrix(adj),
                                       mode="undirected",
                                       weighted=TRUE,
                                       diag=FALSE)

# 为igraph对象添加注释信息


name<-data.frame(V(igraph)$name)
colnames(name)<-"Genus"
name<-merge(name,TaxTable,by="Genus",all.x=TRUE,no.dups = FALSE)
name=name[!duplicated(name),]

V(igraph)$Phylum<-as.character(name$Phylum)
igraph.bal.type1<-igraph
createNetworkFromIgraph(igraph,"BAL_type1_new")
# 添加丰度信息

OTU_TAX_Genus_RA_BAL_type1<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL_type1)]
data<-OTU_TAX_Genus_RA_BAL_type1
data$mean<-rowSums(data)/66
data<-data[rownames(data)%in%row.names(adj),]

# 添加正负相关性

abx<-data.frame(E(igraph)$weight)

abx$abx<-ifelse(abx$E.igraph..weight>0,"positive","negative")
E(igraph)$abx<-abx$abx



# 读取节点信息
V(igraph)$RA<-round(data$mean,3)

## igraph的统计分析

#统计分析
###modularity
fc<-cluster_fast_greedy(igraph,weights = NULL)
modularity<-modularity(igraph,membership(fc))
modularity #高的模块性表示社区内部链接密度高，之间稀疏，一般在0.3-0.7之间

###network property
# 节点数量 Order (number of vertices ) of a graph
num.vertices <- length(V(igraph))
num.vertices

# 边数量 The size of the graph (number of edges)
num.edges <- length(E(igraph))
num.edges 
# 连接数 connectance
connerctance <-edge_density(igraph,loops = FALSE)
connerctance #网络中物种之间实际发生的相互作用数之和（；连接数之和）占潜在的相互作用连接数的比例，可以反应网络的复杂程度

# 平均度 Average degree
average.degree <- mean(igraph::degree(igraph))
average.degree

# 平均路径长度 average.path.length
average.path.length<-average.path.length(igraph)
average.path.length #也称平均最短路径长度
#平均路径长度也称为特征路径长度或平均最短路径长度，指的是一个网络中两点之间最短路径长度（或称距离）的平均值。从一个节点出发，经过与它相连的节点，逐步“走”到另一个节点所经过的路途，称为两点间的路径。其中最短的路径也称为两点间的距离，
# 直径 Diameter
E(igraph)$weight = abs(E(igraph)$weight)
diameter<-diameter(igraph,directed = FALSE,unconnected = TRUE,weights = NULL)
diameter


#连通度 connectivity
#连通度越好，网络越稳定
edge.connectivity<-edge_connectivity(igraph)
edge.connectivity #边连通度
vertex.connectivity<-vertex_connectivity(igraph)
vertex.connectivity #点连通度

# 聚集系数 Clustering coefficient
# 整个网络的全局聚焦系数C表征了整个网络的平均的"成簇性质"
# 分为局部聚焦系数和全局聚焦系数，是反应网络节点中节点的精密关系的参数，也称为传递性
# NaN ,局部聚焦系数为1，及与该节点链接的所有节点均互相链接

clustering.coefficient <- transitivity(igraph,type = "global") #type = "local"
clustering.coefficient
no.clusters<-no.clusters(igraph)
no.clusters

# 度中心性 degree centralization
centralization.degree<-centr_degree(igraph)
centralization.degree

# 接近中心性 closeness centralization
centralization.closeness<-centr_clo(igraph)
centralization.closeness
#美格节点到其他节点的最短路径的平均长度，一个节点的接近中心性越高，到其他节点的网络就越近

# 中介中心性
centralization.betweenness<-centr_betw(igraph)
centralization.betweenness

igraph.BAL.type1<-name
igraph.BAL.type1$centralization.degree<-centralization.degree$res
igraph.BAL.type1$centralization.closeness<-centralization.closeness$res
igraph.BAL.type1$centralization.betweenness<-centralization.betweenness$res

igraph.BAL.type1.info<-data.frame(modularity= numeric(0),num.vertices=numeric(0),num.edges =numeric(0),connerctance=numeric(0),average.degree=numeric(0),average.path.length=numeric(0))
igraph.BAL.type1.info[1,1]<-modularity
igraph.BAL.type1.info[1,2]<-num.vertices
igraph.BAL.type1.info[1,3]<-num.edges 
igraph.BAL.type1.info[1,4]<-connerctance
igraph.BAL.type1.info[1,5]<-average.degree
igraph.BAL.type1.info[1,6]<-average.path.length
row.names(igraph.BAL.type1.info)<-"BAL.type1"


# 验证网络的鲁棒性











write.graph(g, 'Sparcc//type1_network.graphml', format = 'graphml')
write.graph(g, 'Sparcc//type1_network.gml', format = 'gml')




cor_sparcc <- read.table('sparcc/results/BAL_sparcc_table_type2.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/BAL_type2_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj <- as.matrix(cor_sparcc) * as.matrix(pvals)
diag(adj) <- 0
adj<-adj[which(!(colSums(adj)==0)),which(!(colSums(adj)==0))]
adj<-data.frame(adj)
colnames(adj)<-row.names(adj)




g <- graph_from_adjacency_matrix(as.matrix(adj), mode = 'undirected', weighted = TRUE, diag = FALSE)
igraph <-  graph_from_adjacency_matrix(as.matrix(adj),
                                       mode="undirected",
                                       weighted=TRUE,
                                       diag=FALSE)

name<-data.frame(V(igraph)$name)
colnames(name)<-"Genus"
name<-merge(name,TaxTable,by="Genus",all.x=TRUE,no.dups = FALSE)
name=name[!duplicated(name),]

V(igraph)$Phylum<-as.character(name$Phylum)
igraph.bal.type2<-igraph

OTU_TAX_Genus_RA_BAL_type2<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL_type2)]
data<-OTU_TAX_Genus_RA_BAL_type2
data$mean<-rowSums(data)/33
data<-data[rownames(data)%in%row.names(adj),]

# 添加正负相关性

abx<-data.frame(E(igraph)$weight)

abx$abx<-ifelse(abx$E.igraph..weight>0,"positive","negative")
E(igraph)$abx<-abx$abx



# 读取节点信息
V(igraph)$RA<-round(data$mean,3)
createNetworkFromIgraph(igraph,"BAL_type2_new")

igraph.BAL.type2<-name
igraph.BAL.type2$centralization.degree<-centralization.degree$res
igraph.BAL.type2$centralization.closeness<-centralization.closeness$res
igraph.BAL.type2$centralization.betweenness<-centralization.betweenness$res

igraph.BAL.type2.info<-data.frame(modularity= numeric(0),num.vertices=numeric(0),num.edges =numeric(0),connerctance=numeric(0),average.degree=numeric(0),average.path.length=numeric(0))
igraph.BAL.type2.info[1,1]<-modularity
igraph.BAL.type2.info[1,2]<-num.vertices
igraph.BAL.type2.info[1,3]<-num.edges 
igraph.BAL.type2.info[1,4]<-connerctance
igraph.BAL.type2.info[1,5]<-average.degree
igraph.BAL.type2.info[1,6]<-average.path.length
row.names(igraph.BAL.type2.info)<-"BAL.type2"

createNetworkFromIgraph(igraph,"BAL_type2_new")



robu.betweenness.bal.type1<-robustness(igraph.bal.type1, type = "vertex", measure = "btwn.cent", N = 1000)
robu.degree.bal.type1<-robustness(igraph.bal.type1, type = "vertex", measure = "degree", N = 1000)
robu.betweenness.bal.type2<-robustness(igraph.bal.type2, type = "vertex", measure = "btwn.cent", N = 1000)
robu.degree.bal.type2<-robustness(igraph.bal.type2, type = "vertex", measure = "degree", N = 1000)


robu.betweenness.bal.type1<-robustness(igraph.bal.type1, type = "edge", measure = "random")

robu.betweenness.bal.type2<-robustness(igraph.bal.type2, type = "edge", measure = "random")


robu.betweenness.bal.type1$group<-"type1"
robu.betweenness.bal.type2$group<-"type2"
robu.degree.bal.type1$group<-"type1"
robu.degree.bal.type2$group<-"type2"





robu.betweenness.bal<-rbind.data.frame(robu.betweenness.bal.type1,robu.betweenness.bal.type2)
robu.betweenness.bal$sample<-"BAL"

robu.degree.bal<-rbind.data.frame(robu.degree.bal.type1,robu.degree.bal.type2)
robu.degree.bal$sample<-"BAL"
  ggplot(robu.degree.bal,aes(y=comp.pct,x=removed.pct,group=group,color=group))+geom_line()+theme_bw()


# 对比两个网络的特征


igraph.BAL.info<-rbind.data.frame(igraph.BAL.type1.info,igraph.BAL.type2.info)
igraph.BAL.info$group<-rownames(igraph.BAL.info)

ggplot(data= igraph.BAL.info, mapping=aes(x=group,y=num.edges))+
    geom_bar(aes(fill=group),stat="identity",position="dodge")+scale_fill_manual(name = "Group", values = c(`freq` = "blue", `freq.pred`="grey50" ))+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 1,hjust = 1))

```




# 为网络添加注释信息
```{r}
cor_tax<-as.data.frame(rownames(cor_sparcc))
colnames(cor_tax)<-"Genus"

cor_tax<-merge(cor_tax,TAX,by="Genus")
cor_tax<-cor_tax[!duplicated(cor_tax[,1]),]
type1_top10<-Rich_type1_top10[,c("V1","V3")]
type2_top10<-Rich_type2_top10[,c("V1","V3")]

type1_top10<-Rich_type1[,c("V1","V3")]
type2_top10<-Rich_type2[,c("V1","V3")]

top10<-rbind.data.frame(type1_top10,type2_top10)
cor_tax<-merge(cor_tax,top10,by.x="Genus",by.y="V1",all=TRUE)
cor_tax[is.na(cor_tax)]<-0
#替换NA值
cor_tax[is.na(cor_tax)] <- "type3"


dot<-read.table("Sparcc/type1.csv",sep = ",",header = TRUE)
dot<-merge(dot,cor_tax,by.x = "v_name",by.y = "Genus")
dot$Genus<-dot$v_name
dot<-dot[,-1]
write.csv(dot,"Sparcc/type1.csv",sep = ",",row.names = TRUE)
```








# 
```{r}
# 挑选Oroph的样本
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Oroph)]
# 过滤低丰度细菌
OTU_TAX_Genus_Oroph<-OTU_TAX_Genus_Oroph[rownames(OTU_TAX_Genus_Oroph)%in%idx,]


Oroph_sparcc_table_type1<-OTU_TAX_Genus_Oroph[,colnames(OTU_TAX_Genus_Oroph)%in%sample_names(LungCancer_Normalized_Oroph_type1)]
Oroph_sparcc_table_type2<-OTU_TAX_Genus_Oroph[,colnames(OTU_TAX_Genus_Oroph)%in%sample_names(LungCancer_Normalized_Oroph_type2)]




write.table(Oroph_sparcc_table_type1,"sparcc/Oroph_sparcc_table_type1.txt",sep = "\t")
write.table(Oroph_sparcc_table_type2,"sparcc/Oroph_sparcc_table_type2.txt",sep = "\t")




# 挑选Saliva的样本
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Saliva)]
# 过滤低丰度细菌
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus_Saliva[rownames(OTU_TAX_Genus_Saliva)%in%idx,]


Saliva_sparcc_table_type1<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type1)]
Saliva_sparcc_table_type2<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type2)]




write.table(Saliva_sparcc_table_type1,"sparcc/Saliva_sparcc_table_type1.txt",sep = "\t")
write.table(Saliva_sparcc_table_type2,"sparcc/Saliva_sparcc_table_type2.txt",sep = "\t")




# 挑选Nasal的样本
OTU_TAX_Genus_Nasal<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Nasal)]
# 过滤低丰度细菌
OTU_TAX_Genus_Nasal<-OTU_TAX_Genus_Nasal[which(rowSums(OTU_TAX_Genus_Nasal) >860),]


Nasal_sparcc_table_type1<-OTU_TAX_Genus_Nasal[,colnames(OTU_TAX_Genus_Nasal)%in%sample_names(LungCancer_Normalized_Nasal_type1)]
Nasal_sparcc_table_type2<-OTU_TAX_Genus_Nasal[,colnames(OTU_TAX_Genus_Nasal)%in%sample_names(LungCancer_Normalized_Nasal_type2)]




write.table(Nasal_sparcc_table_type1,"sparcc/Nasal_sparcc_table_type1.txt",sep = "\t")
write.table(Nasal_sparcc_table_type2,"sparcc/Nasal_sparcc_table_type2.txt",sep = "\t")



# 挑选Fecal的样本
LungCancer_Normalized_Fecal_type1<-subset_samples(LungCancer_Normalized,sample=="Fecal"&CT=="type1")
LungCancer_Normalized_Fecal_type2<-subset_samples(LungCancer_Normalized,sample=="Fecal"&CT=="type2")
OTU_TAX_Genus_Fecal<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Fecal)]
# 过滤低丰度细菌
OTU_TAX_Genus_Fecal<-OTU_TAX_Genus_Fecal[which(rowSums(OTU_TAX_Genus_Fecal) >870),]


Fecal_sparcc_table_type1<-OTU_TAX_Genus_Fecal[,colnames(OTU_TAX_Genus_Fecal)%in%sample_names(LungCancer_Normalized_Fecal_type1)]
Fecal_sparcc_table_type2<-OTU_TAX_Genus_Fecal[,colnames(OTU_TAX_Genus_Fecal)%in%sample_names(LungCancer_Normalized_Fecal_type2)]


write.table(Fecal_sparcc_table_type1,"sparcc/Fecal_sparcc_table_type1.txt",sep = "\t")
write.table(Fecal_sparcc_table_type2,"sparcc/Fecal_sparcc_table_type2.txt",sep = "\t")













# 挑选Saliva的样本
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Saliva)]
# 过滤低丰度细菌
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus_Saliva[which(rowSums(OTU_TAX_Genus_Saliva) >1000),]

idx<-rownames(OTU_TAX_Genus_Saliva)
Saliva_sparcc_table_type1<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type1)]
Saliva_sparcc_table_type2<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type2)]


write.table(Saliva_sparcc_table_type1,"sparcc/Saliva_sparcc_table_type1.txt",sep = "\t")
write.table(Saliva_sparcc_table_type2,"sparcc/Saliva_sparcc_table_type2.txt",sep = "\t")



cor_sparcc <- read.table('sparcc/results/Saliva_sparcc_table_type1.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/Saliva_type1_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj <- as.matrix(cor_sparcc) * as.matrix(pvals)
diag(adj) <- 0
adj<-adj[which(!(colSums(adj)==0)),which(!(colSums(adj)==0))]
adj<-data.frame(adj)
colnames(adj)<-row.names(adj)


g <- graph_from_adjacency_matrix(as.matrix(adj), mode = 'undirected', weighted = TRUE, diag = FALSE)
igraph <-  graph_from_adjacency_matrix(as.matrix(adj),
                                       mode="undirected",
                                       weighted=TRUE,
                                       diag=FALSE)

name<-data.frame(V(igraph)$name)
colnames(name)<-"Genus"
name<-merge(name,TaxTable,by="Genus",all.x=TRUE,no.dups = FALSE)
name=name[!duplicated(name),]

V(igraph)$Phylum<-as.character(name$Phylum)
igraph.saliva.type1<-igraph



# 添加丰度信息

OTU_TAX_Genus_RA_Saliva_type1<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Saliva_type1)]
data<-OTU_TAX_Genus_RA_Saliva_type1
data$mean<-rowSums(data)/52
data<-data[rownames(data)%in%row.names(adj),]



abx<-data.frame(E(igraph)$weight)

abx$abx<-ifelse(abx$E.igraph..weight>0,"positive","negative")
E(igraph)$abx<-abx$abx

V(igraph)$RA<-round(data$mean,3)

createNetworkFromIgraph(igraph,"Saliva_type1_new")

fc<-cluster_fast_greedy(igraph,weights = NULL)
modularity<-modularity(igraph,membership(fc))
modularity #高的模块性表示社区内部链接密度高，之间稀疏，一般在0.3-0.7之间

###network property
# 节点数量 Order (number of vertices ) of a graph
num.vertices <- length(V(igraph))
num.vertices

# 边数量 The size of the graph (number of edges)
num.edges <- length(E(igraph))
num.edges 
# 连接数 connectance
connerctance <-edge_density(igraph,loops = FALSE)
connerctance #网络中物种之间实际发生的相互作用数之和（；连接数之和）占潜在的相互作用连接数的比例，可以反应网络的复杂程度

# 平均度 Average degree
average.degree <- mean(igraph::degree(igraph))
average.degree

# 平均路径长度 average.path.length
average.path.length<-average.path.length(igraph)
average.path.length #也称平均最短路径长度
#平均路径长度也称为特征路径长度或平均最短路径长度，指的是一个网络中两点之间最短路径长度（或称距离）的平均值。从一个节点出发，经过与它相连的节点，逐步“走”到另一个节点所经过的路途，称为两点间的路径。其中最短的路径也称为两点间的距离，
# 直径 Diameter
E(igraph)$weight = abs(E(igraph)$weight)
diameter<-diameter(igraph,directed = FALSE,unconnected = TRUE,weights = NULL)
diameter


#连通度 connectivity
#连通度越好，网络越稳定
edge.connectivity<-edge_connectivity(igraph)
edge.connectivity #边连通度
vertex.connectivity<-vertex_connectivity(igraph)
vertex.connectivity #点连通度

# 聚集系数 Clustering coefficient
# 整个网络的全局聚焦系数C表征了整个网络的平均的"成簇性质"
# 分为局部聚焦系数和全局聚焦系数，是反应网络节点中节点的精密关系的参数，也称为传递性
# NaN ,局部聚焦系数为1，及与该节点链接的所有节点均互相链接

clustering.coefficient <- transitivity(igraph,type = "gloSaliva") #type = "local"
clustering.coefficient
no.clusters<-no.clusters(igraph)
no.clusters

# 度中心性 degree centralization
centralization.degree<-centr_degree(igraph)
centralization.degree

# 接近中心性 closeness centralization
centralization.closeness<-centr_clo(igraph)
centralization.closeness
#美格节点到其他节点的最短路径的平均长度，一个节点的接近中心性越高，到其他节点的网络就越近

# 中介中心性
centralization.betweenness<-centr_betw(igraph)
centralization.betweenness

igraph.Saliva.type1<-name
igraph.Saliva.type1$centralization.degree<-centralization.degree$res
igraph.Saliva.type1$centralization.closeness<-centralization.closeness$res
igraph.Saliva.type1$centralization.betweenness<-centralization.betweenness$res

igraph.Saliva.type1.info<-data.frame(modularity= numeric(0),num.vertices=numeric(0),num.edges =numeric(0),connerctance=numeric(0),average.degree=numeric(0),average.path.length=numeric(0))
igraph.Saliva.type1.info[1,1]<-modularity
igraph.Saliva.type1.info[1,2]<-num.vertices
igraph.Saliva.type1.info[1,3]<-num.edges 
igraph.Saliva.type1.info[1,4]<-connerctance
igraph.Saliva.type1.info[1,5]<-average.degree
igraph.Saliva.type1.info[1,6]<-average.path.length
row.names(igraph.Saliva.type1.info)<-"Saliva.type1"






createNetworkFromIgraph(igraph,"Saliva_type1_new")








# 挑选Saliva的样本
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_Saliva)]
# 过滤低丰度细菌
OTU_TAX_Genus_Saliva<-OTU_TAX_Genus_Saliva[which(rowSums(OTU_TAX_Genus_Saliva) >1000),]

idx<-rownames(OTU_TAX_Genus_Saliva)
Saliva_sparcc_table_type2<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type2)]
Saliva_sparcc_table_type2<-OTU_TAX_Genus_Saliva[,colnames(OTU_TAX_Genus_Saliva)%in%sample_names(LungCancer_Normalized_Saliva_type2)]


write.table(Saliva_sparcc_table_type2,"sparcc/Saliva_sparcc_table_type2.txt",sep = "\t")
write.table(Saliva_sparcc_table_type2,"sparcc/Saliva_sparcc_table_type2.txt",sep = "\t")



cor_sparcc <- read.table('sparcc/results/Saliva_sparcc_table_type2.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/Saliva_type2_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj <- as.matrix(cor_sparcc) * as.matrix(pvals)
diag(adj) <- 0
adj<-adj[which(!(colSums(adj)==0)),which(!(colSums(adj)==0))]


adj<-data.frame(adj)
colnames(adj)<-row.names(adj)



g <- graph_from_adjacency_matrix(as.matrix(adj), mode = 'undirected', weighted = TRUE, diag = FALSE)
igraph <-  graph_from_adjacency_matrix(as.matrix(adj),
                                       mode="undirected",
                                       weighted=TRUE,
                                       diag=FALSE)

name<-data.frame(V(igraph)$name)
colnames(name)<-"Genus"
name<-merge(name,TaxTable,by="Genus",all.x=TRUE,no.dups = FALSE)
name=name[!duplicated(name),]

V(igraph)$Phylum<-as.character(name$Phylum)

igraph.saliva.type2<-igraph





OTU_TAX_Genus_RA_Saliva_type2<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Saliva_type2)]
data<-OTU_TAX_Genus_RA_Saliva_type2
data$mean<-rowSums(data)/27
data<-data[rownames(data)%in%row.names(adj),]



abx<-data.frame(E(igraph)$weight)

abx$abx<-ifelse(abx$E.igraph..weight>0,"positive","negative")
E(igraph)$abx<-abx$abx

V(igraph)$RA<-round(data$mean,3)

createNetworkFromIgraph(igraph,"Saliva_type2_new")







fc<-cluster_fast_greedy(igraph,weights = NULL)
modularity<-modularity(igraph,membership(fc))
modularity #高的模块性表示社区内部链接密度高，之间稀疏，一般在0.3-0.7之间

###network property
# 节点数量 Order (number of vertices ) of a graph
num.vertices <- length(V(igraph))
num.vertices

# 边数量 The size of the graph (number of edges)
num.edges <- length(E(igraph))
num.edges 
# 连接数 connectance
connerctance <-edge_density(igraph,loops = FALSE)
connerctance #网络中物种之间实际发生的相互作用数之和（；连接数之和）占潜在的相互作用连接数的比例，可以反应网络的复杂程度

# 平均度 Average degree
average.degree <- mean(igraph::degree(igraph))
average.degree

# 平均路径长度 average.path.length
average.path.length<-average.path.length(igraph)
average.path.length #也称平均最短路径长度
#平均路径长度也称为特征路径长度或平均最短路径长度，指的是一个网络中两点之间最短路径长度（或称距离）的平均值。从一个节点出发，经过与它相连的节点，逐步“走”到另一个节点所经过的路途，称为两点间的路径。其中最短的路径也称为两点间的距离，
# 直径 Diameter
E(igraph)$weight = abs(E(igraph)$weight)
diameter<-diameter(igraph,directed = FALSE,unconnected = TRUE,weights = NULL)
diameter


#连通度 connectivity
#连通度越好，网络越稳定
edge.connectivity<-edge_connectivity(igraph)
edge.connectivity #边连通度
vertex.connectivity<-vertex_connectivity(igraph)
vertex.connectivity #点连通度

# 聚集系数 Clustering coefficient
# 整个网络的全局聚焦系数C表征了整个网络的平均的"成簇性质"
# 分为局部聚焦系数和全局聚焦系数，是反应网络节点中节点的精密关系的参数，也称为传递性
# NaN ,局部聚焦系数为1，及与该节点链接的所有节点均互相链接

clustering.coefficient <- transitivity(igraph,type = "gloSaliva") #type = "local"
clustering.coefficient
no.clusters<-no.clusters(igraph)
no.clusters

# 度中心性 degree centralization
centralization.degree<-centr_degree(igraph)
centralization.degree

# 接近中心性 closeness centralization
centralization.closeness<-centr_clo(igraph)
centralization.closeness
#美格节点到其他节点的最短路径的平均长度，一个节点的接近中心性越高，到其他节点的网络就越近

# 中介中心性
centralization.betweenness<-centr_betw(igraph)
centralization.betweenness

igraph.Saliva.type2<-name
igraph.Saliva.type2$centralization.degree<-centralization.degree$res
igraph.Saliva.type2$centralization.closeness<-centralization.closeness$res
igraph.Saliva.type2$centralization.betweenness<-centralization.betweenness$res

igraph.Saliva.type2.info<-data.frame(modularity= numeric(0),num.vertices=numeric(0),num.edges =numeric(0),connerctance=numeric(0),average.degree=numeric(0),average.path.length=numeric(0))
igraph.Saliva.type2.info[1,1]<-modularity
igraph.Saliva.type2.info[1,2]<-num.vertices
igraph.Saliva.type2.info[1,3]<-num.edges 
igraph.Saliva.type2.info[1,4]<-connerctance
igraph.Saliva.type2.info[1,5]<-average.degree
igraph.Saliva.type2.info[1,6]<-average.path.length
row.names(igraph.Saliva.type2.info)<-"Saliva.type2"




createNetworkFromIgraph(igraph,"Saliva_type2_new")






igraph.Saliva.info<-rbind.data.frame(igraph.Saliva.type1.info,igraph.Saliva.type2.info)
igraph.Saliva.info$group<-rownames(igraph.Saliva.info)

ggplot(data= igraph.Saliva.info, mapping=aes(x=group,y=num.edges))+
    geom_bar(aes(fill=group),stat="identity",position="dodge")+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 1,hjust = 1))


#####


robu.betweenness.saliva.type1<-robustness(igraph.saliva.type1, type = "vertex", measure = "btwn.cent", N = 1000)
robu.degree.saliva.type1<-robustness(igraph.saliva.type1, type = "vertex", measure = "degree", N = 1000)
robu.betweenness.saliva.type2<-robustness(igraph.saliva.type2, type = "vertex", measure = "btwn.cent", N = 1000)
robu.degree.saliva.type2<-robustness(igraph.saliva.type2, type = "vertex", measure = "degree", N = 1000)


robu.betweenness.saliva.type1<-robustness(igraph.saliva.type1, type = "edge", measure = "random")

robu.betweenness.saliva.type2<-robustness(igraph.saliva.type2, type = "edge", measure = "random")


robu.betweenness.saliva.type1$group<-"type1"
robu.betweenness.saliva.type2$group<-"type2"
robu.degree.saliva.type1$group<-"type1"
robu.degree.saliva.type2$group<-"type2"





robu.betweenness.saliva<-rbind.data.frame(robu.betweenness.saliva.type1,robu.betweenness.saliva.type2)
robu.betweenness.saliva$sample<-"Saliva"

robu.degree.saliva<-rbind.data.frame(robu.degree.saliva.type1,robu.degree.saliva.type2)
robu.degree.saliva$sample<-"Saliva"
 

robu.degree<-rbind.data.frame(robu.degree.saliva,robu.degree.bal)
robu.betweenness<-rbind.data.frame(robu.betweenness.saliva,robu.betweenness.bal)


ggplot(robu.degree,aes(y=comp.pct,x=removed.pct,color=group))+geom_line(aes(linetype=sample),size=1.2)+theme_bw()+




# 对比Prevetolla_7在网络中的作用

igraph.BAL.type1.Prevotella_7<-subset(igraph.BAL.type1,Genus=="g_Prevotella 7")
rownames(igraph.BAL.type1.Prevotella_7)<-"BAL.type1"
igraph.BAL.type2.Prevotella_7<-subset(igraph.BAL.type2,Genus=="g_Prevotella 7")
rownames(igraph.BAL.type2.Prevotella_7)<-"BAL.type2"
igraph.BAL.Prevotella_7<-rbind.data.frame(igraph.BAL.type1.Prevotella_7,igraph.BAL.type2.Prevotella_7)

igraph.BAL.Prevotella_7$group<-rownames(igraph.BAL.Prevotella_7)



ggplot(data= igraph.BAL.Prevotella_7, mapping=aes(x=group,y=centralization.degree))+
    geom_bar(aes(fill=group),stat="identity",position="dodge")+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 1,hjust = 1))







igraph.Saliva.type1.Prevotella_7<-subset(igraph.Saliva.type1,Genus=="g_Prevotella 7")
rownames(igraph.Saliva.type1.Prevotella_7)<-"Saliva.type1"
igraph.Saliva.type2.Prevotella_7<-subset(igraph.Saliva.type2,Genus=="g_Prevotella 7")
rownames(igraph.Saliva.type2.Prevotella_7)<-"Saliva.type2"
igraph.Saliva.Prevotella_7<-rbind.data.frame(igraph.Saliva.type1.Prevotella_7,igraph.Saliva.type2.Prevotella_7)

igraph.Saliva.Prevotella_7$group<-rownames(igraph.Saliva.Prevotella_7)



ggplot(data= igraph.Saliva.Prevotella_7, mapping=aes(x=group,y=centralization.degree))+
    geom_bar(aes(fill=group),stat="identity",position="dodge")+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 1,hjust = 1))
```


```{r}
 OTU_TAX_Genus_BAL<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL)]
 data<-data.frame(t(OTU_TAX_Genus_BAL))
 data$ID<-rownames(data)
 data<-merge(data,MetaTable,by="ID")
 ggplot(data,aes(x=CT,y=g_Prevotella,fill=CT))+ #”fill=“设置填充颜色
    stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
    geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
    geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=3)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
    scale_fill_manual(values = c("#E69F00", "#56B4E9"))+  #设置填充的颜色
    scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
    ggtitle("Source Tracker")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
    ylab("Saliva")+xlab("")+ #设置x轴和y轴的标题
    stat_compare_means()
```


# 进行聚类分析
```{R}
library(ggtree)
library(ggtreeExtra)
library(ggplot2)
library(ggnewscale)
library(dplyr)
library(tidytree)
library(ggstar)
LungCancer_Normalized


set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵
dis_bray<- as.matrix(phyloseq::distance(subset_samples(LungCancer_Normalized_BS,!(CT=="NA")), "bray"))

## 聚类
hc = eclust(dis_bray,"hclust" ,k = 2)
## 绘制层次聚类图
fviz_dend(hc, rect = TRUE)

## 获得分组信息
group<-data.frame(sample_data(subset_samples(LungCancer_Normalized_BS,!(CT=="NA"))))

#groupInfo <- split(row.names(group), group$sample
# <- groupOTU(res.hc, groupInfo)

den <- as.dendrogram(hc)

clus <- cutree(hc, 2)
g <- split(names(clus), clus)


p <- ggtree(hc,layout="circular")

clades <- sapply(g, function(n) MRCA(p, n))

p <- groupClade(p, clades, group_name='subtree') + aes(color=subtree)


p1<-p+scale_colour_manual(
        values=c("black","#69B920","#9C2E88","#F74B00","#60C3DB"),
        labels=c("","I", "II"),
        guide=guide_legend(keywidth=0.5,
                           keyheight=0.5,
                           order=1,
                           override.aes=list(linetype=c("0"=NA,
                                                        "Clade1"=1,
                                                        "Clade2"=1
                                                        
                           )
                           )
        )
    ) + 
    new_scale_colour()



p2<-p1 +geom_tippoint(aes(colour=cyl), alpha=0) +geom_tiplab(aes(colour=cyl),
                                                       align=TRUE,
                                                       linetype=3,
                                                       size=2,
                                                       linesize=0.2,
                                                       show.legend=FALSE
) +scale_colour_manual(
        name="Sample",
        values=c("blue","red"),
        guide=guide_legend(keywidth=0.5,
                           keyheight=0.5,
                           order=2,
                           override.aes=list(size=2,alpha=1))
    )

d <- data.frame(label = names(clus),cyl = group$sample,CT=group$CT)

p<-p %<+%d

p + geom_tippoint(size=2.5, aes(fill=factor(cyl), x=x+.5,shape=factor(cyl)), color='black') +aes(color=CT)+scale_shape_manual(values=c(21,22))
```

# 绘制BAL和Saliva的中高丰度物种的热图
```{r}

# 获取样本ID
ID<-sample_names(subset_samples(LungCancer_Normalized_BS,!(CT=="NA")))
idx<-c("g_Prevotella 7","g_Veillonella","g_Prevotella 6","g_Butyrivibrio 2","g_Porphyromonas","g_Megasphaera")
data<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%ID]
data$Genus<-rownames(data)

data<-melt(data,id.vars = "Genus")


data$value[which(data$value==0)]<-0.001

p2+ new_scale_fill() +
    geom_fruit(data=data, geom=geom_tile,mapping=aes(y=variable, x=Genus, alpha=log10(value),fill=Genus),color = "grey50", offset = 0.1,size = 0.02)+scale_alpha_continuous(range=c(0, 1),guide=guide_legend(keywidth = 0.3, keyheight = 0.3, order=5),name="log10(Relative Abundance)")+scale_fill_manual(values=c("#0000FF","#FFA500","#FF0000","#800000","#006400","#800080","#696969"),guide=guide_legend(keywidth = 0.3, keyheight = 0.3, order=4))
```



# 双分组对角线热图
```{r}
geom_rectriangle <- function(mapping = NULL, data = NULL,
                        stat = "identity", position = "identity",
                        ...,
                        linejoin = "mitre",
                        na.rm = FALSE,
                        show.legend = NA,
                        inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomRectriangle,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      linejoin = linejoin,
      na.rm = na.rm,
      ...
    )
  )
}


GeomRectriangle <- ggproto(
    "GeomRectriangle", Geom,
    default_aes = aes(r = 1, colour = "#e0e0e0", fill = NA, size = 0.25, linetype = 1,
                      alpha = NA,type = "upper"),
    required_aes = c("x", "y"),
    draw_panel = function(self, data, panel_params, coord, linejoin = "mitre",type = "upper") {
        aesthetics <- setdiff(names(data), c("x", "y"))
        
        polys <- lapply(split(data, seq_len(nrow(data))), function(row) {
            rectriangle <- point_to_rectriangle(row$x, row$y, row$r, row$type)
            aes <- new_data_frame(row[aesthetics])[rep(1, 4), ]
            GeomPolygon$draw_panel(cbind(rectriangle, aes), panel_params, coord)
        })
        
        ggplot2:::ggname("geom_rectriangle", do.call("grobTree", polys))
    },
    draw_key = draw_key_polygon
)



point_to_rectriangle <- function(x, y, r, type = type) {
    r <- 0.5 * sign(r) * sqrt(abs(r))
    #r0 = 0.5
    xmin <- - r + x
    xmax <- r + x
    ymin <- - r + y
    ymax <- r + y 
    if(type == "upper"){
        df = new_data_frame(list(
            y = c(ymax, ymax, ymin, ymax),
            x = c(xmin, xmax, xmin, xmin)
        ))
    }else if(type == "lower"){
        df = new_data_frame(list(
            y = c(ymax, ymin, ymin, ymax),
            x = c(xmax, xmax, xmin, xmax) 
        ))
    }
    df
}


# 选择需要进行相关性的细菌
idx<-c("g_Prevotella 7","g_Veillonella","g_Prevotella 6","g_Butyrivibrio 2","g_Porphyromonas","g_Megasphaera")
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_BAL)]
OTU_TAX_Genus_BAL<-data.frame(t(OTU_TAX_Genus_BAL))
OTU_TAX_Genus_BAL$ID<-rownames(OTU_TAX_Genus_BAL)
# 选择需要进行相关性的因子
idx<-c("EGF","IL.1a","IL.1b","IP.10","RANTES","IL.1RA","PDGF.AA","TNFb","TGFa","VEGF.A","MCP.1","CXCL9","CT","IL.10","IL.27")
cytokine_data<-cytokine[,colnames(cytokine)%in%idx]
cytokine_data$ID <- row.names(cytokine_data)
# 合并细菌和细胞因子


data<-merge(cytokine_data,OTU_TAX_Genus_BAL,by="ID")
data<-merge(data,MetaTable,by="ID")
data<-data[,c(2:21,64)]
data_type1<-subset(data,CT=="type1")
data_type1<-data_type1[,-21]
data_type2<-subset(data,CT=="type2")
data_type2<-data_type2[,-21]
data_type1<-cor(data_type1,method = "spearman")

pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0

data_type1_R<-as.matrix((corr.test(data_type1,method = "spearman",adjust = "none"))$r)
data_type1_P<-as.matrix((corr.test(data_type1,method = "spearman",adjust = "none"))$p)

data_type1_P[data_type1_P>=0.05] <- -1
data_type1_P[data_type1_P<0.05 & data_type1_P>=0] <- 1
data_type1_P[data_type1_P==-1] <- 0


data_type1_cor <- as.matrix(data_type1_R) * as.matrix(data_type1_P)
data_type1_cor <- melt(data_type1_cor)


data_type2_R<-as.matrix((corr.test(data_type2,method = "spearman",adjust = "none"))$r)
data_type2_P<-as.matrix((corr.test(data_type2,method = "spearman",adjust = "none"))$p)

data_type2_P[data_type2_P>=0.05] <- -1
data_type2_P[data_type2_P<0.05 & data_type2_P>=0] <- 1
data_type2_P[data_type2_P==-1] <- 0


data_type2_cor <- as.matrix(data_type2_R) * as.matrix(data_type2_P)
data_type2_cor <- melt(data_type2_cor)


data_cor <- cbind.data.frame(data_type1_cor,data_type2_cor)
data_cor <- data_cor[,-4:-5]
colnames(data_cor) <- c("Var1","Var2","Type1","Type2")
ggplot()+
    geom_rectriangle(data = data_cor, aes(Var1, Var2, fill = Type1),type = "upper", r = 1)+
    scale_fill_gradient2(high = "#762a83",low = "#1b7837", na.value = "#e0e0e0",limits=c(-0.35,1))+
    ggnewscale::new_scale_fill()+
    geom_rectriangle(data = data_cor, aes(Var1, Var2, fill = Type2),type = "lower", r = 1)+
    scale_fill_gradient2(high = "#762a83",low = "#1b7837", na.value = "#e0e0e0",limits=c(-0.35,1))+
    labs(x = "", y = "")+
    theme_bw()+theme(axis.text.x = element_text(hjust=1, vjust=1, angle = 35,size = 12))
```






# 四部位菌群关系
```{r}
name<-as.character(pair$index.x)
count<-data.frame()
for (i in 1:64){
idx<-name[i]
data<-subset_samples(LungCancer_Normalized,index==idx&!(sample=="Fecal"))


table<-data.frame(otu_table(data))

colnames(table)<-c("BAL","Nasal","Oroph","Saliva")


table[table<2]<-0
table[table>=2]<-1
#table$BAL<-as.logical(table$BAL)
#table$Nasal<-as.logical(table$Nasal)
#table$Oroph<-as.logical(table$Oroph)
#table$Saliva<-as.logical(table$Saliva)
a<-make_comb_mat(table,mode = "intersect")
b<-data.frame(comb_size(a))

#补充四个都存在的个数

table<-subset(table,(BAL==1 )& (Nasal==1) & (Oroph==1) & (Saliva==1))
num_all<-nrow(table)
b["1111",]<-num_all
b$group<-rownames(b)
b$index<-idx
count<-rbind.data.frame(count,b)
}

count<-merge(count,BAL_Interindividual,by="index")

count<-subset(count,!(group.x=="0000"))
count.x<-count

count$group.x[which(count$group.x=="1000")]<-list(c("BAL"))
count$group.x[which(count$group.x=="0100")]<-list(c("Nasal"))
count$group.x[which(count$group.x=="0010")]<-list(c("Oroph"))
count$group.x[which(count$group.x=="0001")]<-list(c("Saliva"))
count$group.x[which(count$group.x=="1100")]<-list(c("BAL","Nasal"))
count$group.x[which(count$group.x=="1010")]<-list(c("BAL","Oroph"))
count$group.x[which(count$group.x=="1001")]<-list(c("BAL","Saliva"))
count$group.x[which(count$group.x=="1110")]<-list(c("BAL","Nasal","Oroph"))
count$group.x[which(count$group.x=="1101")]<-list(c("BAL","Nasal","Saliva"))
count$group.x[which(count$group.x=="1011")]<-list(c("BAL","Oroph","Saliva"))
count$group.x[which(count$group.x=="1111")]<-list(c("BAL","Nasal","Oroph","Saliva"))
count$group.x[which(count$group.x=="0110")]<-list(c("Nasal","Oroph"))
count$group.x[which(count$group.x=="0101")]<-list(c("Nasal","Saliva"))
count$group.x[which(count$group.x=="0111")]<-list(c("Nasal","Oroph","Saliva"))
count$group.x[which(count$group.x=="0011")]<-list(c("Oroph","Saliva"))



count.x$group.x[which(count.x$group.x=="1000")]<-(c("BAL"))
count.x$group.x[which(count.x$group.x=="0100")]<-(c("Nasal"))
count.x$group.x[which(count.x$group.x=="0010")]<-(c("Oroph"))
count.x$group.x[which(count.x$group.x=="0001")]<-(c("Saliva"))
count.x$group.x[which(count.x$group.x=="1100")]<-(c("BAL_Nasal"))
count.x$group.x[which(count.x$group.x=="1010")]<-(c("BAL_Oroph"))
count.x$group.x[which(count.x$group.x=="1001")]<-(c("BAL_Saliva"))
count.x$group.x[which(count.x$group.x=="1110")]<-(c("BAL_Nasal_Oroph"))
count.x$group.x[which(count.x$group.x=="1101")]<-(c("BAL_Nasal_Saliva"))
count.x$group.x[which(count.x$group.x=="1011")]<-(c("BAL_Oroph_Saliva"))
count.x$group.x[which(count.x$group.x=="1111")]<-(c("BAL_Nasal_Oroph_Saliva"))
count.x$group.x[which(count.x$group.x=="0110")]<-(c("Nasal_Oroph"))
count.x$group.x[which(count.x$group.x=="0101")]<-(c("Nasal_Saliva"))
count.x$group.x[which(count.x$group.x=="0111")]<-(c("Nasal_Oroph_Saliva"))
count.x$group.x[which(count.x$group.x=="0011")]<-(c("Oroph_Saliva"))







tidy_pathway_member <- table %>%
  as_tibble(rownames = "Site") %>%
  gather(ASV, Member, -Site) %>%
  filter(Member) %>%
  select(- Member)


data<-tidy_pathway_member %>%
  group_by(ASV) %>%
  tidytree::summarize(Site = list(Site))

data$group<-0
count<-t(data.frame(table((data$Site[1:nrow(data)])%in%test)))

#%>%
 # ggplot(aes(x = Site)) +
  #  geom_bar() +
   # scale_x_upset()

count$CT<-factor(count$CT)


 
p<-ggplot(data=count,aes(x=group.x, y=comb_size.a.)) +
    geom_boxplot() +
    geom_jitter(aes(color=CT), width=0.1) +
    scale_x_upset(order_by = "degree",
                  sets = c("BAL", "Nasal", "Oroph", "Saliva"),
                  position="top", name = "") +
    theme_combmatrix(combmatrix.label.text = element_text(size=12),
                     combmatrix.label.extra_spacing = 5)
count.x$group.x<-factor( count.x$group.x,levels = c("BAL","Oroph","Saliva","Nasal","Oroph_Saliva","Nasal_Oroph","BAL_Oroph","BAL_Saliva","Nasal_Saliva","","BAL_Nasal","BAL_Oroph_Saliva","Nasal_Oroph_Saliva","BAL_Nasal_Oroph","BAL_Nasal_Saliva","BAL_Nasal_Oroph_Saliva"))
count.x$group.x<-factor( count.x$group.x,levels =
c("BAL","Nasal","Oroph","Saliva","Oroph_Saliva","Nasal_Oroph","BAL_Oroph","BAL_Saliva","Nasal_Saliva","","BAL_Nasal","BAL_Oroph_Saliva","Nasal_Oroph_Saliva","BAL_Nasal_Oroph","BAL_Nasal_Saliva","BAL_Nasal_Oroph_Saliva"))


count.x$group.x<-factor( count.x$group.x,levels =c("BAL","Nasal","Oroph","Saliva","BAL_Oroph","Nasal_Oroph","Oroph_Saliva","BAL_Nasal","BAL_Saliva","Nasal_Saliva","BAL_Oroph_Saliva","Nasal_Oroph_Saliva","BAL_Nasal_Oroph","BAL_Nasal_Saliva","BAL_Nasal_Oroph_Saliva"))







p1<-ggplot(data=count.x,aes(x=group.x, y=comb_size.a.,fill=CT))+stat_boxplot(geom = "errorbar",width=0.3) +geom_boxplot(width=0.3)+stat_compare_means(aes(group=CT),label = "p.signif")+ 
    ggtitle("")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("ASV Number")+xlab("")

 
 
 
 
 
("non","BAL","Nasal","BAL_Nasal","Oroph","BAL_Oroph","Saliva","Nasal_Oroph","BAL_Saliva","BAL_Nasal_Oroph","Nasal_Saliva","Oroph_Saliva","BAL_Oroph_Saliva","Nasal_Oroph_Saliva","BAL_Nasal_Oroph_Saliva")
 
 
```




# 绘制
```{r}
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%BAL_Interindividual$ID]

OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL[which(rowSums(OTU_TAX_Genus_BAL)>50),]

count<-data.frame(t(OTU_TAX_Genus_BAL))

count$CT<-BAL_Interindividual$CT


count$CT<-ifelse(count$CT == "type1", 1, 0)
count$CT<-as.factor(count$CT)
colnames( count)<-gsub(" ", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("-", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("(", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub(")", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("[", "", colnames( count), fixed = TRUE)
colnames( count)<-gsub("]", "", colnames( count), fixed = TRUE)




randomforest <- randomForest(CT ~.,
                                          count,
                                          ntree =500,
                                          mtry=3,
                                          importance=TRUE ,
                                          proximity=TRUE)
data<-varImpPlot(randomforest, main = "variable importance")



idx<-c("g_Prevotella 7","g_Acinetobacter","g_Fusobacterium","g_Leptotrichia","g_Prevotella")


data<-OTU_TAX_Genus_BAL[rownames(OTU_TAX_Genus_BAL)%in%idx,]
data$Genus<-rownames(data)
data<-melt(data,id.vars = "Genus")
data<-merge(data,BAL_Interindividual,by.x="variable",by.y="ID")
data$Genus<-factor(data$Genus,levels = c("g_Prevotella 7","g_Acinetobacter","g_Fusobacterium","g_Leptotrichia","g_Prevotella"))

ggplot(data,aes(x=Genus, y=value.x,fill=CT))+stat_boxplot(geom = "errorbar",width=0.3) +geom_boxplot(width=0.3)+stat_compare_means(aes(group=CT),label = "p.signif")+ 
    ggtitle("")+ #设置总的标题
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=12,angle = 30,vjust = 1,hjust = 1), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+
    ylab("Relative Abundance (%)")+xlab("")+ylim(c(0,25))
```


```{r}

idx<-c("g_Prevotella 7","g_Acinetobacter","g_Fusobacterium","g_Leptotrichia","g_Prevotella")
rf_phy<-subset_taxa(LungCancer_Normalized_BAL,Genus%in%idx)






## 计算Bray_curtis距离矩阵
dis_bray<- phyloseq::distance(rf_phy, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(rf_phy, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(rf_phy, dis_bray.pcoa, color="CT" ) + geom_point(size=3)







## 绘制初始图形
bray.pcoa <- plot_ordination(subset_samples(LungCancer_Normalized_BAL,ID%in%rownames(heatmap_data)), dis_bray.pcoa, color="CT" ) + geom_point(size=3)
bray.pcoa 
## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")




ggscatter(data, x = "PC1", y = "PC2",
          color = "CT",
          ellipse = FALSE, mean.point = TRUE,
          star.plot = TRUE)
```


# PCA分析
```{r}

idx<-c("g_Prevotella 7","g_Acinetobacter","g_Fusobacterium","g_Leptotrichia","g_Prevotella")
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%BAL_Interindividual$ID]
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL[which(rowSums(OTU_TAX_Genus_BAL)>1),]
PCA_table<-OTU_TAX_Genus_BAL[,colnames(OTU_TAX_Genus_BAL)%in%sample.names(LungCancer_Normalized_BAL_type2)]

# 筛选mad(median absolute deviation,中位数偏差的绝对值的中位数,衡量特异波动的方法)值大于0.5的ASV,
mad.5 <- PCA_table[apply(PCA_table,1,mad)>0.5,]
otu.pca <- prcomp(t(PCA_table))


otu.pca <- prcomp((PCA_table), scale. = TRUE)




p <- fviz_pca_ind(otu.pca,
                  geom.ind = "point", # show points only ( not "text")
                  palette = c("#00AFBB", "#E7B800", "#FC4E07"),
                  addEllipses = TRUE, # Concentration ellipses
                  legend.title = "Groups")
```


#  口肺趋同菌和鼻咽趋同菌的相关性

```{r}
idx<-c("g_Prevotella 7","g_Fusobacterium","g_Leptotrichia","g_Prevotella")
BAL_cor<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%BAL_Interindividual$ID]
BAL_cor<-data.frame(t(BAL_cor))
BAL_cor$index<-BAL_Interindividual$index
BAL_cor$CT<-BAL_Interindividual$CT
BAL_cor$day<-BAL_Interindividual$住院天数
BAL_Interindividual$g_Prevotella.7<-BAL_cor$g_Prevotella.7

idx<-c("g_Corynebacterium 1")
Oroph_cor<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%metadata_Oroph$ID]
Oroph_cor<-data.frame(t(Oroph_cor))
Oroph_cor$index<-metadata_Oroph$index



idx<-c("g_Porphyromonas")
Nasal_cor<-OTU_TAX_Genus_RA[rownames(OTU_TAX_Genus_RA)%in%idx,colnames(OTU_TAX_Genus_RA)%in%metadata_Nasal$ID]
Nasal_cor<-data.frame(t(Nasal_cor))
Nasal_cor$index<-metadata_Nasal$index



ALL_cor<-merge(Nasal_cor,Oroph_cor,by="index")
ALL_cor<-merge(ALL_cor,BAL_cor,by="index")




ggplot(ALL_cor, aes(x=g_Porphyromonas, y= g_Prevotella.7,color=CT,group=CT)) +
    geom_point(aes(fill=CT),shape = 21,size=2.5) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1, span = 1,aes(fill=CT)) +
theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")



ggplot(ALL_cor, aes(x=g_Corynebacterium.1, y= g_Prevotella.7)) +
    geom_point(aes(fill=CT),shape = 21,size=3) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1 ) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")+xlab("Porphyromonas in Nasal")+ylab("Leptotrichia in BAL")
```




# 绘制整体样本的圈图
```{r}
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(tidytree)
library(ggstar)
library(ggplot2)
library(ggnewscale)



tree <- read.tree("treedata-book/data/HMP_tree/hmptree.nwk")
# the abundance and types of microbes
dat1 <- read.csv("treedata-book/data/HMP_tree/tippoint_attr.csv")
# the abundance of microbes at different body sites.
dat2 <- read.csv("treedata-book/data/HMP_tree/ringheatmap_attr.csv")
# the abundance of microbes at the body sites of greatest prevalence.
dat3 <- read.csv("treedata-book/data/HMP_tree/barplot_attr.csv")



# adjust the order
dat2$Sites <- factor(dat2$Sites, levels=c("Stool (prevalence)", "Cheek (prevalence)",
                                          "Plaque (prevalence)","Tongue (prevalence)",
                                          "Nose (prevalence)", "Vagina (prevalence)",
                                          "Skin (prevalence)"))
dat3$Sites <- factor(dat3$Sites, levels=c("Stool (prevalence)", "Cheek (prevalence)",
                                          "Plaque (prevalence)", "Tongue (prevalence)",
                                          "Nose (prevalence)", "Vagina (prevalence)",
                                          "Skin (prevalence)"))
# extract the clade label information. Because some nodes of tree are annotated to genera,
# which can be displayed with high light using ggtree.
nodeids <- nodeid(tree, tree$node.label[nchar(tree$node.label)>4])
nodedf <- data.frame(node=nodeids)
nodelab <- gsub("[\\.0-9]", "", tree$node.label[nchar(tree$node.label)>4])
# The layers of clade and hightlight
poslist <- c(1.6, 1.4, 1.6, 0.8, 0.1, 0.25, 1.6, 1.6, 1.2, 0.4,
             1.2, 1.8, 0.3, 0.8, 0.4, 0.3, 0.4, 0.4, 0.4, 0.6,
             0.3, 0.4, 0.3)

labdf <- data.frame(node=nodeids, label=nodelab, pos=poslist)



# The circular layout tree.
p <- ggtree(tree, layout="fan", size=0.15, open.angle=5) +
     geom_hilight(data=nodedf, mapping=aes(node=node),
                  extendto=6.8, alpha=0.3, fill="grey", color="grey50",
                  size=0.05) +
     geom_cladelabel2(data=labdf, 
                   mapping=aes(node=node, 
                               label=label,
                               offset.text=pos),
                   hjust=0.5,
                   angle="auto",
                   barsize=NA,
                   horizontal=FALSE, 
                   fontsize=1.4,
                   fontface="italic"
                   )







```



# 绘制tree
```{r}
# 构建带进化树的phyloseq对象

## 读取数据表格
TableOneData<-read.table("TableOne_data.txt",sep = "\t",header = TRUE,row.names = 1)
## 过滤掉没有测序数据且非肺癌的患者信息
TableOneData<-subset(TableOneData,data=="yes"&diagnosis%in%c("Cancer","Health"))
#读取由QIIME生成的biom文件
OtuTable.biom = read_biom("feature-table.10000.biom")
#将biom对象转化成矩阵对象
OtuTable<-as(biom_data(OtuTable.biom), "matrix")
#读取TAX
TaxTable<-read.table("silva.taxonomy.txt",sep="\t",row.names = 1,header = TRUE,stringsAsFactors = FALSE)
TaxTable <- as.matrix(TaxTable)
#读取metadata
MetaTable<-read.table("metadata_new.txt",sep="\t",row.names = 1,header = TRUE)
TableOneData<-read.table("TableOne_data.txt",sep = "\t",header = TRUE,row.names = 1)
TableOneData$index<-row.names(TableOneData)
MetaTable<-merge(MetaTable,TableOneData,by = "index")
rownames(MetaTable)<-MetaTable$ID
#生成phyloseq对象
OTU = otu_table(OtuTable, taxa_are_rows = TRUE)
samples = sample_data(MetaTable)
TAX = tax_table(TaxTable)
root_tree<- read.tree("tree.nwk") 

tree_phyloseq <- phyloseq::phyloseq(OTU, TAX, samples,root_tree)


#过滤掉未确诊或非肺癌患者
tree_phyloseq<-subset_samples(tree_phyloseq,diagnosis%in%c("Cancer","Health"))

#过滤掉肺癌患者的健康侧BAL
tree_phyloseq<-subset_samples(tree_phyloseq,!(BALgroup=="HS"))

#选择腺癌患者和对照组
tree_phyloseq<-subset_samples(tree_phyloseq, type%in%c("腺癌","Health"))

#过滤掉使用过抗生素的人群
tree_phyloseq<-subset_samples(tree_phyloseq,抗生素使用=="N")
#过滤掉reads数少于10的ASV
tree_phyloseq<- filter_taxa(tree_phyloseq,function(x) sum(x) > 10 , TRUE )

tree_phyloseq<- filter_taxa(tree_phyloseq,function(x) sum(x) > 1000 , TRUE )


tree_phyloseq<- subset_taxa(tree_phyloseq,!(Phylum==""))

tree_phyloseq<- subset_taxa(tree_phyloseq,!(Genus==""))
# 提取物种信息
tippoint_attr<-data.frame(tax_table(tree_phyloseq))

# 读取致病菌信息
pathogen<-read.table("rep_sequences_pathogen.txt",header = TRUE)

# 注释致病菌信息

tippoint_attr$ID<-rownames(tippoint_attr)
tippoint_attr<-merge(tippoint_attr,pathogen,by="ID",all.x=TRUE)

tippoint_attr<-tippoint_attr[,c("ID","Phylum","Type","Genus")]
tippoint_attr$Type<-as.character(tippoint_attr$Type)
tippoint_attr$Type[is.na(tippoint_attr$Type)]<-"Commensal_microbes"





# 计算每个物种最优势的群体
# BAL
tree_phyloseq_BAL<-subset_samples(tree_phyloseq,sample=="BAL")
tree_table_BAL<-data.frame(otu_table(tree_phyloseq_BAL))

tree_table_BAL<-t(tree_table_BAL)
tree_table_BAL<-tree_table_BAL/apply(tree_table_BAL,1,sum)
tree_table_BAL<-data.frame(t(tree_table_BAL))
tree_table_BAL$mean<-rowMeans(tree_table_BAL)

# Saliva
tree_phyloseq_Saliva<-subset_samples(tree_phyloseq,sample=="Saliva")
tree_table_Saliva<-data.frame(otu_table(tree_phyloseq_Saliva))

tree_table_Saliva<-t(tree_table_Saliva)
tree_table_Saliva<-tree_table_Saliva/apply(tree_table_Saliva,1,sum)
tree_table_Saliva<-data.frame(t(tree_table_Saliva))
tree_table_Saliva$mean<-rowMeans(tree_table_Saliva)


# Oroph
tree_phyloseq_Oroph<-subset_samples(tree_phyloseq,sample=="Oroph")
tree_table_Oroph<-data.frame(otu_table(tree_phyloseq_Oroph))

tree_table_Oroph<-t(tree_table_Oroph)
tree_table_Oroph<-tree_table_Oroph/apply(tree_table_Oroph,1,sum)
tree_table_Oroph<-data.frame(t(tree_table_Oroph))
tree_table_Oroph$mean<-rowMeans(tree_table_Oroph)

# Nasal
tree_phyloseq_Nasal<-subset_samples(tree_phyloseq,sample=="Nasal")
tree_table_Nasal<-data.frame(otu_table(tree_phyloseq_Nasal))

tree_table_Nasal<-t(tree_table_Nasal)
tree_table_Nasal<-tree_table_Nasal/apply(tree_table_Nasal,1,sum)
tree_table_Nasal<-data.frame(t(tree_table_Nasal))
tree_table_Nasal$mean<-rowMeans(tree_table_Nasal)

# Fecal
#tree_phyloseq_Fecal<-subset_samples(tree_phyloseq,sample=="Fecal")
#tree_table_Fecal<-data.frame(otu_table(tree_phyloseq_Fecal))

#tree_table_Fecal<-t(tree_table_Fecal)
#tree_table_Fecal<-tree_table_Fecal/apply(tree_table_Fecal,1,sum)
#tree_table_Fecal<-data.frame(t(tree_table_Fecal))
#tree_table_Fecal$mean<-rowMeans(tree_table_Fecal)
# 计算相对丰度
#all_mean<-data.frame(BAL=tree_table_BAL$mean,Saliva=tree_table_Saliva$mean,Oroph=tree_table_Oroph$mean,Nasal=tree_table_Nasal$mean,Fecal=tree_table_Fecal$mean)

all_mean<-data.frame(BAL=tree_table_BAL$mean,Saliva=tree_table_Saliva$mean,Oroph=tree_table_Oroph$mean,Nasal=tree_table_Nasal$mean)
row.names(all_mean)<-row.names(tree_table_BAL)

max<-c()
# 挑选最高相对丰度

for (i in 1:378){
  max[i]<-rownames(data.frame(which.max(all_mean[i,])))
  
}
max.value<-c()
for (i in 1:378){
  max.value[i]<-max(all_mean[i,])
  
}
all_mean$Sites<-max
all_mean$HigherAbundance<-max.value
all_mean$ID<-rownames(all_mean)

barplot_attr<-all_mean[,c("ID","Sites","HigherAbundance")]




# ringheatmap_attrringheatmap_attr

ringheatmap_attr <- all_mean[,1:4]
ringheatmap_attr <- data.frame(t(ringheatmap_attr))
colnames(ringheatmap_attr)<- all_mean$ID
ringheatmap_attr$Sites<-rownames(ringheatmap_attr)
ringheatmap_attr<-melt(ringheatmap_attr,id.vars = "Sites")
ringheatmap_attr<-data.frame(ID=ringheatmap_attr$variable,Sites=ringheatmap_attr$Sites,Abundance=ringheatmap_attr$value)




write.table(ringheatmap_attr,"tree/ringheatmap_attr.txt",sep="\t")
write.table(tippoint_attr,"tree/tippoint_attr.txt",sep="\t")
write.table(barplot_attr,"tree/barplot_attr.txt",sep="\t")
#提取树文件

tree_zjx<-phy_tree(tree_phyloseq)

tree_zjx$tip.label<-as.character(tippoint_attr$Genus)
ggtree(tree_zjx, layout="fan", size=0.15, open.angle=5)+ geom_tiplab()



# 重新制作进化树
write.table(tippoint_attr,"tippoint_attr.txt",sep="\t")

new_tree<- read.tree("new_tree.nwk") 

ggtree(new_tree, layout="fan", size=0.15, open.angle=5)

tip_label<-data.frame(new_tree$tip.label)


tip_label<-merge(tip_label,tippoint_attr,by="ID",sort =FALSE)


new_tree$tip.label<-as.character(tip_label$Genus)



# 绘制进化树
# 读取Genus的nodes文件
Genus_nodes<-read.table("Genus_nodes.txt",header = TRUE)

dat1 <- tippoint_attr
dat2 <- ringheatmap_attr
dat3 <- barplot_attr
node_ID<-data.frame(paste("N", 1:378, sep = ""))
colnames(node_ID)<-"ID"
new_tree$node.label<-paste("N", 0:376, sep = "")
nodedf<-data.frame(node=Genus_nodes$node)
ggtree(new_tree, layout="fan", size=0.15, open.angle=5)+
  geom_hilight(data=nodedf, mapping=aes(node=node),
                  extendto=6.8, alpha=0.3, fill="grey", color="grey50",
                  size=0.05) + geom_cladelabel(data=Genus_nodes, 
                   mapping=aes(node=node, 
                               label=Genus,
                               offset.text=1),
                   hjust=0.5,
                   angle="auto",
                   barsize=NA,
                   horizontal=FALSE, 
                   fontsize=1.4,
                   fontface="italic"
                   )
```



# 读取表型数据 
```{r}
BugBase<-read.table("BugBase/predicted_phenotypes/predictions.txt",header = TRUE)
BugBase$ID<-row.names(BugBase)
BugBase <- merge(BAL_Interindividual,BugBase,by="ID")
ggplot(BugBase,aes(x=CT,y=BugBase$Aerobic,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()

test<-BugBase[,46:54]
test<-scale(x =test )
test<-data.frame(t(test))
colnames(test)<-BugBase$ID


library(circlize)
col_fun = colorRamp2(c(-3, 0,3), c("white", "yellow", "red"))
col_fun(seq(-3, 3))

ha = HeatmapAnnotation(CT =BugBase$CT,
                       gas =BugBase$gastroenteropathy,
                       smoking =BugBase$smoking,
                       drinking=BugBase$drinking,
                       gender=BugBase$gender,
                       native=BugBase$native,
                       college=BugBase$college,
                       age=BugBase$age,
                       col = list(CT = c("type1" = "#F8766D", "type2" = "#00BFC4"),
                                  gas = c("Yes" = "#FF8C00", "No" = "#228B22"),
                                  smoking =c("Yes" = "#8B0000", "No" = "#008080"),
                                  drinking=c("Yes" = "#8B0000", "No" = "#008080"),
                                  gender=c("male" = "#4169E1", "female" = "#FF4500"),
                                  native=c("Yes" = "#9400D3", "No" = "#CD853F"),
                                  college=c("Yes" = "#FFC0CB", "No" = "#F0E68C")),
                       na_col = "black")



col_fun = colorRamp2(c(-3, 0,3), c("white", "yellow", "red"))
col_fun(seq(-3, 3))

Heatmap(test,name = "Z-score",top_annotation = ha,column_order =order(BugBase$CT),col = col_fun,show_column_names = FALSE)
``` 


# Bugbase预测kegg数据
```{r}

Bugbase_kegg<-read.table("BugBase/kegg/Bugbase_kegg_predictions.txt",header = TRUE)
module_class<- read.table("BugBase/kegg/ko00002.txt",header = TRUE)
module_describe <- read.table("BugBase/kegg/module_function.txt",header = TRUE)


Bugbase_kegg$ID<-row.names(Bugbase_kegg)
Bugbase_kegg <- merge(BAL_Interindividual,Bugbase_kegg,by="ID")

kegg_data <- Bugbase_kegg[,c(1,46:603)]
rownames(kegg_data) <-kegg_data [,1]
kegg_data<-kegg_data [,-1]
kegg_data<-data.frame(t(kegg_data))


kegg_data$Module<-rownames(kegg_data)

kegg_data<-melt(kegg_data,id.vars ="Module")

kegg_data<-merge(kegg_data,BAL_Interindividual,by.x="variable",by.y="ID")

results<-compare_means(value.x~CT, data=kegg_data,group.by = "Module")
results<-subset(results,p < 0.05)


# 合并相同的模块
Bugbase_kegg<-read.table("BugBase/kegg/Bugbase_kegg_predictions.txt",header = TRUE)
module_class<- read.table("BugBase/kegg/ko00002.txt",header = TRUE)

kegg_table <- data.frame(t(Bugbase_kegg))
kegg_table$module_ID <- row.names(kegg_table)

kegg_module_B<-merge(module_class,kegg_table,by="module_ID")
kegg_module_B<-kegg_module_B[,-1]


kegg_module_B<-aggregate(kegg_module_B[,4:762],list(kegg_module_B$B),sum)
colnames(kegg_module_B)[1]<-"module_classB"

kegg_module_B$module_classB<-as.character(kegg_module_B$module_classB)
kegg_module_B[2,1]<-"other_secondary_metabolites"
kegg_module_B[3,1]<-"terpenoids_and_polyketides"
# 筛选BAL样本
kegg_module_B_BAL<-kegg_module_B[,colnames(kegg_module_B)%in%c(BAL_Interindividual$ID,"module_classB")]

kegg_module_B_BAL<-melt(kegg_module_B_BAL,id.vars ="module_classB")

kegg_module_B_BAL<-merge(kegg_module_B_BAL,BAL_Interindividual,by.x="variable",by.y="ID")

results<-compare_means(value.x~CT, data=kegg_module_B_BAL,group.by = "module_classB")


results<-subset(results,p < 0.05)

# 

kegg_module_B_BAL<-kegg_module_B[,colnames(kegg_module_B)%in%c(BAL_Interindividual$ID,"module_classB")]
rownames(kegg_module_B_BAL)<-kegg_module_B_BAL[,1]
kegg_module_B_BAL<-kegg_module_B_BAL[,-1]
kegg_module_B_BAL<-data.frame(t(kegg_module_B_BAL))
test.1<-scale(x =kegg_module_B_BAL)
test.1<-data.frame(t(test.1))



library(circlize)


ha = HeatmapAnnotation(CT =BugBase$CT,
                       gas =BugBase$gastroenteropathy,
                       smoking =BugBase$smoking,
                       drinking=BugBase$drinking,
                       gender=BugBase$gender,
                       native=BugBase$native,
                       college=BugBase$college,
                       age=BugBase$age,
                       col = list(CT = c("type1" = "#F8766D", "type2" = "#00BFC4"),
                                  gas = c("Yes" = "#FF8C00", "No" = "#228B22"),
                                  smoking =c("Yes" = "#8B0000", "No" = "#008080"),
                                  drinking=c("Yes" = "#8B0000", "No" = "#008080"),
                                  gender=c("male" = "#4169E1", "female" = "#FF4500"),
                                  native=c("Yes" = "#9400D3", "No" = "#CD853F"),
                                  college=c("Yes" = "#FFC0CB", "No" = "#F0E68C")),
                       na_col = "black")

col_fun = colorRamp2(c(-3, 0,3), c("#00008B", "white", "#8B0000"))
col_fun(seq(-3, 3))


Heatmap(test.1,name = "Z-score",top_annotation = ha,column_order =order(BugBase$CT),col = col_fun,show_column_names = FALSE,cluster_rows = FALSE )



data<-kegg_module_B_BAL
data$ID<-rownames(data)
data<-merge(data,BAL_Interindividual,by="ID")

ggplot(data,aes(x=CT,y=data$Amino_acid_metabolism,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()
```

# 读取毒力和耐药基因
```{r}
CARD<-read.table("CARD_VFDB/CARD.txt",header = TRUE)
CARD_DISCRIBE<-read.table("CARD_VFDB/CARD_DISCRIBE.txt",header = TRUE)
CARD_BAL<-CARD[,colnames(CARD)%in%c(BAL_Interindividual$ID,"ID")]


colnames(CARD_BAL)[1]<-"ARO"

CARD_BAL<-melt(CARD_BAL,id.vars ="ARO")

CARD_BAL<-merge(CARD_BAL,BAL_Interindividual,by.x="variable",by.y="ID")

results<-compare_means(value.x~CT, data=CARD_BAL,group.by = "ARO")
results<-subset(results,p < 0.05)

# 给显著的ARO添加标签
results<-merge(results,CARD_DISCRIBE,by.x="ARO",by.y="ID")
results<-unite(results, "AROS", ARO, genu,remove=FALSE)

# 获取完整的CARD表

CARD_BAL<-CARD[,colnames(CARD)%in%c(BAL_Interindividual$ID,"ID")]
CARD_BAL<-subset(CARD_BAL,ID%in%results$ARO)

rownames(CARD_BAL)<-CARD_BAL$ID
CARD_BAL<-CARD_BAL[,-1]
CARD_BAL<-data.frame(t(CARD_BAL))
CARD_BAL$ID<-rownames(CARD_BAL)
CARD_BAL<-merge(BAL_Interindividual,CARD_BAL,by="ID",all.x=TRUE)

CARD_BAL<-CARD_BAL[,46:55]

rownames(CARD_BAL)<-BAL_Interindividual$ID

CARD_BAL<-scale(x =CARD_BAL)
CARD_BAL<-data.frame(t(CARD_BAL))

col_fun = colorRamp2(c(-0.5, 0,3), c("#00008B", "white", "#8B0000"))
col_fun(seq(-0.5, 3))
rownames(CARD_BAL)<-results$AROS
Heatmap(CARD_BAL,name = "Z-score",top_annotation = ha,column_order =order(BugBase$CT),col = col_fun,show_column_names = FALSE,cluster_rows = FALSE )



# 绘制存在差异耐药基因的箱线图
data<-data.frame(t(CARD_BAL))
data$ID<-rownames(data)
data<-merge(data,BAL_Interindividual,by="ID")







ggplot(data,aes(x=CT,y=ARO_3000375_ErmB,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()
```



```{r}
VFDB<-read.table("CARD_VFDB/VFDB.txt",header = TRUE)

VFDB_BAL<-VFDB[,colnames(VFDB)%in%c(BAL_Interindividual$ID,"ID")]
rownames(VFDB_BAL)<-VFDB_BAL$ID
VFDB_BAL<-VFDB_BAL[,-1]
VFDB_BAL$sum<-rowSums(VFDB_BAL)
VFDB_BAL<-subset(VFDB_BAL,sum >0)


VFDB_BAL$VF<-rownames(VFDB_BAL)

VFDB_BAL<-melt(VFDB_BAL,id.vars ="VF")

VFDB_BAL<-merge(VFDB_BAL,BAL_Interindividual,by.x="variable",by.y="ID")

results<-compare_means(value.x~CT, data=VFDB_BAL,group.by = "VF")
results<-subset(results,p < 0.05)

results<-subset(results,p <7.151395e-03)
# 获取完整的VFDB表
idx<-c(14,10,9,8,7,6,5,4,2)
results<-results[idx,]

VFDB_BAL<-VFDB[,colnames(VFDB)%in%c(BAL_Interindividual$ID,"ID")]
VFDB_BAL<-subset(VFDB_BAL,ID%in%results$VF)

rownames(VFDB_BAL)<-VFDB_BAL$ID
VFDB_BAL<-VFDB_BAL[,-1]
VFDB_BAL<-data.frame(t(VFDB_BAL))
VFDB_BAL$ID<-rownames(VFDB_BAL)
VFDB_BAL<-merge(BAL_Interindividual,VFDB_BAL,by="ID",all.x=TRUE)

VFDB_BAL<-VFDB_BAL[,46:54]

rownames(VFDB_BAL)<-BAL_Interindividual$ID

VFDB_BAL<-scale(x =VFDB_BAL)
VFDB_BAL<-data.frame(t(VFDB_BAL))

col_fun = colorRamp2(c(-0.5, 0,3), c("#00008B", "white", "#8B0000"))
col_fun(seq(-0.5, 3))
rownames(VFDB_BAL)<-c("ErmBP_virulence","CfxA2-like_virulence","hypothetical protein SP_0728","5,10-methylenetetrahydrofolate reductase","transcriptional regulator","metallo-beta-lactamase superfamily protein","PTS system transporter subunit IID","UDP-N-acetylglucosamine acyltransferase","7,8-dihydro-8-oxoguanine-triphosphatase")
Heatmap(VFDB_BAL,name = "Z-score",top_annotation = ha,column_order =order(BugBase$CT),col = col_fun,show_column_names = FALSE,cluster_rows = FALSE )



# 绘制存在差异耐药基因的箱线图
data<-data.frame(t(VFDB_BAL))
data$ID<-rownames(data)
data<-merge(data,BAL_Interindividual,by="ID")







ggplot(data,aes(x=CT,y=ARO_3000375_ErmB,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()

```







# 16s物种热图
```{r}
idx<-c("g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rothia","g_Leptotrichia","g_Streptococcus","g_Lactobacillus","g_Acinetobacter","g_Pseudomonas")


Genus_BAL<-OTU_TAX_Genus_BAL[rownames(OTU_TAX_Genus_BAL)%in%idx,]
Genus_BAL<-data.frame(t(Genus_BAL))
Genus_BAL<-scale(x=Genus_BAL)
Genus_BAL<-log10(Genus_BAL+1)

col_fun = colorRamp2(c(0, 0.3,1.5), c( "#04040D", "#5B94B3","#FAFAD2"))
col_fun(seq(0, 1.5))

Heatmap(Genus_BAL,name = "log10(RA)",top_annotation = ha,column_order =order(BugBase$CT),col = col_fun,show_column_names = FALSE)
```



# 厌氧菌和好氧菌比例与肺功能之间的关系
```{r}
BugBase<-read.table("BugBase/predicted_phenotypes/predictions.txt",header = TRUE)
BugBase<-BugBase[rownames(BugBase)%in%lung_function_merge$ID,]

BugBase_lungfunction<-cbind.data.frame(lung_function_merge,BugBase)

BugBase_lungfunction<-subset(BugBase_lungfunction,CT=="type2")

LF<-c("FEV1实际.预测","FEV1.FVC实际预测","PEF实际.预测","FVC实际.预测","MEF75实际.预测" ,"MEF25实际.预测","MEF75.25实际.预测","MEF50实际.预测","DLCO.SB实际.预测")
 

BugBase_lungfunction_corr <-corr.test(x=BugBase_lungfunction[,95:103],y=BugBase_lungfunction[,47:94], method = 'spearman',adjust = "none")
corrplot(BugBase_lungfunction_corr$r, p.mat=BugBase_lungfunction_corr$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")


ggplot(BugBase_lungfunction, aes(x=BugBase_lungfunction$Facultatively_Anaerobic, y= BugBase_lungfunction$FEV1.FVC实际预测)) +
    geom_point(aes(fill=CT),shape = 21,size=3) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1 ) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")+xlab("Porphyromonas in Nasal")+ylab("Leptotrichia in BAL")

```

# 厌氧菌和好氧菌比例与细胞因子之间的关系
```{R}
BugBase<-read.table("BugBase/predicted_phenotypes/predictions.txt",header = TRUE)
BugBase<-BugBase[rownames(BugBase)%in%rownames(cytokine),]

BugBase_cytokine<-cbind.data.frame(cytokine,BugBase)
BugBase_cytokine$ID<-row.names(BugBase_cytokine)
BugBase_cytokine<-merge(BugBase_cytokine,BAL_Interindividual,by="ID")


BugBase_cytokine_corr <-corr.test(x=BugBase_cytokine[,35:43],y=BugBase_cytokine[,2:34], method = 'spearman',adjust = "none")
corrplot(BugBase_cytokine_corr$r, p.mat=BugBase_cytokine_corr$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")




ggplot(BugBase_lungfunction, aes(x=BugBase_lungfunction$Aerobic, y= BugBase_lungfunction$CT)) +
    geom_point(aes(fill=CT),shape = 21,size=3) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1 ) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")+xlab("Porphyromonas in Nasal")+ylab("Leptotrichia in BAL")
```






# 厌氧菌和好氧菌比例与物种组成之间的关系
```{R}
BugBase<-read.table("BugBase/predicted_phenotypes/predictions.txt",header = TRUE)
idx<-c("g_Prevotella 7","g_Fusobacterium","g_Veillonella","g_Porphyromonas","g_Rothia","g_Leptotrichia","g_Streptococcus","g_Lactobacillus","g_Acinetobacter","g_Pseudomonas")
BugBase<-BugBase[rownames(BugBase)%in%BAL_Interindividual$ID,]
Genus_BAL<-OTU_TAX_Genus_BAL[rownames(OTU_TAX_Genus_BAL)%in%idx,]
Genus_BAL<-data.frame(t(Genus_BAL))
BugBase_Genus<-cbind.data.frame(Genus_BAL,BugBase)
BugBase_Genus$ID<-row.names(BugBase_Genus)
BugBase_Genus<-merge(BugBase_Genus,BAL_Interindividual,by="ID")


BugBase_Genus_corr <-corr.test(x=BugBase_Genus[,12:20],y=BugBase_Genus[,2:11], method = 'spearman',adjust = "none")
corrplot(BugBase_Genus_corr$r, p.mat=BugBase_Genus_corr$p, insig = "label_sig",sig.level = c(.001, .01, .05), pch.cex = .9, pch.col = "black")




ggplot(BugBase_Genus, aes(x=BugBase_lungfunction$Facultatively_Anaerobic, y= BugBase_Genus$g_Lactobacillus)) +
    geom_point(aes(fill=CT),shape = 21,size=3) +
    geom_smooth(method = "lm", alpha = 0.3, size = 1 ) +
    theme_bw()+ #背景变为白色
    theme(legend.position="none", #不需要图例
          axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
          axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
          axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
          axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
          plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
          panel.grid.major = element_blank(), #不显示网格线
          panel.grid.minor = element_blank())+stat_cor(method = "spearman")+xlab("Porphyromonas in Nasal")+ylab("Leptotrichia in BAL")
```



# 粪便中B/P的分布特征
```{R}
# 提取Bacteroides和Prevotella

Bacteroides_Prevotella<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%colnames(OTU_TAX_Genus_Fecal)]

idx<-c("g_Prevotella","g_Prevotella 1","g_Prevotella 2","g_Prevotella 6","g_Prevotella 7","g_Prevotella 9","g_Bacteroides")
Bacteroides_Prevotella<-Bacteroides_Prevotella[rownames(Bacteroides_Prevotella)%in%idx,]
Bacteroides_Prevotella<-sum()
Bacteroides_Prevotella["Prevotella",]<-Bacteroides_Prevotella[2,]+Bacteroides_Prevotella[3,]+Bacteroides_Prevotella[4,]+Bacteroides_Prevotella[5,]+Bacteroides_Prevotella[6,]+Bacteroides_Prevotella[7,]

Bacteroides_Prevotella<-Bacteroides_Prevotella[-2:-7,]

rownames(Bacteroides_Prevotella)[1]<-"Bacteroides"

Bacteroides_Prevotella<-data.frame(t(Bacteroides_Prevotella))


Bacteroides_Prevotella$sigma<-Bacteroides_Prevotella$Prevotella-Bacteroides_Prevotella$Bacteroides
Bacteroides_Prevotella$delta<-100-(Bacteroides_Prevotella$Prevotella+Bacteroides_Prevotella$Bacteroides)

Bacteroides_Prevotella$CT<-sample_data(LungCancer_Normalized_Fecal)$CT
plot(x=Bacteroides_Prevotella$sigma,Bacteroides_Prevotella$delta)



ggplot(Bacteroides_Prevotella, aes(sigma, delta)) +
  #绘制样本点，根据分组匹配颜色和形状，size调整点的大小
  geom_point(aes(colour=CT,shape=CT,fill=CT),size=6)+
  #匹配形状、边框和填充的图例
  scale_shape_manual(values=pich)+
  scale_colour_manual(values=Palette)+
  #scale_fill_manual(values=cbbPalette)+
  #设置标题和横纵坐标label文字
  labs(title="PCoA - The composition of BAL microbiome") + 
  xlab(paste("PC1 ( ",pc1,"%"," )",sep="")) + 
  ylab(paste("PC2 ( ",pc2,"%"," )",sep=""))+
  theme(text=element_text(size=30))+
  #添加横纵两条虚线
  geom_vline(aes(xintercept = 0),linetype="dotted")+
  geom_hline(aes(yintercept = 0),linetype="dotted")+
  #调整背景、坐标轴、图例的格式
  theme(panel.background = element_rect(fill='white', colour='black'),
        panel.grid=element_blank(), 
        axis.title = element_text(color='black',size=14),
        axis.ticks.length = unit(0.4,"lines"), axis.ticks = element_line(color='black'),
        axis.line = element_line(colour = "black"), 
        axis.title.x=element_text(colour='black', size=14),
        axis.title.y=element_text(colour='black', size=14),
        axis.text=element_text(colour='black',size=14),
        legend.title=element_blank(),
        legend.text=element_text(size=14),
        legend.key=element_blank(),
        legend.background = element_rect(colour = "black"),
        legend.key.height=unit(1.6,"cm"),
        legend.position="none")+
  #设置标题的格式
  theme(plot.title = element_text(size=14,colour = "black",hjust = 0.5,face = "bold"))+geom_line(aes(x = sigma,y = poly2))



lmp2<-lm(delta~poly(sigma, 2),data = Bacteroides_Prevotella)
poly2 <- predict(lmp2,Bacteroides_Prevotella)

Bacteroides_Prevotella$poly2<-poly2 
```


#个位点的CT相关性
```{r}
count<-as.matrix(OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sample.names(LungCancer_Normalized_Fecal)])
count<-t(count[which(rowSums(count) >100),])

set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)
best 


  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  #d1 <-d1[20:47,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
 # d2 <-d2[20:47,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))






Fecal_CT<-data.frame(ass)
Fecal_CT$ID<-rownames(Fecal_CT)

colnames(Fecal_CT)<-c("Fecal_CT","ID")

# 挑选全套配套样本

data<-new_MetaTable[,c(1,5)]
data<-subset(data,!(sample=="Env"))
data<-dcast(data = data,index~sample)
data<-subset(data,BAL>=1&Nasal>=1&Fecal>=1&Saliva>=1&Oroph>=1)
data<-subset(data,index%in%BAL_Interindividual$index)
idx<-data$index
idx<-idx[-19]
ALL_pair_name<-sample.names(subset_samples(LungCancer_Normalized,index%in%idx))

BAL_CT<-subset(BAL_Interindividual,ID%in%ALL_pair_name)
BAL_CT<-BAL_CT[,c("index","CT")]


Nasal_CT<-subset(Nasal_CT,ID%in%ALL_pair_name)


Saliva_CT<-subset(Saliva_CT,ID%in%ALL_pair_name)


Oroph_CT<-subset(Oroph_CT,ID%in%ALL_pair_name)


Fecal_CT<-subset(Fecal_CT,ID%in%ALL_pair_name)

BAL_CT$CT<-ifelse(BAL_CT$CT=="type1","1","2")

CT_ALL_pair<-cbind.data.frame(BAL_CT$CT,Nasal_CT$Nasal_CT,Saliva_CT$Saliva_CT,Oroph_CT$Oroph_CT,Fecal_CT$Fecal_CT)


colnames(CT_ALL_pair)<-c("BAL","Nasal","Saliva","Oroph","Fecal")



```




# 绘制BAL与唾液属水平中性模型
```{r}
BAL_spp<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%BAL_Interindividual$ID]
Saliva_pool<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%colnames(OTU_TAX_Genus_Saliva)]

spp <-t(BAL_spp)
pool <-t(Saliva_pool)

tax_id<-data.frame(row.names(OTU_TAX_Genus))
row.names(tax_id)<-tax_id$row.names.OTU_TAX_Genus.
colnames(tax_id)<-"Genus"
spp.out <- fit_sncm(spp, pool=pool, taxon=tax_id)
P_BS <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())



BS_midmod<-spp.out$predictions
















BAL_spp<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%BAL_Interindividual$ID]
Nasal_pool<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%colnames(tree_table_Nasal)]

spp <-t(BAL_spp)
pool <-t(Nasal_pool)

tax_id<-data.frame(row.names(OTU_TAX_Genus))
row.names(tax_id)<-tax_id$row.names.OTU_TAX_Genus.
colnames(tax_id)<-"Genus"
spp.out <- fit_sncm(spp, pool=pool, taxon=tax_id)
P_BN <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())



BN_midmod<-spp.out$predictions





BAL_spp<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%BAL_Interindividual$ID]
Oroph_pool<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%colnames(OTU_TAX_Genus_Oroph)]

spp <-t(BAL_spp)
pool <-t(Oroph_pool)

tax_id<-data.frame(row.names(OTU_TAX_Genus))
row.names(tax_id)<-tax_id$row.names.OTU_TAX_Genus.
colnames(tax_id)<-"Genus"
spp.out <- fit_sncm(spp, pool=pool, taxon=tax_id)
P_BO <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())



BO_midmod<-spp.out$predictions

















BS_midmod_above<-subset(BS_midmod,fit_class=="Above prediction")
BS_midmod_predicted<-subset(BS_midmod,fit_class=="As predicted")
BS_midmod_below<-subset(BS_midmod,fit_class=="Below prediction")



















# 挑选各组前10的菌


BS_midmod_above<-subset(BS_midmod_above,freq>0.63)
BS_midmod_predicted<-subset(BS_midmod_predicted,freq>=0.96969696)
BS_midmod_predicted<-subset(BS_midmod_predicted,pred.upr>=1)
BS_midmod_below<-subset(BS_midmod_below,freq>=	0.84848484)
BS_midmod_below<-subset(BS_midmod_below,pred.upr>0.9699190)
BS_midmod_above<-melt(BS_midmod_above,id.vars = "Genus")
BS_midmod_above<-subset(BS_midmod_above,variable%in%c("freq","freq.pred"))
BS_midmod_above$value<-as.numeric(BS_midmod_above$value)
BS_midmod_below<-melt(BS_midmod_below,id.vars = "Genus")
BS_midmod_below<-subset(BS_midmod_below,variable%in%c("freq","freq.pred"))
BS_midmod_below$value<-as.numeric(BS_midmod_below$value)
BS_midmod_predicted<-melt(BS_midmod_predicted,id.vars = "Genus")
BS_midmod_predicted<-subset(BS_midmod_predicted,variable%in%c("freq","freq.pred"))
BS_midmod_predicted$value<-as.numeric(BS_midmod_predicted$value)

ggplot(data=BS_midmod_above, mapping=aes(x=Genus,y=value))+
    geom_bar(aes(fill=variable),stat="identity",position="dodge")+scale_fill_manual(name = "Group", values = c(`freq` = "cyan4", `freq.pred`="grey50" ))+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 0.1))




ggplot(data= BS_midmod_predicted, mapping=aes(x=Genus,y=value))+
    geom_bar(aes(fill=variable),stat="identity",position="dodge")+scale_fill_manual(name = "Group", values = c(`freq` = "blue", `freq.pred`="grey50" ))+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 1,hjust = 1))


number<-c(101,94,47)
class<-c("Above prediction","As predicted","Below prediction")

data<-data.frame(number,class)

ggplot(data=data, mapping=aes(x=class,y=number,fill=class))+
    geom_bar(stat="identity")+scale_fill_manual(name = "Prediction", values = c(`Above prediction` = "cyan4", `As predicted` = "blue", `Below prediction` = "goldenrod"))+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 0.6))
```










```{r}
CARD<-read.table("CARD_VFDB/CARD.txt",header = TRUE)

CARD_DISCRIBE<-read.table("CARD_VFDB/CARD_DISCRIBE.txt",header = TRUE)

Fecal_meatadata<-data.frame(sample_data(LungCancer_Normalized_Fecal))

CARD_Fecal<-CARD[,colnames(CARD)%in%(Fecal_meatadata$ID)]

CARD_Fecal$ARO<-CARD$ID


CARD_Fecal<-melt(CARD_Fecal,id.vars ="ARO")

CARD_Fecal<-merge(CARD_Fecal,Fecal_meatadata,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=CARD_Fecal,group.by = "ARO")
results<-subset(results,p < 0.05)

CARD_Fecal<-subset(CARD_Fecal,ARO%in%results$ARO)



ggplot(CARD_Fecal,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(~ARO)















VFDB<-read.table("CAR",header = TRUE)



Fecal_meatadata<-data.frame(sample_data(LungCancer_Normalized_Fecal))

VFDB_Fecal<-VFDB[,colnames(VFDB)%in%(Fecal_meatadata$ID)]

VFDB_Fecal$ARO<-VFDB$ID


VFDB_Fecal<-melt(VFDB_Fecal,id.vars ="ARO")

VFDB_Fecal<-merge(VFDB_Fecal,Fecal_meatadata,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=VFDB_Fecal,group.by = "ARO")
results<-subset(results,p < 0.05)

VFDB_Fecal<-subset(VFDB_Fecal,ARO%in%results$ARO)



ggplot(VFDB_Fecal,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()+facet_wrap(~ARO)
```



# 中性模型的优化

```{r}
BS_midmod_best<-subset(BS_midmod,fit_class=="As predicted")

BN_midmod_best<-subset(BN_midmod,fit_class=="As predicted")

BO_midmod_best<-subset(BO_midmod,fit_class=="As predicted")
SS<-BS_midmod_best$Genus
SN<-setdiff(BN_midmod_best$Genus,BS_midmod_best$Genus)
SO<-setdiff(BO_midmod_best$Genus,BS_midmod_best$Genus)
idx<-pair$index.x

name<-sample_names(subset_samples(LungCancer_Normalized,index%in%idx))

Nasal_pool_best<-Nasal_pool[rownames(Nasal_pool)%in%SN,]
Nasal_pool_best<-Nasal_pool_best[,colnames(Nasal_pool_best)%in%name]

Oroph_pool_best<-Oroph_pool[rownames(Oroph_pool)%in%SO,]
Oroph_pool_best<-Oroph_pool_best[,colnames(Oroph_pool_best)%in%name]
Saliva_pool_best<-Saliva_pool[rownames(Saliva_pool)%in%SS,]
Saliva_pool_best<-Saliva_pool_best[,colnames(Saliva_pool_best)%in%name]

colnames(Saliva_pool_best)<-idx

colnames(Oroph_pool_best)<-idx

colnames(Nasal_pool_best)<-idx
best_pool<-rbind.data.frame(Saliva_pool_best,Oroph_pool_best,Nasal_pool_best,Saliva_pool_ab)



SS_ab<-setdiff(SS_ab,rownames(best_pool))
Saliva_pool_ab<-Saliva_pool[rownames(Saliva_pool)%in%SS_ab,]
Saliva_pool_ab<-Saliva_pool_ab[,colnames(Saliva_pool_ab)%in%name]
colnames(Saliva_pool_ab)<-idx

BAL_spp<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%BAL_Interindividual$ID]


spp <-t(BAL_spp)
pool <-t(best_pool)

tax_id<-data.frame(row.names(OTU_TAX_Genus))
row.names(tax_id)<-tax_id$row.names.OTU_TAX_Genus.
colnames(tax_id)<-"Genus"
spp.out <- fit_sncm(spp, pool=pool, taxon=tax_id)
BEST_MODEL <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())



BEST_midmod<-spp.out$predictions

```










```{r}
BN_unit<-data.frame(setdiff(BN_midmod$Genus,union(BS_midmod$Genus,BO_midmod$Genus)))
BN_unit$site<-"Nasal"
BS_unit<-data.frame(setdiff(BS_midmod$Genus,union(BN_midmod$Genus,BO_midmod$Genus)))
BS_unit$site<-"Saliva"
BO_unit<-data.frame(setdiff(BO_midmod$Genus,union(BN_midmod$Genus,BS_midmod$Genus)))
BO_unit$site<-"Oroph"
OS_unit<-data.frame(setdiff(intersect(BS_midmod$Genus,BO_midmod$Genus),BN_midmod$Genus))
row.names(OS_unit)<-OS_unit$setdiff.intersect.BS_midmod.Genus..BO_midmod.Genus...BN_midmod.Genus.
OS_unit$site<-"no"
for (i in OS_unit$setdiff.intersect.BS_midmod.Genus..BO_midmod.Genus...BN_midmod.Genus.){
  if (abs(BS_midmod[i,]$freq.pred/BS_midmod[i,]$freq-1)>abs(BO_midmod[i,]$freq.pred/BO_midmod[i,]$freq-1))
    OS_unit[i,2]<-"Oroph"
  else
    OS_unit[i,2]<-"Saliva"
  
} 


ON_unit<-data.frame(setdiff(intersect(BN_midmod$Genus,BO_midmod$Genus),BS_midmod$Genus))
row.names(ON_unit)<-ON_unit$setdiff.intersect.BN_midmod.Genus..BO_midmod.Genus...BS_midmod.Genus.
ON_unit$site<-"no"
for (i in ON_unit$setdiff.intersect.BN_midmod.Genus..BO_midmod.Genus...BS_midmod.Genus.){
  if (abs(BN_midmod[i,]$freq.pred/BN_midmod[i,]$freq-1)>abs(BO_midmod[i,]$freq.pred/BO_midmod[i,]$freq-1))
    ON_unit[i,2]<-"Oroph"
  else
    ON_unit[i,2]<-"Nasal"
} 



SN_unit<-data.frame(setdiff(intersect(BN_midmod$Genus,BS_midmod$Genus),BO_midmod$Genus))
row.names(SN_unit)<-SN_unit$setdiff.intersect.BN_midmod.Genus..BS_midmod.Genus...BO_midmod.Genus.
SN_unit$site<-"no"
for (i in SN_unit$setdiff.intersect.BN_midmod.Genus..BS_midmod.Genus...BO_midmod.Genus.){
  if (abs(BN_midmod[i,]$freq.pred/BN_midmod[i,]$freq-1)>abs(BS_midmod[i,]$freq.pred/BS_midmod[i,]$freq-1))
    SN_unit[i,2]<-"Saliva"
  else
    SN_unit[i,2]<-"Nasal"
} 





SNO_unit<-data.frame(intersect(intersect(BN_midmod$Genus,BS_midmod$Genus),BO_midmod$Genus))
row.names(SNO_unit)<-SNO_unit$intersect.intersect.BN_midmod.Genus..BS_midmod.Genus...BO_midmod.Genus.
SNO_unit$site<-"no"
for (i in SNO_unit$intersect.intersect.BN_midmod.Genus..BS_midmod.Genus...BO_midmod.Genus.){
  if (abs(BN_midmod[i,]$freq.pred/BN_midmod[i,]$freq-1)>abs(BS_midmod[i,]$freq.pred/BS_midmod[i,]$freq-1))
   if (abs(BO_midmod[i,]$freq.pred/BO_midmod[i,]$freq-1)>abs(BS_midmod[i,]$freq.pred/BS_midmod[i,]$freq-1))
     SNO_unit[i,2]<-"Saliva"
   else
     SNO_unit[i,2]<-"Oroph"
  else
  if (abs(BO_midmod[i,]$freq.pred/BO_midmod[i,]$freq-1)>abs(BN_midmod[i,]$freq.pred/BS_midmod[i,]$freq-1))
    SNO_unit[i,2]<-"Nasal"
  else
    SNO_unit[i,2]<-"Oroph"
     
} 


# 合并所有的最优解
# 统一列名
colnames(SNO_unit)<-c("Genus","site")
colnames(SN_unit)<-c("Genus","site")
colnames(OS_unit)<-c("Genus","site")
colnames(ON_unit)<-c("Genus","site")
colnames(BN_unit)<-c("Genus","site")
colnames(BO_unit)<-c("Genus","site")
colnames(BS_unit)<-c("Genus","site")
best_site<-rbind.data.frame(SNO_unit,SN_unit,OS_unit,ON_unit,BN_unit,BO_unit,BS_unit)
best_site_Saliva<-subset(best_site,site=="Saliva")
best_site_Nasal<-subset(best_site,site=="Nasal")
best_site_Oroph<-subset(best_site,site=="Oroph")

Oroph_pool_best<-Oroph_pool[rownames(Oroph_pool)%in%best_site_Oroph$Genus,]
Saliva_pool_best<-Saliva_pool[rownames(Saliva_pool)%in%best_site_Saliva$Genus,]
Nasal_pool_best<-Nasal_pool[rownames(Nasal_pool)%in%best_site_Nasal$Genus,]


Nasal_pool_best<-Nasal_pool_best[,colnames(Nasal_pool_best)%in%name]

Oroph_pool_best<-Oroph_pool_best[,colnames(Oroph_pool_best)%in%name]

Saliva_pool_best<-Saliva_pool_best[,colnames(Saliva_pool_best)%in%name]

colnames(Saliva_pool_best)<-idx

colnames(Oroph_pool_best)<-idx

colnames(Nasal_pool_best)<-idx
best_pool<-rbind.data.frame(Saliva_pool_best,Oroph_pool_best,Nasal_pool_best)

BAL_spp<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%BAL_Interindividual$ID]



# 统计BAL中存在的物种树
table<-BAL_spp[rowSums(BAL_spp)>0,]



spp <-t(BAL_spp)
pool <-t(best_pool)

tax_id<-data.frame(row.names(OTU_TAX_Genus))
row.names(tax_id)<-tax_id$row.names.OTU_TAX_Genus.
colnames(tax_id)<-"Genus"
spp.out <- fit_sncm(spp, pool=pool, taxon=tax_id)
BEST_MODEL <- plot_sncm_fit(spp.out)+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme( #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())

data<-BEST_MODEL$data
data<-merge(data,best_site,by="Genus")

BEST_MODEL $data<-data
BEST_MODEL$data$site<-as.factor(BEST_MODEL$data$site)

BEST_midmod<-spp.out$predictions






ggplot(data,aes(x=site,y=freq))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=site),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("BAL")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("Shannon Index")+xlab("")






number<-c(84,170,58)
class<-c("Above prediction","As predicted","Below prediction")

data<-data.frame(number,class)

ggplot(data=data, mapping=aes(x=class,y=number,fill=class))+
    geom_bar(stat="identity")+scale_fill_manual(name = "Prediction", values = c(`Above prediction` = "cyan4", `As predicted` = "blue", `Below prediction` = "goldenrod"))+theme_bw()+theme(axis.text.x=element_text(colour="black",family="Times",size=14,angle =30,vjust = 0.6))
```



#四部位DMM
```{r}

pair_idx<-pair$index.x

sampel_BAL<-sample.names(subset_samples(LungCancer_Normalized,(index%in%pair_idx)&sample=="BAL"))
table_BAL<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sampel_BAL]
table_BAL$Genus<-rownames(table_BAL)
table_BAL$site<-"BAL"

table_BAL<-tidyr::unite(table_BAL, "Site_Genus", site,Genus,remove=FALSE)

rownames(table_BAL)<-table_BAL$Site_Genus
table_BAL<-table_BAL[,-65:-67]
colnames(table_BAL)<-pair_idx











pair_idx<-pair$index.x

sampel_Oroph<-sample.names(subset_samples(LungCancer_Normalized,(index%in%pair_idx)&sample=="Oroph"))
table_Oroph<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sampel_Oroph]
table_Oroph$Genus<-rownames(table_Oroph)
table_Oroph$site<-"Oroph"

table_Oroph<-tidyr::unite(table_Oroph, "Site_Genus", site,Genus,remove=FALSE)

rownames(table_Oroph)<-table_Oroph$Site_Genus
table_Oroph<-table_Oroph[,-65:-67]
colnames(table_Oroph)<-pair_idx





pair_idx<-pair$index.x

sampel_Saliva<-sample.names(subset_samples(LungCancer_Normalized,(index%in%pair_idx)&sample=="Saliva"))
table_Saliva<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sampel_Saliva]
table_Saliva$Genus<-rownames(table_Saliva)
table_Saliva$site<-"Saliva"

table_Saliva<-tidyr::unite(table_Saliva, "Site_Genus", site,Genus,remove=FALSE)

rownames(table_Saliva)<-table_Saliva$Site_Genus
table_Saliva<-table_Saliva[,-65:-67]
colnames(table_Saliva)<-pair_idx




pair_idx<-pair$index.x

sampel_Nasal<-sample.names(subset_samples(LungCancer_Normalized,(index%in%pair_idx)&sample=="Nasal"))
table_Nasal<-OTU_TAX_Genus_RA[,colnames(OTU_TAX_Genus_RA)%in%sampel_Nasal]
table_Nasal$Genus<-rownames(table_Nasal)
table_Nasal$site<-"Nasal"

table_Nasal<-tidyr::unite(table_Nasal, "Site_Genus", site,Genus,remove=FALSE)

rownames(table_Nasal)<-table_Nasal$Site_Genus
table_Nasal<-table_Nasal[,-65:-67]
colnames(table_Nasal)<-pair_idx


table_four_site<-rbind.data.frame(table_BAL,table_Oroph,table_Saliva,table_Nasal)

table_four_site<-rbind.data.frame(table_BAL,table_Oroph,table_Saliva)

count<-table_four_site
count<-t(count[which(rowSums(count) >80),])

set.seed(1024)
fit <- lapply(1:5, dmn, count = count, verbose=TRUE)

lplc <- sapply(fit, laplace) # AIC / BIC / Laplace
aic  <- sapply(fit, AIC) # AIC / BIC / Laplace
bic  <- sapply(fit, BIC) # AIC / BIC / Laplace




best <- fit[[which.min(lplc)]]
mixturewt(best)
ass <- apply(mixture(best), 1, which.max)
best 


  d <- melt(fitted(best))
  colnames(d) <- c("OTU", "cluster", "value")
  d1 <- subset(d, cluster == 1) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d1 <-d1[31:51,] 
  d1<- mutate(d1,OTU = factor(OTU, levels = unique(OTU)))
  d2 <- subset(d, cluster == 2) %>%
     # Arrange OTUs by assignment strength
     arrange(value) 
  d2 <-d2[31:51,]
  d2<- mutate(d2,OTU = factor(OTU, levels = unique(OTU)))
p1 <- ggplot(d1, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 1))

    

p2 <- ggplot(d2, aes(x = OTU, y = value)) +
       geom_bar(stat = "identity") +
       coord_flip() +
       labs(title = paste("Top drivers: community type", 2))


ass<-data.frame(ass)
ass$CT<-pair$CT.x


table(ass$CT,ass$ass)
```





# Pre.meta 泛基因组分析
```{r}

# 读取泛基因组表格

panphlan<-read.table("result_gene_presence_absence.tsv",header = TRUE,row.names = 1)

panphlan<-panphlan[rownames(panphlan)%in%results.sig$Metacyc,]

panphlan_annot<-read.table("panphlan_Prevotella_melaninogenica_annot.tsv",header = TRUE)


panphlan_annot<-subset(panphlan_annot,NR90%in%results.sig$Metacyc)

rownames(panphlan_annot)<-panphlan_annot$Metacyc
# 筛选合适的唾液样本

panphlan<-panphlan[,colnames(panphlan)%in%Saliva_CT$ID]

# 筛选metadata

panphlan_meta<-subset(new_MetaTable,ID%in%colnames(panphlan))

rownames(panphlan_meta)<-panphlan_meta$ID
# 构建phyloseq对象

panphlan = otu_table(panphlan, taxa_are_rows = TRUE)
panphlan_meta = sample_data(panphlan_meta)

panphlan_phyloseq<- phyloseq(panphlan,panphlan_annot,panphlan_meta)








alpha_meas = c("InvSimpson", "Observed","Shannon")
alpha_meas = c("Observed")



p<-plot_richness(panphlan_phyloseq, "group", "group",nrow = 1,measures=alpha_meas)


P_BAL <- ggplot(data=p$data,aes(x=CT,y=value,fill=CT))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()











set.seed(1024)
#BAL

## 计算Bray_curtis距离矩阵


dis_bray<- phyloseq::distance(panphlan_phyloseq, "bray")

## 采用PCoA的方法对距离矩阵进行降维
dis_bray.pcoa = ordinate(panphlan_phyloseq, method="PCoA", distance=dis_bray)

## 绘制初始图形
bray.pcoa <- plot_ordination(panphlan_phyloseq, dis_bray.pcoa, color="CT" ) + geom_point(size=3)

## 提取图形数据
data<-bray.pcoa$data

## 修改列名
colnames(data)[1:2]<-c("PC1","PC2")

## 获取主坐标轴1,2的解释度
pc1<-17
pc2<-7


#用于填充样本点的颜色
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442")
#样本点的边框颜色
Palette <- c("#000000", "#000000","#000000","#000000")
#用于样本点的形状，注意：有一些样本点形状边框和填充的颜色是一起的不能独立
pich=c(21,21)

#匹配横纵坐标
ggscatter(data, x = "PC1", y = "PC2",
          color = "CT",
          ellipse = FALSE, mean.point = TRUE,
          star.plot = TRUE)

set.seed(2020)
otu.adonis<-adonis(dis_bray ~ CT, data = data.frame(sample_data(panphlan_phyloseq)))




data<-panphlan
data$Metacyc<-rownames(data)


data<-melt(data,id.vars ="Metacyc")

data<-merge(data,panphlan_meta,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=data,group.by = "Metacyc")
results.sig<-subset(results,p < 0.05)








ha_1 = HeatmapAnnotation(CT =panphlan_meta$CT,
                       col = list(CT = c("type1" = "#F8766D", "type2" = "#00BFC4")),
                       na_col = "black")



col_fun = colorRamp2(c(0,1), c("yellow", "red"))
col_fun(seq(-3, 3))

Heatmap(panphlan[rownames(anphlan)%in%results.sig$Metacyc,],name = "Z-score",top_annotation = ha_1,column_order =order(panphlan_meta$CT),col = col_fun,show_column_names = FALSE)



```





# 网络分析


```{r}

# 挑选BAL的样本
OTU_TAX_Genus_BAL<-OTU_TAX_Genus[,colnames(OTU_TAX_Genus)%in%sample.names(LungCancer_Normalized_BAL)]
# 过滤低丰度细菌
OTU_TAX_Genus_BAL<-OTU_TAX_Genus_BAL[which(rowSums(OTU_TAX_Genus_BAL) >1000),]

idx<-rownames(OTU_TAX_Genus_BAL)
BAL_sparcc_table_type1<-OTU_TAX_Genus_BAL[,colnames(OTU_TAX_Genus_BAL)%in%sample_names(LungCancer_Normalized_BAL_type1)]
BAL_sparcc_table_type2<-OTU_TAX_Genus_BAL[,colnames(OTU_TAX_Genus_BAL)%in%sample_names(LungCancer_Normalized_BAL_type2)]


write.table(BAL_sparcc_table_type1,"sparcc/BAL_sparcc_table_type1.txt",sep = "\t")
write.table(BAL_sparcc_table_type2,"sparcc/BAL_sparcc_table_type2.txt",sep = "\t")

cor_sparcc <- read.table('sparcc/results/BAL_sparcc_table_type1.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/BAL_type1_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj.BAL.type1 <- as.matrix(cor_sparcc) * as.matrix(pvals)
adj.BAL.type1[!(adj.BAL.type1==0)]<-1
adj.BAL.type1<-data.frame(adj.BAL.type1)




cor_sparcc <- read.table('sparcc/results/Saliva_sparcc_table_type1.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/Saliva_type1_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj.Saliva.type1  <- as.matrix(cor_sparcc) * as.matrix(pvals)
adj.Saliva.type1[!(adj.Saliva.type1==0)]<-1
adj.Saliva.type1<-data.frame(adj.Saliva.type1)


counts<-data.frame(colSums(adj.BAL.type1),colSums(adj.Saliva.type1))

counts<-subset(counts,colSums.adj.BAL.type1.>=0&colSums.adj.Saliva.type1.>=0)
name<-rownames(counts)

distance<-c()
for (i in name){
  table<-cbind.data.frame(adj.BAL.type1[,i],adj.Saliva.type1[,i])
  value<-matrix(proxy::dist(t(table),"Jaccard"))
  
  distance[i]<-value[1,1]
  
}

distance_type1<-data.frame(distance)














cor_sparcc <- read.table('sparcc/results/BAL_sparcc_table_type2.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/BAL_type2_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj.BAL.type2 <- as.matrix(cor_sparcc) * as.matrix(pvals)
adj.BAL.type2[!(adj.BAL.type2==0)]<-1
adj.BAL.type2<-data.frame(adj.BAL.type2)




cor_sparcc <- read.table('sparcc/results/Saliva_sparcc_table_type2.txt_cor_sparcc.out.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
pvals <- read.table('sparcc/results/Saliva_type2_pvals.txt', row.names = 1, sep = '\t', header = TRUE,check.names = FALSE)
cor_sparcc[abs(cor_sparcc) < 0.4] <- 0
pvals[pvals>=0.05] <- -1
pvals[pvals<0.05 & pvals>=0] <- 1
pvals[pvals==-1] <- 0
adj.Saliva.type2  <- as.matrix(cor_sparcc) * as.matrix(pvals)
adj.Saliva.type2[!(adj.Saliva.type2==0)]<-1
adj.Saliva.type2<-data.frame(adj.Saliva.type2)


counts<-data.frame(colSums(adj.BAL.type2),colSums(adj.Saliva.type2))

counts<-subset(counts,colSums.adj.BAL.type2.>=0&colSums.adj.Saliva.type2.>=0)
name<-rownames(counts)

distance<-c()
for (i in name){
  table<-cbind.data.frame(adj.BAL.type2[,i],adj.Saliva.type2[,i])
  value<-matrix(proxy::dist(t(table),"Jaccard"))
  
  distance[i]<-value[1,1]
  
}

distance_type2<-data.frame(distance)

distance_type1$group<-"type1"
distance_type2$group<-"type2"
distance<-rbind.data.frame(distance_type1,distance_type2)

ggplot(distance,aes(x=group,y=distance,fill=group))+ #”fill=“设置填充颜色
  stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
  geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
  geom_jitter(aes(fill=group),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
  scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
  ggtitle("")+ #设置总的标题
  theme_bw()+ #背景变为白色
  theme(legend.position="none", #不需要图例
        axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
        axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
        axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
        axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
        plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
        panel.grid.major = element_blank(), #不显示网格线
        panel.grid.minor = element_blank())+
  ylab("")+xlab("")+ #设置x轴和y轴的标题
  stat_compare_means()
```


# 读取SNVs数据
```{r}

Pre_SNV<-read.table("Pre.count.txt",header = TRUE,row.names = 1)
Str_SNV<-read.table("Str.count.txt",header = TRUE,row.names = 1)
Vel_SNV<-read.table("Vel.count.txt",header = TRUE,row.names = 1)





Pre_SNV$ID<-row.names(Pre_SNV)

Pre_SNV<-merge(Pre_SNV,new_MetaTable,by="ID")


Pre_SNV<-subset(Pre_SNV,!(CT=="NA"))
ggplot(Pre_SNV,aes(x=CT,y=nucl_diversity,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()









Str_SNV$ID<-row.names(Str_SNV)

Str_SNV<-merge(Str_SNV,new_MetaTable,by="ID")


Str_SNV<-subset(Str_SNV,!(CT=="NA"))
ggplot(Str_SNV,aes(x=CT,y=nucl_diversity,fill=CT))+ #”fill=“设置填充颜色
stat_boxplot(geom = "errorbar",width=0.15,aes(color="black"))+ #由于自带的箱形图没有胡须末端没有短横线，使用误差条的方式补上
geom_boxplot(size=0.5,fill="white",outlier.fill="white",outlier.color="white")+ #size设置箱线图的边框线和胡须的线宽度，fill设置填充颜色，outlier.fill和outlier.color设置异常点的属性
geom_jitter(aes(fill=CT),width =0.2,shape = 21,size=2.5)+ #设置为向水平方向抖动的散点图，width指定了向水平方向抖动，不改变纵轴的值
  #scale_fill_manual(values = c("#E69F00", "#0072B2","#F0E442"))+  #设置填充的颜色
scale_color_manual(values=c("black","black","black"))+ #设置散点图的圆圈的颜色为黑色
ggtitle("")+ #设置总的标题
theme_bw()+ #背景变为白色
theme(legend.position="none", #不需要图例
axis.text.x=element_text(colour="black",family="Times",size=14), #设置x轴刻度标签的字体属性
axis.text.y=element_text(family="Times",size=14,face="plain"), #设置x轴刻度标签的字体属性
axis.title.y=element_text(family="Times",size = 14,face="plain"), #设置y轴的标题的字体属性
axis.title.x=element_text(family="Times",size = 14,face="plain"), #设置x轴的标题的字体属性
plot.title = element_text(family="Times",size=15,face="bold",hjust = 0.5), #设置总标题的字体属性
panel.grid.major = element_blank(), #不显示网格线
panel.grid.minor = element_blank())+
ylab("")+xlab("")+ #设置x轴和y轴的标题
stat_compare_means()



data<-all_SNP_Pre


data<-data[rowSums(data)>5,]
data$Metacyc<-rownames(data)


data<-melt(data,id.vars ="Metacyc")

data<-merge(data,new_MetaTable,by.x="variable",by.y="ID")



results<-compare_means(value~CT, data=data,group.by = "Metacyc")
results.sig<-subset(results,p < 0.05)


all_SNP<-read.table("all.SNV.rst.tsv")



all_SNP_Pre<-subset(all_SNP,group=="Pre")


all_SNP_Pre<-all_SNP_Pre[,-1]

all_SNP_Pre<-all_SNP_Pre[,colnames(all_SNP_Pre)%in%all_SNP_Pre_meta$ID]

all_SNP_Pre_meta<-subset(new_MetaTable,ID%in%colnames(all_SNP_Pre))
all_SNP_Pre_meta<-subset(all_SNP_Pre_meta,!(CT=="NA"))


ha_1 = HeatmapAnnotation(CT =all_SNP_Pre_meta$CT,
                       col = list(CT = c("type1" = "#F8766D", "type2" = "#00BFC4")),
                       na_col = "black")



col_fun = colorRamp2(c(0,1), c("yellow", "red"))
col_fun(seq(-3, 3))


all_SNP_Pre_sig<-all_SNP_Pre[row.names(all_SNP_Pre)%in%results.sig$Metacyc,]

Heatmap(all_SNP_Pre_sig,name = "Z-score",top_annotation = ha_1,column_order =order(all_SNP_Pre_meta$CT),col = col_fun,show_column_names = FALSE)
```



















